// @progress/kendo-vue-layout/dist/npm/expansionpanel/ExpansionPanelContent.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ExpansionPanelContentVue2 = exports.ExpansionPanelContent = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var package_metadata_1 = require("../package-metadata");
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
/**
 * @hidden
 */
var ExpansionPanelContentVue2 = {
  name: 'KendoExpansionPanelContent',
  created: function created() {
    (0, kendo_vue_common_1.validatePackage)(package_metadata_1.packageMetadata);
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = (0, kendo_vue_common_1.getDefaultSlots)(this);
    return h("div", {
      "class": 'k-expander-content-wrapper'
    }, [h("div", {
      "class": 'k-expander-content'
    }, [defaultSlots])]);
  }
};
exports.ExpansionPanelContentVue2 = ExpansionPanelContentVue2;
/**
 * @hidden
 */
var ExpansionPanelContent = ExpansionPanelContentVue2;
exports.ExpansionPanelContent = ExpansionPanelContent;
