// @progress/kendo-vue-layout/dist/npm/expansionpanel/ExpansionPanel.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ExpansionPanelVue2 = exports.ExpansionPanel = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
var kendo_svg_icons_1 = require("@progress/kendo-svg-icons");
var package_metadata_1 = require("../package-metadata");
/**
 * @hidden
 */
var ExpansionPanelVue2 = {
  name: 'KendoExpansionPanel',
  props: {
    dir: String,
    id: String,
    tabIndex: Number,
    title: String,
    subtitle: String,
    titleRender: [String, Object, Function],
    subtitleRender: [String, Object, Function],
    expandIcon: String,
    expandSvgIcon: Object,
    collapseIcon: String,
    collapseSvgIcon: Object,
    expanded: Boolean,
    disabled: Boolean
  },
  created: function created() {
    (0, kendo_vue_common_1.validatePackage)(package_metadata_1.packageMetadata);
  },
  computed: {
    wrapperClass: function wrapperClass() {
      var _a = this.$props,
        expanded = _a.expanded,
        disabled = _a.disabled;
      return {
        'k-expander': true,
        'k-expanded': expanded,
        'k-focus': this.focused && !disabled,
        'k-disabled': disabled
      };
    },
    expandIconName: function expandIconName() {
      var _a = this.$props,
        expanded = _a.expanded,
        expandIcon = _a.expandIcon,
        collapseIcon = _a.collapseIcon;
      return expanded ? expandIcon ? (0, kendo_vue_common_1.getIconName)(expandIcon) : 'chevron-up' : collapseIcon ? (0, kendo_vue_common_1.getIconName)(collapseIcon) : 'chevron-down';
    },
    expandSVGIcon: function expandSVGIcon() {
      var _a = this.$props,
        expanded = _a.expanded,
        expandSvgIcon = _a.expandSvgIcon,
        collapseSvgIcon = _a.collapseSvgIcon;
      return expanded ? expandSvgIcon || kendo_svg_icons_1.chevronUpIcon : collapseSvgIcon || kendo_svg_icons_1.chevronDownIcon;
    }
  },
  data: function data() {
    return {
      focused: false
    };
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = (0, kendo_vue_common_1.getDefaultSlots)(this);
    var _a = this.$props,
      _b = _a.expanded,
      expanded = _b === void 0 ? false : _b,
      disabled = _a.disabled,
      title = _a.title,
      titleRender = _a.titleRender,
      subtitle = _a.subtitle,
      subtitleRender = _a.subtitleRender,
      ariaControls = _a.ariaControls,
      tabIndex = _a.tabIndex,
      id = _a.id;
    var titleTemplate = kendo_vue_common_1.templateRendering.call(this, titleRender, kendo_vue_common_1.getListeners.call(this));
    var titleToRender = kendo_vue_common_1.getTemplate.call(this, {
      h: h,
      template: titleTemplate,
      defaultRendering: h("div", {
        "class": "k-expander-title"
      }, [title])
    });
    var subtitleTemplate = kendo_vue_common_1.templateRendering.call(this, subtitleRender, kendo_vue_common_1.getListeners.call(this));
    var subtitleToRender = kendo_vue_common_1.getTemplate.call(this, {
      h: h,
      template: subtitleTemplate,
      defaultRendering: h("div", {
        "class": "k-expander-sub-title"
      }, [subtitle])
    });
    return h("div", {
      "class": this.wrapperClass,
      onFocusin: this.handleFocus,
      on: this.v3 ? undefined : {
        "focusin": this.handleFocus,
        "focusout": this.handleBlur,
        "keydown": this.onKeyDown
      },
      onFocusout: this.handleBlur,
      id: id
      // dir={useRtl(elementRef, dir)}
      ,
      attrs: this.v3 ? undefined : {
        id: id
      },
      onKeydown: this.onKeyDown
    }, [h("div", {
      role: 'button',
      attrs: this.v3 ? undefined : {
        role: 'button',
        "aria-controls": ariaControls,
        "aria-expanded": expanded,
        "aria-disabled": disabled,
        tabIndex: (0, kendo_vue_common_1.getTabIndex)(tabIndex, disabled)
      },
      ref: 'header',
      "aria-controls": ariaControls,
      "aria-expanded": expanded,
      "aria-disabled": disabled,
      tabIndex: (0, kendo_vue_common_1.getTabIndex)(tabIndex, disabled),
      "class": "k-expander-header",
      onClick: this.onClick,
      on: this.v3 ? undefined : {
        "click": this.onClick
      }
    }, [titleToRender, h("span", {
      "class": "k-spacer"
    }), subtitleToRender, h("span", {
      "class": "k-expander-indicator"
    }, [h(kendo_vue_common_1.Icon, {
      name: this.expandIconName,
      attrs: this.v3 ? undefined : {
        name: this.expandIconName,
        icon: this.expandSVGIcon
      },
      icon: this.expandSVGIcon
    })])]), defaultSlots]);
  },
  methods: {
    onClick: function onClick(event) {
      if (this.$props.disabled) {
        return;
      }
      this.$emit('action', {
        event: event,
        expanded: this.expanded
      });
    },
    onKeyDown: function onKeyDown(event) {
      if (this.$props.disabled) {
        return;
      }
      if ((event.keyCode === kendo_vue_common_1.Keys.enter || event.keyCode === kendo_vue_common_1.Keys.space) && event.target === this.$refs.header) {
        event.preventDefault();
        this.$emit('action', {
          event: event,
          expanded: this.expanded
        });
      }
    },
    handleFocus: function handleFocus() {
      if (this.$props.disabled) {
        return;
      }
      this.focused = true;
    },
    handleBlur: function handleBlur() {
      if (this.$props.disabled) {
        return;
      }
      this.focused = false;
    }
  }
};
exports.ExpansionPanelVue2 = ExpansionPanelVue2;
/**
 * @hidden
 */
var ExpansionPanel = ExpansionPanelVue2;
exports.ExpansionPanel = ExpansionPanel;
