// @progress/kendo-vue-grid/dist/esm/columnMenu/GridColumnMenuItemContent.js
// @ts-ignore
import * as Vue from 'vue';
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
import { getDefaultSlots } from '@progress/kendo-vue-common';
import { Reveal } from '@progress/kendo-vue-animation';
/**
 * @hidden
 */
var GridColumnMenuItemContentVue2 = {
  props: {
    show: Boolean
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  // @ts-ignore
  render: function render(createElement) {
    var _this = this;
    var h = gh || createElement;
    var defaultSlot = getDefaultSlots(this);
    return h("div", {
      "class": 'k-columnmenu-item-content'
    }, [
    // @ts-ignore function children
    h(Reveal, {
      appear: this.$props.show,
      attrs: this.v3 ? undefined : {
        appear: this.$props.show
      },
      style: {
        position: 'relative',
        display: 'block'
      }
    }, this.v3 ? function () {
      return [_this.$props.show ? defaultSlot : null];
    } : [_this.$props.show ? defaultSlot : null])]);
  }
};
/**
 * @hidden
 */
var GridColumnMenuItemContent = GridColumnMenuItemContentVue2;
export { GridColumnMenuItemContent, GridColumnMenuItemContentVue2 };
