// @progress/kendo-vue-inputs/dist/npm/input-separator/InputSeparator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InputSeparatorVue2 = exports.InputSeparator = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var package_metadata_1 = require("../package-metadata");
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
/**
 * @hidden
 */
var InputSeparatorVue2 = {
  created: function created() {
    (0, kendo_vue_common_1.validatePackage)(package_metadata_1.packageMetadata);
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    return h("span", {
      "class": 'k-input-separator'
    });
  }
};
exports.InputSeparatorVue2 = InputSeparatorVue2;
/**
 * Represents the Kendo UI for Vue InputSeparator component. It can be used as a semarator
 *  between horizontal icons in prefi and suffic templates.
 *
 *  ```jsx
 *  <template>
 *    <InputSeparator>
 *  </template>
 *  ```
 */
var InputSeparator = InputSeparatorVue2;
exports.InputSeparator = InputSeparator;
