// @progress/kendo-vue-data-tools/dist/es/filter/filters/TextFilter.js
var __assign = this && this.__assign || function () {
  __assign = Object.assign || function (t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
      s = arguments[i];
      for (var p in s) {
        if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
      }
    }
    return t;
  };
  return __assign.apply(this, arguments);
};
// @ts-ignore
import * as Vue from 'vue';
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
import { TextBox } from '@progress/kendo-vue-inputs';
import { provideLocalizationService } from '@progress/kendo-vue-intl';
import { filterTextFilterAriaLabel, messages } from '../../messages/main';
/**
 * @hidden
 */
var TextFilterVue2 = {
  name: 'kendoTextFilter',
  props: {
    filter: {
      type: Object,
      required: true
    },
    ariaLabel: String
  },
  inject: {
    kendoLocalizationService: {
      default: null
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var locService = provideLocalizationService(this);
    var _a = this.$props.ariaLabel,
      ariaLabel = _a === void 0 ? locService.toLanguageString(filterTextFilterAriaLabel, messages[filterTextFilterAriaLabel]) : _a;
    return (
      // @ts-ignore
      h(TextBox, {
        value: this.$props.filter.value || '',
        attrs: this.v3 ? undefined : {
          value: this.$props.filter.value || '',
          ariaLabel: ariaLabel
        },
        onInput: this.onChange,
        on: this.v3 ? undefined : {
          "input": this.onChange
        },
        ariaLabel: ariaLabel
      })
    );
  },
  methods: {
    onChange: function onChange(event) {
      this.$emit('filterchange', {
        nextFilter: __assign(__assign({}, this.$props.filter), {
          value: event.value
        })
      });
    }
  }
};
/**
 * @hidden
 */
var TextFilter = TextFilterVue2;
export { TextFilter, TextFilterVue2 };
