// @progress/kendo-vue-inputs/dist/npm/slider/Slider.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SliderVue2 = exports.Slider = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var ref = allVue.ref;
var inject = allVue.inject;
var kendo_vue_buttons_1 = require("@progress/kendo-vue-buttons");
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
var kendo_vue_intl_1 = require("@progress/kendo-vue-intl");
var main_1 = require("../messages/main");
var SliderLabel_1 = require("./SliderLabel");
var package_metadata_1 = require("../package-metadata");
var kendo_svg_icons_1 = require("@progress/kendo-svg-icons");
/**
 * @hidden
 */
var SliderVue2 = {
  name: 'KendoSlider',
  model: {
    event: 'changemodel'
  },
  // @ts-ignore
  emits: {
    'changemodel': null,
    'update:modelValue': null,
    'change': null,
    'blur': null,
    'focus': null
  },
  props: {
    modelValue: {
      type: Number,
      default: undefined
    },
    defaultValue: {
      type: Number,
      default: undefined
    },
    name: String,
    buttons: Boolean,
    tabIndex: Number,
    disabled: Boolean,
    dir: String,
    step: Number,
    min: {
      type: Number,
      required: true
    },
    max: {
      type: Number,
      required: true
    },
    value: Number,
    vertical: Boolean,
    id: String,
    ariaLabelledBy: String,
    ariaDescribedBy: String
  },
  provide: function provide() {
    return {
      kendoMin: this.$props.min,
      kendoMax: this.$props.max,
      kendoVertical: this.$props.vertical
    };
  },
  inject: {
    kendoLocalizationService: {
      default: null
    }
  },
  data: function data() {
    return {
      currentValue: undefined,
      currentFocused: false,
      currentDir: 'ltr'
    };
  },
  computed: {
    computedValue: function computedValue() {
      var value = this.$props.value !== undefined ? this.$props.value : this.currentValue;
      var _a = this.$props,
        min = _a.min,
        max = _a.max;
      return value === undefined ? value : Math.min(Math.max(value, min), max);
    },
    sliderTrack: function sliderTrack() {
      return this._sliderTrack;
    }
  },
  created: function created() {
    (0, kendo_vue_common_1.validatePackage)(package_metadata_1.packageMetadata);
    if (this.$props.value !== undefined) {
      this.$data.currentValue = this.$props.value;
    } else if (this.$props.modelValue !== undefined) {
      this.$data.currentValue = this.$props.modelValue;
    } else if (this.$props.defaultValue !== undefined) {
      this.$data.currentValue = this.$props.defaultValue;
    } else {
      this.$data.currentValue = this.$props.min;
    }
    this.currentFocused = false;
    this.currentDir = this.$props.dir;
  },
  mounted: function mounted() {
    this._sliderTrack = this.$refs.sliderTrack;
    if (this.$el) {
      this.draggable = this.$refs.draggable;
    }
    if (!this.currentDir && window && this.$el) {
      // Note: getComputedStyle forces reflow
      var direction = window.getComputedStyle(this.$el).direction;
      if (direction) {
        this.currentDir = direction;
      }
    }
  },
  updated: function updated() {
    if (this.$el) {
      this.draggable = this.$refs.draggable;
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    var inputRef = ref(null);
    var kendoLocalizationService = inject('kendoLocalizationService', {});
    return {
      v3: v3,
      inputRef: inputRef,
      kendoLocalizationService: kendoLocalizationService
    };
  },
  render: function render(createElement) {
    var _this = this;
    var _a;
    var h = gh || createElement;
    var lS = (0, kendo_vue_intl_1.provideLocalizationService)(this);
    var percentValue = (this.computedValue - this.$props.min) / (this.$props.max - this.$props.min) * 100;
    var defaultSlot = (0, kendo_vue_common_1.getDefaultSlots)(this);
    var vertical = this.$props.vertical;
    var increaseSvgArrow = this.currentDir === 'rtl' ? vertical ? kendo_svg_icons_1.caretAltUpIcon : kendo_svg_icons_1.caretAltLeftIcon : vertical ? kendo_svg_icons_1.caretAltUpIcon : kendo_svg_icons_1.caretAltRightIcon;
    var decreaseSvgArrow = this.currentDir === 'rtl' ? vertical ? kendo_svg_icons_1.caretAltDownIcon : kendo_svg_icons_1.caretAltRightIcon : vertical ? kendo_svg_icons_1.caretAltDownIcon : kendo_svg_icons_1.caretAltLeftIcon;
    var increaseFontArrow = this.currentDir === 'rtl' ? vertical ? 'caret-alt-up' : 'caret-alt-left' : vertical ? 'caret-alt-up' : 'caret-alt-right';
    var decreaseFontArrow = this.currentDir === 'rtl' ? vertical ? 'caret-alt-down' : 'caret-alt-right' : vertical ? 'caret-alt-down' : 'caret-alt-left';
    return h("div", {
      dir: this.currentDir,
      attrs: this.v3 ? undefined : {
        dir: this.currentDir,
        id: this.$props.id
      },
      id: this.$props.id,
      style: this.$props.style,
      onFocus: this.onFocus,
      on: this.v3 ? undefined : {
        "focus": this.onFocus,
        "blur": this.onBlur,
        "keydown": this.onKeyDown
      },
      onBlur: this.onBlur,
      onKeydown: this.onKeyDown,
      "class": (0, kendo_vue_common_1.classNames)('k-slider', {
        'k-focus': this.currentFocused,
        'k-disabled': this.$props.disabled,
        'k-slider-horizontal': !this.$props.vertical,
        'k-slider-vertical': this.$props.vertical
      }, this.$props.className)
    }, [h("div", {
      "class": (0, kendo_vue_common_1.classNames)('k-slider-wrap')
    }, [this.$props.buttons &&
    // @ts-ignore
    h(kendo_vue_buttons_1.Button, {
      type: "button",
      attrs: this.v3 ? undefined : {
        type: "button",
        tabindex: -1,
        icon: decreaseFontArrow,
        svgIcon: decreaseSvgArrow,
        rounded: 'full',
        title: lS.toLanguageString(main_1.sliderDecreaseValue, main_1.messages[main_1.sliderDecreaseValue])
      },
      tabindex: -1,
      icon: decreaseFontArrow,
      svgIcon: decreaseSvgArrow,
      style: {
        position: 'relative'
      },
      rounded: 'full',
      "class": "k-button-decrease",
      title: lS.toLanguageString(main_1.sliderDecreaseValue, main_1.messages[main_1.sliderDecreaseValue]),
      onClick: this.decrement,
      on: this.v3 ? undefined : {
        "click": this.decrement
      }
    }),
    // @ts-ignore  function children
    h(kendo_vue_common_1.Draggable, {
      onDrag: this.dragOver,
      on: this.v3 ? undefined : {
        "drag": this.dragOver,
        "press": this.dragStart
      },
      onPress: this.dragStart,
      ref: 'draggable'
    }, this.v3 ? function () {
      return [h("div", {
        "class": "k-slider-track-wrap",
        style: {
          touchAction: 'none'
        }
      }, [defaultSlot && h("ul", {
        "class": "k-reset k-slider-items"
      }, [defaultSlot]), h("div", {
        "class": "k-slider-track",
        ref: 'sliderTrack'
      }, [h("div", {
        "class": "k-slider-selection",
        style: _this.$props.vertical ? {
          height: percentValue + '%'
        } : {
          width: percentValue + '%'
        }
      }), h("a", {
        "class": "k-draghandle",
        role: "slider",
        attrs: _this.v3 ? undefined : {
          role: "slider",
          tabIndex: "0",
          "aria-valuenow": _this.computedValue,
          "aria-valuemin": _this.$props.min,
          "aria-valuemax": _this.$props.max,
          "aria-valuetext": _this.computedValue.toString(),
          "aria-disabled": _this.$props.disabled ? 'true' : undefined,
          title: lS.toLanguageString(main_1.sliderDragTitle, main_1.messages[main_1.sliderDragTitle])
        },
        tabIndex: "0",
        "aria-valuenow": _this.computedValue,
        "aria-valuemin": _this.$props.min,
        "aria-valuemax": _this.$props.max,
        "aria-valuetext": _this.computedValue.toString(),
        "aria-disabled": _this.$props.disabled ? 'true' : undefined,
        title: lS.toLanguageString(main_1.sliderDragTitle, main_1.messages[main_1.sliderDragTitle]),
        style: _this.$props.vertical ? {
          bottom: percentValue + '%',
          zIndex: 1
        } : (_a = {}, _a[_this.currentDir === 'rtl' ? 'right' : 'left'] = percentValue + '%', _a.zIndex = 1, _a)
      })])])];
    } : [h("div", {
      "class": "k-slider-track-wrap",
      style: {
        touchAction: 'none'
      }
    }, [defaultSlot && h("ul", {
      "class": "k-reset k-slider-items"
    }, [defaultSlot]), h("div", {
      "class": "k-slider-track",
      ref: 'sliderTrack'
    }, [h("div", {
      "class": "k-slider-selection",
      style: _this.$props.vertical ? {
        height: percentValue + '%'
      } : {
        width: percentValue + '%'
      }
    }), h("a", {
      "class": "k-draghandle",
      role: "slider",
      attrs: _this.v3 ? undefined : {
        role: "slider",
        tabIndex: "0",
        "aria-valuenow": _this.computedValue,
        "aria-valuemin": _this.$props.min,
        "aria-valuemax": _this.$props.max,
        "aria-valuetext": _this.computedValue.toString(),
        "aria-disabled": _this.$props.disabled ? 'true' : undefined,
        title: lS.toLanguageString(main_1.sliderDragTitle, main_1.messages[main_1.sliderDragTitle])
      },
      tabIndex: "0",
      "aria-valuenow": _this.computedValue,
      "aria-valuemin": _this.$props.min,
      "aria-valuemax": _this.$props.max,
      "aria-valuetext": _this.computedValue.toString(),
      "aria-disabled": _this.$props.disabled ? 'true' : undefined,
      title: lS.toLanguageString(main_1.sliderDragTitle, main_1.messages[main_1.sliderDragTitle]),
      style: _this.$props.vertical ? {
        bottom: percentValue + '%',
        zIndex: 1
      } : (_a = {}, _a[_this.currentDir === 'rtl' ? 'right' : 'left'] = percentValue + '%', _a.zIndex = 1, _a)
    })])])]), this.$props.buttons &&
    // @ts-ignore
    h(kendo_vue_buttons_1.Button, {
      type: "button",
      attrs: this.v3 ? undefined : {
        type: "button",
        tabindex: -1,
        icon: increaseFontArrow,
        svgIcon: increaseSvgArrow,
        rounded: 'full',
        title: lS.toLanguageString(main_1.sliderIncreaseValue, main_1.messages[main_1.sliderIncreaseValue])
      },
      tabindex: -1,
      icon: increaseFontArrow,
      svgIcon: increaseSvgArrow,
      rounded: 'full',
      style: {
        position: 'relative'
      },
      "class": "k-button-increase",
      title: lS.toLanguageString(main_1.sliderIncreaseValue, main_1.messages[main_1.sliderIncreaseValue]),
      onClick: this.increment,
      on: this.v3 ? undefined : {
        "click": this.increment
      }
    })])]);
  },
  methods: {
    focus: function focus() {
      if (this.$el) {
        // @ts-ignore
        this.$el.focus();
      }
    },
    isLabel: function isLabel(target) {
      var currentTarget = target;
      while (currentTarget) {
        if (Boolean(currentTarget.getAttribute(SliderLabel_1.SLIDER_LABEL_ATTRIBUTE))) {
          return true;
        }
        currentTarget = currentTarget.parentElement;
      }
      return false;
    },
    onFocus: function onFocus(event) {
      this.currentFocused = true;
      this.$emit('focus', event);
    },
    onBlur: function onBlur(event) {
      this.currentFocused = false;
      this.$emit('blur', event);
    },
    onKeyDown: function onKeyDown(e) {
      var newValue = undefined;
      if (e.keyCode === kendo_vue_common_1.Keys.left || e.keyCode === kendo_vue_common_1.Keys.down) {
        newValue = this.currentValue - (this.$props.step || 0);
      } else if (e.keyCode === kendo_vue_common_1.Keys.right || e.keyCode === kendo_vue_common_1.Keys.up) {
        newValue = this.currentValue + (this.$props.step || 0);
      } else if (e.keyCode === kendo_vue_common_1.Keys.home) {
        newValue = this.$props.min;
      } else if (e.keyCode === kendo_vue_common_1.Keys.end) {
        newValue = this.$props.max;
      }
      if (newValue !== undefined) {
        e.preventDefault();
        this.change(e, newValue);
      }
    },
    decrement: function decrement(e) {
      e.preventDefault();
      this.change(e, this.currentValue - (this.$props.step || 0));
    },
    increment: function increment(e) {
      e.preventDefault();
      this.change(e, this.currentValue + (this.$props.step || 0));
    },
    dragStart: function dragStart(e) {
      if (!this.isLabel(e.originalEvent.target)) {
        if (e.isTouch) {
          e.originalEvent.preventDefault();
        }
        this.drag(e);
      }
    },
    dragOver: function dragOver(e) {
      e.originalEvent.preventDefault();
      this.drag(e);
    },
    drag: function drag(e) {
      var computed = this.draggable.element.getBoundingClientRect();
      var distance = this.$props.vertical ? computed.bottom - e.clientY : this.currentDir === 'rtl' ? computed.right - e.clientX : e.clientX - computed.left;
      var size = this.$props.vertical ? computed.height : computed.width;
      var percentage = distance / size;
      this.change(e, this.$props.min + percentage * (this.$props.max - this.$props.min));
    },
    change: function change(e, value) {
      value = Math.min(Math.max(value, this.$props.min), this.$props.max);
      this.currentValue = value;
      this.$emit('changemodel', value);
      this.$emit('update:modelValue', value);
      this.$emit('change', {
        event: e,
        value: value,
        component: this,
        target: {
          name: this.$props.name,
          value: value
        }
      });
    }
  }
};
exports.SliderVue2 = SliderVue2;
/**
 * @hidden
 */
var Slider = SliderVue2;
exports.Slider = Slider;
