// @progress/kendo-vue-dateinputs/dist/npm/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.domContainerFactory = exports.dateInRange = exports.shiftWeekNames = exports.range = exports.isInSelectionRange = exports.isInDateRange = exports.isInRange = exports.getToday = exports.setTime = exports.isValidRange = exports.MAX_TIME = exports.MIN_TIME = exports.MAX_DATE = exports.MIN_DATE = exports.MIDNIGHT_DATE = exports.viewInRange = exports.isEqualRange = void 0;
var kendo_date_math_1 = require("@progress/kendo-date-math");
var kendo_date_math_2 = require("@progress/kendo-date-math");
var SelectionRange_1 = require("./calendar/models/SelectionRange");
/**
 * @hidden
 */
var isEqualRange = function (initial, updated) {
    var _a = initial || SelectionRange_1.EMPTY_SELECTIONRANGE, initialStart = _a.start, initialEnd = _a.end;
    var _b = updated || SelectionRange_1.EMPTY_SELECTIONRANGE, updatedStart = _b.start, updatedEnd = _b.end;
    if (initialStart === null || initialEnd === null || updatedStart === null || updatedEnd === null) {
        return false;
    }
    return (0, kendo_date_math_1.isEqual)(initialStart, updatedStart) && (0, kendo_date_math_1.isEqual)(initialEnd, updatedEnd);
};
exports.isEqualRange = isEqualRange;
/**
 * @hidden
 */
var viewInRange = function (candidate, min, max) {
    if (min === undefined || max === undefined) {
        return candidate;
    }
    return min <= candidate && candidate <= max
        ? candidate
        : candidate < min
            ? min
            : max;
};
exports.viewInRange = viewInRange;
/**
 * @hidden
 */
exports.MIDNIGHT_DATE = new Date(1980, 0, 1);
/**
 * @hidden
 */
exports.MIN_DATE = new Date(1900, 0, 1);
/**
 * @hidden
 */
exports.MAX_DATE = new Date(2099, 11, 31);
/**
 * @hidden
 */
exports.MIN_TIME = new Date(1980, 0, 1);
/**
 * @hidden
 */
exports.MAX_TIME = new Date(1980, 0, 1, 23, 59, 59);
var isSet = function (value) { return value !== null && value !== undefined; };
/**
 * @hidden
 */
var isValidRange = function (min, max) { return (!isSet(min) || !isSet(max) || min <= max); };
exports.isValidRange = isValidRange;
/**
 * @hidden
 */
var setTime = function (origin, candidate) {
    var date = (0, kendo_date_math_1.cloneDate)(origin);
    date.setHours(candidate.getHours(), candidate.getMinutes(), candidate.getSeconds(), candidate.getMilliseconds());
    return date;
};
exports.setTime = setTime;
/**
 * @hidden
 */
var getToday = function () { return (0, kendo_date_math_2.getDate)(new Date()); };
exports.getToday = getToday;
/**
 * @hidden
 */
var isInRange = function (candidate, min, max) { return (!candidate || !((min && min > candidate) || (max && max < candidate))); };
exports.isInRange = isInRange;
/**
 * @hidden
 */
var isInDateRange = function (candidate, min, max) { return (candidate === null
    || !((min && (0, kendo_date_math_2.getDate)(min) > (0, kendo_date_math_2.getDate)(candidate))
        || (max && (0, kendo_date_math_2.getDate)(max) < (0, kendo_date_math_2.getDate)(candidate)))); };
exports.isInDateRange = isInDateRange;
/**
 * @hidden
 */
var isInSelectionRange = function (value, selectionRange) {
    var _a = selectionRange || SelectionRange_1.EMPTY_SELECTIONRANGE, start = _a.start, end = _a.end;
    if (!start || !end) {
        return false;
    }
    return start < value && value < end;
};
exports.isInSelectionRange = isInSelectionRange;
/**
 * @hidden
 */
var range = function (start, end, step) {
    if (step === void 0) { step = 1; }
    var result = [];
    for (var i = start; i < end; i = i + step) {
        result.push(i);
    }
    return result;
};
exports.range = range;
/**
 * @hidden
 */
var shiftWeekNames = function (names, offset) { return (names.slice(offset).concat(names.slice(0, offset))); };
exports.shiftWeekNames = shiftWeekNames;
/**
 * @hidden
 */
var dateInRange = function (candidate, min, max) {
    if (!candidate) {
        return candidate;
    }
    if (min && candidate < min) {
        return (0, kendo_date_math_1.cloneDate)(min);
    }
    if (max && candidate > max) {
        return (0, kendo_date_math_1.cloneDate)(max);
    }
    return candidate;
};
exports.dateInRange = dateInRange;
/**
 * @hidden
 */
var domContainerFactory = function (type) { return function (children, classes, styles) {
    if (classes === void 0) { classes = ''; }
    if (styles === void 0) { styles = {}; }
    var container = document.createElement(type);
    container.className = classes;
    Object.keys(styles).map(function (key) { return container.style[key] = styles[key]; });
    if (typeof children === 'string') {
        container.innerHTML = children || '';
    }
    else {
        (children || []).forEach(function (child) { return child && container.appendChild(child); });
    }
    return container;
}; };
exports.domContainerFactory = domContainerFactory;
