// @progress/kendo-vue-data-tools/dist/npm/filter/filters/BooleanFilter.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BooleanFilterVue2 = exports.BooleanFilter = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var EnumFilter_1 = require("./EnumFilter");
var main_1 = require("../../messages/main");
/**
 * @hidden
 */
var BooleanFilterVue2 = {
  name: 'KendoBooleanFilter',
  props: {
    filter: {
      type: Object,
      required: true
    },
    dataItems: {
      type: Array,
      default: function _default() {
        return undefined;
      }
    },
    defaultItem: Object
  },
  data: function data() {
    return {
      currentData: [{
        text: main_1.filterIsTrue,
        value: true
      }, {
        text: main_1.filterIsFalse,
        value: false
      }]
    };
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    return (
      // @ts-ignore
      h(EnumFilter_1.EnumFilter, {
        filter: this.$props.filter,
        attrs: this.v3 ? undefined : {
          filter: this.$props.filter,
          dataItems: this.$props.dataItems || this.currentData,
          defaultItem: this.$props.defaultItem
        },
        onFilterchange: this.handleFilterChange,
        on: this.v3 ? undefined : {
          "filterchange": this.handleFilterChange
        },
        dataItems: this.$props.dataItems || this.currentData,
        defaultItem: this.$props.defaultItem
      })
    );
  },
  methods: {
    handleFilterChange: function handleFilterChange(event) {
      this.$emit('filterchange', event);
    }
  }
};
exports.BooleanFilterVue2 = BooleanFilterVue2;
/**
 * @hidden
 */
var BooleanFilter = BooleanFilterVue2;
exports.BooleanFilter = BooleanFilter;
