// @progress/kendo-vue-data-tools/dist/npm/filter/filters/DateFilter.js
"use strict";

var __assign = undefined && undefined.__assign || function () {
  __assign = Object.assign || function (t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
      s = arguments[i];
      for (var p in s) {
        if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
      }
    }
    return t;
  };
  return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DateFilterVue2 = exports.DateFilter = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var kendo_vue_dateinputs_1 = require("@progress/kendo-vue-dateinputs");
/**
 * @hidden
 */
var DateFilterVue2 = {
  name: 'KendoDateFilter',
  props: {
    filter: {
      type: Object,
      required: true
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var filter = this.$props.filter;
    return (
      // @ts-ignore
      h(kendo_vue_dateinputs_1.DatePicker, {
        value: filter.value || null,
        attrs: this.v3 ? undefined : {
          value: filter.value || null
        },
        onChange: this.onChange,
        on: this.v3 ? undefined : {
          "change": this.onChange
        }
      })
    );
  },
  methods: {
    onChange: function onChange(event) {
      this.$emit('filterchange', {
        nextFilter: __assign(__assign({}, this.$props.filter), {
          value: event.value
        })
      });
    }
  }
};
exports.DateFilterVue2 = DateFilterVue2;
/**
 * @hidden
 */
var DateFilter = DateFilterVue2;
exports.DateFilter = DateFilter;
