// @progress/kendo-vue-data-tools/dist/npm/filter/GroupToolbar.js
"use strict";

var __assign = undefined && undefined.__assign || function () {
  __assign = Object.assign || function (t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
      s = arguments[i];
      for (var p in s) {
        if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
      }
    }
    return t;
  };
  return __assign.apply(this, arguments);
};
var __spreadArray = undefined && undefined.__spreadArray || function (to, from, pack) {
  if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
    if (ar || !(i in from)) {
      if (!ar) ar = Array.prototype.slice.call(from, 0, i);
      ar[i] = from[i];
    }
  }
  return to.concat(ar || Array.prototype.slice.call(from));
};
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GroupToolbarVue2 = exports.GroupToolbar = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var kendo_vue_buttons_1 = require("@progress/kendo-vue-buttons");
var kendo_vue_intl_1 = require("@progress/kendo-vue-intl");
var operators_1 = require("./operators");
var main_1 = require("../messages/main");
var kendo_svg_icons_1 = require("@progress/kendo-svg-icons");
/**
 * @hidden
 */
var GroupToolbarVue2 = {
  name: 'KendoFilterGroup',
  // @ts-ignore
  emits: {
    change: null,
    remove: null
  },
  props: {
    filter: {
      type: Object,
      required: true
    },
    fields: {
      type: Array,
      required: true
    },
    defaultGroupFilter: {
      type: Object,
      required: true
    },
    ariaLabel: {
      type: String,
      default: undefined
    }
  },
  inject: {
    kendoLocalizationService: {
      default: null
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var _this = this;
    var h = gh || createElement;
    var filter = this.$props.filter;
    var locService = (0, kendo_vue_intl_1.provideLocalizationService)(this);
    return h("div", {
      "class": "k-filter-toolbar"
    }, [
    // @ts-ignore function children   
    h(kendo_vue_buttons_1.Toolbar, {
      keyboardNavigation: false,
      attrs: this.v3 ? undefined : {
        keyboardNavigation: false,
        role: "toolbar",
        "aria-label": this.$props.ariaLabel
      },
      role: "toolbar",
      "aria-label": this.$props.ariaLabel
    }, this.v3 ? function () {
      return [
      // @ts-ignore function children 
      h(kendo_vue_buttons_1.ButtonGroup, {
        "class": "k-toolbar-button-group"
      }, _this.v3 ? function () {
        return [
        // @ts-ignore function children 
        h(kendo_vue_buttons_1.Button, {
          togglable: true,
          attrs: _this.v3 ? undefined : {
            togglable: true,
            selected: filter.logic === 'and',
            type: "button"
          },
          onClick: _this.onLogicAnd,
          on: _this.v3 ? undefined : {
            "click": _this.onLogicAnd
          },
          "class": "k-toolbar-button",
          selected: filter.logic === 'and',
          type: "button"
        }, _this.v3 ? function () {
          return [locService.toLanguageString(main_1.filterAndLogic, main_1.messages[main_1.filterAndLogic])];
        } : [locService.toLanguageString(main_1.filterAndLogic, main_1.messages[main_1.filterAndLogic])]),
        // @ts-ignore function children 
        h(kendo_vue_buttons_1.Button, {
          togglable: true,
          attrs: _this.v3 ? undefined : {
            togglable: true,
            selected: filter.logic === 'or',
            type: "button"
          },
          onClick: _this.onLogicOr,
          on: _this.v3 ? undefined : {
            "click": _this.onLogicOr
          },
          "class": "k-toolbar-button",
          selected: filter.logic === 'or',
          type: "button"
        }, _this.v3 ? function () {
          return [locService.toLanguageString(main_1.filterOrLogic, main_1.messages[main_1.filterOrLogic])];
        } : [locService.toLanguageString(main_1.filterOrLogic, main_1.messages[main_1.filterOrLogic])])];
      } : [h(kendo_vue_buttons_1.Button, {
        togglable: true,
        attrs: _this.v3 ? undefined : {
          togglable: true,
          selected: filter.logic === 'and',
          type: "button"
        },
        onClick: _this.onLogicAnd,
        on: _this.v3 ? undefined : {
          "click": _this.onLogicAnd
        },
        "class": "k-toolbar-button",
        selected: filter.logic === 'and',
        type: "button"
      }, _this.v3 ? function () {
        return [locService.toLanguageString(main_1.filterAndLogic, main_1.messages[main_1.filterAndLogic])];
      } : [locService.toLanguageString(main_1.filterAndLogic, main_1.messages[main_1.filterAndLogic])]), h(kendo_vue_buttons_1.Button, {
        togglable: true,
        attrs: _this.v3 ? undefined : {
          togglable: true,
          selected: filter.logic === 'or',
          type: "button"
        },
        onClick: _this.onLogicOr,
        on: _this.v3 ? undefined : {
          "click": _this.onLogicOr
        },
        "class": "k-toolbar-button",
        selected: filter.logic === 'or',
        type: "button"
      }, _this.v3 ? function () {
        return [locService.toLanguageString(main_1.filterOrLogic, main_1.messages[main_1.filterOrLogic])];
      } : [locService.toLanguageString(main_1.filterOrLogic, main_1.messages[main_1.filterOrLogic])])]),
      // @ts-ignore function children 
      h(kendo_vue_buttons_1.Button, {
        title: locService.toLanguageString(main_1.filterAddExpression, main_1.messages[main_1.filterAddExpression]),
        attrs: _this.v3 ? undefined : {
          title: locService.toLanguageString(main_1.filterAddExpression, main_1.messages[main_1.filterAddExpression]),
          icon: "filter-add-expression",
          svgIcon: kendo_svg_icons_1.filterAddExpressionIcon,
          type: "button"
        },
        icon: "filter-add-expression",
        svgIcon: kendo_svg_icons_1.filterAddExpressionIcon,
        type: "button",
        "class": "k-toolbar-button",
        onClick: _this.onAddExpression,
        on: _this.v3 ? undefined : {
          "click": _this.onAddExpression
        }
      }),
      // @ts-ignore function children 
      h(kendo_vue_buttons_1.Button, {
        title: locService.toLanguageString(main_1.filterAddGroup, main_1.messages[main_1.filterAddGroup]),
        attrs: _this.v3 ? undefined : {
          title: locService.toLanguageString(main_1.filterAddGroup, main_1.messages[main_1.filterAddGroup]),
          icon: "filter-add-group",
          svgIcon: kendo_svg_icons_1.filterAddGroupIcon,
          type: "button"
        },
        icon: "filter-add-group",
        svgIcon: kendo_svg_icons_1.filterAddGroupIcon,
        type: "button",
        "class": "k-toolbar-button",
        onClick: _this.onAddGroup,
        on: _this.v3 ? undefined : {
          "click": _this.onAddGroup
        }
      }),
      // @ts-ignore function children 
      h(kendo_vue_buttons_1.Button, {
        title: locService.toLanguageString(main_1.filterClose, main_1.messages[main_1.filterClose]),
        attrs: _this.v3 ? undefined : {
          title: locService.toLanguageString(main_1.filterClose, main_1.messages[main_1.filterClose]),
          icon: "x",
          svgIcon: kendo_svg_icons_1.xIcon,
          fillMode: "flat",
          type: "button"
        },
        icon: "x",
        svgIcon: kendo_svg_icons_1.xIcon,
        fillMode: "flat",
        type: "button",
        "class": "k-toolbar-button",
        onClick: _this.onGroupRemove,
        on: _this.v3 ? undefined : {
          "click": _this.onGroupRemove
        }
      })];
    } : [h(kendo_vue_buttons_1.ButtonGroup, {
      "class": "k-toolbar-button-group"
    }, _this.v3 ? function () {
      return [h(kendo_vue_buttons_1.Button, {
        togglable: true,
        attrs: _this.v3 ? undefined : {
          togglable: true,
          selected: filter.logic === 'and',
          type: "button"
        },
        onClick: _this.onLogicAnd,
        on: _this.v3 ? undefined : {
          "click": _this.onLogicAnd
        },
        "class": "k-toolbar-button",
        selected: filter.logic === 'and',
        type: "button"
      }, _this.v3 ? function () {
        return [locService.toLanguageString(main_1.filterAndLogic, main_1.messages[main_1.filterAndLogic])];
      } : [locService.toLanguageString(main_1.filterAndLogic, main_1.messages[main_1.filterAndLogic])]), h(kendo_vue_buttons_1.Button, {
        togglable: true,
        attrs: _this.v3 ? undefined : {
          togglable: true,
          selected: filter.logic === 'or',
          type: "button"
        },
        onClick: _this.onLogicOr,
        on: _this.v3 ? undefined : {
          "click": _this.onLogicOr
        },
        "class": "k-toolbar-button",
        selected: filter.logic === 'or',
        type: "button"
      }, _this.v3 ? function () {
        return [locService.toLanguageString(main_1.filterOrLogic, main_1.messages[main_1.filterOrLogic])];
      } : [locService.toLanguageString(main_1.filterOrLogic, main_1.messages[main_1.filterOrLogic])])];
    } : [h(kendo_vue_buttons_1.Button, {
      togglable: true,
      attrs: _this.v3 ? undefined : {
        togglable: true,
        selected: filter.logic === 'and',
        type: "button"
      },
      onClick: _this.onLogicAnd,
      on: _this.v3 ? undefined : {
        "click": _this.onLogicAnd
      },
      "class": "k-toolbar-button",
      selected: filter.logic === 'and',
      type: "button"
    }, _this.v3 ? function () {
      return [locService.toLanguageString(main_1.filterAndLogic, main_1.messages[main_1.filterAndLogic])];
    } : [locService.toLanguageString(main_1.filterAndLogic, main_1.messages[main_1.filterAndLogic])]), h(kendo_vue_buttons_1.Button, {
      togglable: true,
      attrs: _this.v3 ? undefined : {
        togglable: true,
        selected: filter.logic === 'or',
        type: "button"
      },
      onClick: _this.onLogicOr,
      on: _this.v3 ? undefined : {
        "click": _this.onLogicOr
      },
      "class": "k-toolbar-button",
      selected: filter.logic === 'or',
      type: "button"
    }, _this.v3 ? function () {
      return [locService.toLanguageString(main_1.filterOrLogic, main_1.messages[main_1.filterOrLogic])];
    } : [locService.toLanguageString(main_1.filterOrLogic, main_1.messages[main_1.filterOrLogic])])]), h(kendo_vue_buttons_1.Button, {
      title: locService.toLanguageString(main_1.filterAddExpression, main_1.messages[main_1.filterAddExpression]),
      attrs: _this.v3 ? undefined : {
        title: locService.toLanguageString(main_1.filterAddExpression, main_1.messages[main_1.filterAddExpression]),
        icon: "filter-add-expression",
        svgIcon: kendo_svg_icons_1.filterAddExpressionIcon,
        type: "button"
      },
      icon: "filter-add-expression",
      svgIcon: kendo_svg_icons_1.filterAddExpressionIcon,
      type: "button",
      "class": "k-toolbar-button",
      onClick: _this.onAddExpression,
      on: _this.v3 ? undefined : {
        "click": _this.onAddExpression
      }
    }), h(kendo_vue_buttons_1.Button, {
      title: locService.toLanguageString(main_1.filterAddGroup, main_1.messages[main_1.filterAddGroup]),
      attrs: _this.v3 ? undefined : {
        title: locService.toLanguageString(main_1.filterAddGroup, main_1.messages[main_1.filterAddGroup]),
        icon: "filter-add-group",
        svgIcon: kendo_svg_icons_1.filterAddGroupIcon,
        type: "button"
      },
      icon: "filter-add-group",
      svgIcon: kendo_svg_icons_1.filterAddGroupIcon,
      type: "button",
      "class": "k-toolbar-button",
      onClick: _this.onAddGroup,
      on: _this.v3 ? undefined : {
        "click": _this.onAddGroup
      }
    }), h(kendo_vue_buttons_1.Button, {
      title: locService.toLanguageString(main_1.filterClose, main_1.messages[main_1.filterClose]),
      attrs: _this.v3 ? undefined : {
        title: locService.toLanguageString(main_1.filterClose, main_1.messages[main_1.filterClose]),
        icon: "x",
        svgIcon: kendo_svg_icons_1.xIcon,
        fillMode: "flat",
        type: "button"
      },
      icon: "x",
      svgIcon: kendo_svg_icons_1.xIcon,
      fillMode: "flat",
      type: "button",
      "class": "k-toolbar-button",
      onClick: _this.onGroupRemove,
      on: _this.v3 ? undefined : {
        "click": _this.onGroupRemove
      }
    })])]);
  },
  methods: {
    onGroupRemove: function onGroupRemove() {
      this.$emit('remove', {
        filter: this.$props.filter
      });
    },
    onAddExpression: function onAddExpression() {
      var prevFilter = this.$props.filter;
      var firstField = this.$props.fields[0];
      var newExpression = {
        field: firstField.name,
        operator: firstField.operators[0].operator
      };
      newExpression.value = (0, operators_1.stringOperator)(newExpression.operator) ? '' : null;
      this.$emit('change', {
        nextFilter: __assign(__assign({}, prevFilter), {
          filters: __spreadArray(__spreadArray([], prevFilter.filters, true), [newExpression], false)
        }),
        prevFilter: prevFilter
      });
    },
    onAddGroup: function onAddGroup() {
      var prevFilter = this.$props.filter;
      this.$emit('change', {
        nextFilter: __assign(__assign({}, prevFilter), {
          filters: __spreadArray(__spreadArray([], prevFilter.filters, true), [__assign({}, this.$props.defaultGroupFilter)], false)
        }),
        prevFilter: prevFilter
      });
    },
    onLogicAnd: function onLogicAnd() {
      this.changeLogic('and');
    },
    onLogicOr: function onLogicOr() {
      this.changeLogic('or');
    },
    changeLogic: function changeLogic(logic) {
      var prevFilter = this.$props.filter;
      if (prevFilter.logic !== logic) {
        this.$emit('change', {
          nextFilter: __assign(__assign({}, prevFilter), {
            logic: logic
          }),
          prevFilter: prevFilter
        });
      }
    }
  }
};
exports.GroupToolbarVue2 = GroupToolbarVue2;
/**
 * @hidden
 */
var GroupToolbar = GroupToolbarVue2;
exports.GroupToolbar = GroupToolbar;
