// @progress/kendo-vue-labels/dist/npm/Hint.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HintVue2 = exports.Hint = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var package_metadata_1 = require("./package-metadata");
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
/**
 * @hidden
 */
var HintVue2 = {
  name: 'KendoHint',
  props: {
    id: String,
    direction: {
      type: String,
      default: 'start',
      validator: function validator(value) {
        return ['start', 'end'].includes(value);
      }
    }
  },
  created: function created() {
    (0, kendo_vue_common_1.validatePackage)(package_metadata_1.packageMetadata);
  },
  computed: {
    hintClassName: function hintClassName() {
      return {
        'k-form-hint': true,
        'k-text-start': this.$props.direction === 'start',
        'k-text-end': this.$props.direction === 'end'
      };
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = (0, kendo_vue_common_1.getDefaultSlots)(this);
    return h("div", {
      id: this.$props.id,
      attrs: this.v3 ? undefined : {
        id: this.$props.id
      },
      "class": this.hintClassName
    }, [defaultSlots]);
  }
};
exports.HintVue2 = HintVue2;
/**
 * @hidden
 */
var Hint = HintVue2;
exports.Hint = Hint;
