// @progress/kendo-vue-labels/dist/npm/main.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HintVue2 = exports.Hint = exports.ErrorVue2 = exports.Error = exports.LabelVue2 = exports.Label = exports.FloatingLabelVue2 = exports.FloatingLabel = void 0;
var Label_1 = require("./Label");
Object.defineProperty(exports, "Label", { enumerable: true, get: function () { return Label_1.Label; } });
Object.defineProperty(exports, "LabelVue2", { enumerable: true, get: function () { return Label_1.LabelVue2; } });
var Error_1 = require("./Error");
Object.defineProperty(exports, "Error", { enumerable: true, get: function () { return Error_1.Error; } });
Object.defineProperty(exports, "ErrorVue2", { enumerable: true, get: function () { return Error_1.ErrorVue2; } });
var Hint_1 = require("./Hint");
Object.defineProperty(exports, "Hint", { enumerable: true, get: function () { return Hint_1.Hint; } });
Object.defineProperty(exports, "HintVue2", { enumerable: true, get: function () { return Hint_1.HintVue2; } });
var FloatingLabel_1 = require("./FloatingLabel");
Object.defineProperty(exports, "FloatingLabel", { enumerable: true, get: function () { return FloatingLabel_1.FloatingLabel; } });
Object.defineProperty(exports, "FloatingLabelVue2", { enumerable: true, get: function () { return FloatingLabel_1.FloatingLabelVue2; } });
