// @progress/kendo-vue-dateinputs/dist/npm/calendar/components/CalendarHeaderTitle.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CalendarHeaderTitleVue2 = exports.CalendarHeaderTitle = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
var kendo_vue_buttons_1 = require("@progress/kendo-vue-buttons");
/**
 * @hidden
 */
var CalendarHeaderTitleVue2 = {
  name: 'KendoHeaderTitle',
  // @ts-ignore
  emits: {
    'click': null
  },
  props: {
    id: String,
    value: String,
    view: Number
  },
  methods: {
    handleClick: function handleClick(event) {
      this.$emit('click', event);
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  /**
   * @return
   * Returns a `<button />` element with the [`value`]({% slug api_dateinputs_calendarheadertitleprops %}#toc-value) of the title as a child.
   */
  // @ts-ignore
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlot = (0, kendo_vue_common_1.getDefaultSlots)(this);
    return (
      // @ts-ignore  function children
      h(kendo_vue_buttons_1.Button, {
        type: "button",
        attrs: this.v3 ? undefined : {
          type: "button",
          value: this.$props.value,
          id: this.$props.id,
          fillMode: 'flat',
          tabIndex: 0
        },
        value: this.$props.value,
        id: this.$props.id,
        fillMode: 'flat',
        onClick: this.handleClick,
        on: this.v3 ? undefined : {
          "click": this.handleClick
        },
        tabIndex: 0
      }, this.v3 ? function () {
        return [defaultSlot];
      } : [defaultSlot])
    );
  }
};
exports.CalendarHeaderTitleVue2 = CalendarHeaderTitleVue2;
/**
 * @hidden
 */
var CalendarHeaderTitle = CalendarHeaderTitleVue2;
exports.CalendarHeaderTitle = CalendarHeaderTitle;
