// @progress/kendo-vue-dateinputs/dist/npm/calendar/components/CalendarCell.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CalendarCellVue2 = exports.CalendarCell = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
/**
 * @hidden
 */
var compareProps = function compareProps(x, y) {
  var xProps = Object.getOwnPropertyNames(x);
  var yProps = Object.getOwnPropertyNames(y);
  if (xProps.length !== yProps.length) {
    return false;
  }
  for (var i = 0; i < xProps.length; i++) {
    var propName = xProps[i];
    if (x[propName] !== y[propName]) {
      return false;
    }
  }
  return true;
};
/**
 * @hidden
 */
var CalendarCellVue2 = {
  name: 'KendoCalendarCell',
  // @ts-ignore
  emits: {
    'click': null,
    'mousedown': null,
    'mouseenter': null,
    'mouseleave': null
  },
  props: {
    isDisabled: Boolean,
    view: Number,
    formattedValue: String,
    id: String,
    isWeekend: Boolean,
    isFocused: Boolean,
    isSelected: Boolean,
    isInRange: Boolean,
    isRangeStart: Boolean,
    isRangeEnd: Boolean,
    isRangeMid: Boolean,
    isRangeSplitEnd: Boolean,
    isRangeSplitStart: Boolean,
    isToday: Boolean,
    title: String,
    value: Date
  },
  // Manually checking if the component needs an update
  // due to date object being compared by instance
  // and new Date object is created
  // every time and fails the shallow compare of the Vue.
  /**
   * @hidden
   */
  updated: function updated() {
    // const { value, ...props } = this.props;
    // const { value: newValue, ...newProps } = nextProps;
    // const valueEqual = !(value && newValue) || value.getTime() === newValue.getTime();
    // return !(valueEqual && compareProps(props, newProps));
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  // tslint:disable:max-line-length
  /**
   * @return
   * Returns a `<td />` element with a `<span class="k-link" />` inside to apply the styles. The text inside is the [`formattedValue`]({% slug api_dateinputs_calendarcellprops %}#toc-formattedvalue) of the `cell`.
   */
  // @ts-ignore
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlot = (0, kendo_vue_common_1.getDefaultSlots)(this);
    var _a = this.$props,
      formattedValue = _a.formattedValue,
      isWeekend = _a.isWeekend,
      isFocused = _a.isFocused,
      isInRange = _a.isInRange,
      isSelected = _a.isSelected,
      isRangeStart = _a.isRangeStart,
      isRangeMid = _a.isRangeMid,
      isRangeEnd = _a.isRangeEnd,
      isRangeSplitStart = _a.isRangeSplitStart,
      isRangeSplitEnd = _a.isRangeSplitEnd,
      isToday = _a.isToday,
      isDisabled = _a.isDisabled,
      view = _a.view,
      value = _a.value;
    var isEndActive = this.$props.activeRangeEnd === 'end' && isRangeEnd;
    var isStartActive = this.$props.activeRangeEnd === 'start' && isRangeStart;
    return h("td", {
      id: this.$props.id,
      attrs: this.v3 ? undefined : {
        id: this.$props.id,
        title: this.$props.title
      },
      title: this.$props.title,
      "class": {
        'k-calendar-td': true,
        'k-range-end': isRangeEnd,
        'k-range-mid': isRangeMid,
        'k-range-split-end': isRangeSplitEnd,
        'k-range-split-start': isRangeSplitStart,
        'k-range-start': isRangeStart,
        'k-active': isStartActive || isEndActive,
        'k-focus': isFocused,
        'k-selected': isSelected || isRangeStart || isRangeEnd,
        'k-today': isToday,
        'k-weekend': isWeekend,
        'k-disabled': isDisabled
      },
      onClick: this.handleClick,
      on: this.v3 ? undefined : {
        "click": this.handleClick,
        "mousedown": this.handleMouseDown,
        "mouseenter": this.handleMouseEnter,
        "mouseleave": this.handleMouseLeave
      },
      onMousedown: this.handleMouseDown,
      onMouseenter: this.handleMouseEnter,
      onMouseleave: this.handleMouseLeave
    }, [h("span", {
      "class": "k-link"
    }, [defaultSlot])]);
  },
  // tslint:enable:max-line-length
  methods: {
    handleClick: function handleClick(event) {
      var value = this.$props.value;
      this.$emit('click', value, event);
    },
    handleMouseDown: function handleMouseDown(event) {
      var value = this.$props.value;
      this.$emit('mousedown', value, event);
    },
    handleMouseEnter: function handleMouseEnter() {
      var value = this.$props.value;
      this.$emit('mouseenter', value);
    },
    handleMouseLeave: function handleMouseLeave() {
      var value = this.$props.value;
      this.$emit('mouseleave', value);
    }
  }
};
exports.CalendarCellVue2 = CalendarCellVue2;
/**
 * @hidden
 */
var CalendarCell = CalendarCellVue2;
exports.CalendarCell = CalendarCell;
