// @progress/kendo-vue-dateinputs/dist/npm/calendar/components/CalendarWeekCell.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CalendarWeekCellVue2 = exports.CalendarWeekCell = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
/**
 * @hidden
 */
var CalendarWeekCellVue2 = {
  props: {
    id: String,
    value: Number
  },
  // @ts-ignore
  emits: {
    'click': null
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  /**
   * @return
   * Returns a `<td />` element with the [`value`]({% slug api_dateinputs_calendarweekcellprops %}#toc-value) as a child.
   */
  // @ts-ignore
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlot = (0, kendo_vue_common_1.getDefaultSlots)(this);
    return h("td", {
      onClick: this.handleClick,
      on: this.v3 ? undefined : {
        "click": this.handleClick
      },
      id: this.$props.id,
      attrs: this.v3 ? undefined : {
        id: this.$props.id
      }
    }, [this.$props.value !== undefined ? this.$props.value : defaultSlot]);
  },
  methods: {
    handleClick: function handleClick(event) {
      var value = this.$props.value;
      this.$emit('click', value, event);
    }
  }
};
exports.CalendarWeekCellVue2 = CalendarWeekCellVue2;
/**
 * @hidden
 */
var CalendarWeekCell = CalendarWeekCellVue2;
exports.CalendarWeekCell = CalendarWeekCell;
