// @progress/kendo-vue-data-tools/dist/esm/header/Header.js
// @ts-ignore
import * as Vue from 'vue';
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var ref = allVue.ref;
import { isRtl, getRef, setRef, BrowserSupportService, kendoThemeMaps } from '@progress/kendo-vue-common';
/**
 * @hidden
 */
var HeaderVue2 = {
  name: 'KendoHeader',
  props: {
    staticHeaders: Boolean,
    headerRow: Object,
    columnResize: Object,
    cols: Array,
    draggable: Boolean,
    size: String
  },
  data: function data() {
    return {
      divStyle: {},
      element: null,
      headerWrap: null,
      table: null
    };
  },
  computed: {
    wrapperClass: function wrapperClass() {
      return {
        'k-grid-header': true,
        'k-grid-draggable-header': this.$props.draggable
      };
    },
    tableClass: function tableClass() {
      var _a;
      var size = this.$props.size;
      return _a = {
        'k-table': true,
        'k-grid-header-table': true
      }, _a["k-table-".concat(kendoThemeMaps.sizeMap[size] || size)] = size, _a;
    }
  },
  mounted: function mounted() {
    this.headerWrap = getRef(this, 'headerWrap');
    this.table = getRef(this, 'table');
    this.$props.columnResize.colGroupHeader = getRef(this, 'colGroupHeader');
    var scrollbarWidth = new BrowserSupportService().scrollbarWidth;
    var rtl = isRtl(this.$el);
    if (this.$props.columnResize) {
      this.$props.columnResize.setIsRtl(rtl);
    }
    var padding = Math.max(0, scrollbarWidth) + 'px';
    var right = rtl ? 0 : padding;
    var left = rtl ? padding : 0;
    this.divStyle = {
      padding: "0 ".concat(right, " 0 ").concat(left)
    };
  },
  methods: {
    setScrollLeft: function setScrollLeft(scrollLeft) {
      if (this.headerWrap) {
        this.headerWrap.scrollLeft = scrollLeft;
      }
    },
    setWidth: function setWidth(width) {
      if (this.table) {
        this.table.style.width = width + 'px';
      }
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    var headerWrapRef = ref(null);
    var tableRef = ref(null);
    var colGroupHeaderRef = ref(null);
    return {
      v3: v3,
      headerWrapRef: headerWrapRef,
      tableRef: tableRef,
      colGroupHeaderRef: colGroupHeaderRef
    };
  },
  // @ts-ignore
  render: function render(createElement) {
    var h = gh || createElement;
    if (!this.$props.staticHeaders) {
      return this.$props.headerRow;
    }
    return h("div", {
      "class": this.wrapperClass,
      style: this.divStyle,
      role: "presentation",
      attrs: this.v3 ? undefined : {
        role: "presentation"
      }
    }, [h("div", {
      ref: setRef(this, 'headerWrap'),
      "class": "k-grid-header-wrap",
      role: "presentation",
      attrs: this.v3 ? undefined : {
        role: "presentation"
      }
    }, [h("table", {
      "class": this.tableClass,
      ref: setRef(this, 'table'),
      role: "presentation",
      attrs: this.v3 ? undefined : {
        role: "presentation"
      }
    }, [h("colgroup", {
      ref: setRef(this, 'colGroupHeader')
    }, [this.$props.cols]), this.$props.headerRow])])]);
  }
};
/**
 * @hidden
 */
var Header = HeaderVue2;
export { Header, HeaderVue2 };
