// @progress/kendo-vue-dialogs/dist/esm/WindowTitlebar.js
// @ts-ignore
import { getTemplate } from '@progress/kendo-vue-common';
import * as Vue from 'vue';
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
import { Button } from '@progress/kendo-vue-buttons';
import { provideLocalizationService } from '@progress/kendo-vue-intl';
import { windowStage } from './StageEnum.js';
import { dialogsWindowCloseButton, dialogsWindowMaximizeButton, dialogsWindowMinimizeButton, dialogsWindowRestoreButton, messages } from './messages/main.js';
import { windowIcon, windowMinimizeIcon, windowRestoreIcon, xIcon } from '@progress/kendo-svg-icons';
/**
 * @hidden
 */
var WindowTitleBarVue2 = {
  name: 'KendoWindowTitleBar',
  props: {
    id: String,
    stage: String,
    closeButton: [String, Function, Object, Boolean],
    minimizeButton: [String, Function, Object, Boolean],
    maximizeButton: [String, Function, Object, Boolean],
    restoreButton: [String, Function, Object, Boolean],
    title: String,
    titleId: String,
    titleRender: [String, Function, Object, Boolean]
  },
  inject: {
    kendoLocalizationService: {
      default: null
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  methods: {
    onDoubleClick: function onDoubleClick(e) {
      this.$emit('doubleclick', e);
    },
    onMinimizeClick: function onMinimizeClick(e) {
      this.$emit('minimizeclick', e);
    },
    onFullScreenClick: function onFullScreenClick(e) {
      this.$emit('fullscreenclick', e);
    },
    onRestoreClick: function onRestoreClick(e) {
      this.$emit('restoreclick', e);
    },
    onCloseClick: function onCloseClick(e) {
      this.$emit('closeclick', e);
    }
  },
  // @ts-ignore
  render: function render(createElement) {
    var h = gh || createElement;
    var props = this.$props;
    var stage = props.stage,
      title = props.title,
      titleRender = props.titleRender,
      minimizeButton = props.minimizeButton,
      maximizeButton = props.maximizeButton,
      restoreButton = props.restoreButton,
      closeButton = props.closeButton;
    var ls = provideLocalizationService(this);
    var titleElement = getTemplate.call(this, {
      h: h,
      template: titleRender,
      defaultRendering: title
    });
    var minimizeButtonDefault =
    // @ts-ignore
    h(Button, {
      type: "button",
      attrs: this.v3 ? undefined : {
        type: "button",
        icon: 'window-minimize',
        svgIcon: windowMinimizeIcon,
        fillMode: 'flat',
        "aria-label": ls.toLanguageString(dialogsWindowMinimizeButton, messages[dialogsWindowMinimizeButton])
      },
      "class": "k-window-titlebar-action",
      icon: 'window-minimize',
      svgIcon: windowMinimizeIcon,
      fillMode: 'flat',
      onClick: this.onMinimizeClick,
      on: this.v3 ? undefined : {
        "click": this.onMinimizeClick
      },
      "aria-label": ls.toLanguageString(dialogsWindowMinimizeButton, messages[dialogsWindowMinimizeButton])
    });
    var minimizeButtonRender = getTemplate.call(this, {
      h: h,
      template: minimizeButton,
      defaultRendering: minimizeButtonDefault
    });
    var maximizeButtonDefault =
    // @ts-ignore
    h(Button, {
      type: "button",
      attrs: this.v3 ? undefined : {
        type: "button",
        icon: 'window',
        svgIcon: windowIcon,
        fillMode: 'flat',
        "aria-label": ls.toLanguageString(dialogsWindowMaximizeButton, messages[dialogsWindowMaximizeButton])
      },
      "class": "k-window-titlebar-action",
      icon: 'window',
      svgIcon: windowIcon,
      fillMode: 'flat',
      onClick: this.onFullScreenClick,
      on: this.v3 ? undefined : {
        "click": this.onFullScreenClick
      },
      "aria-label": ls.toLanguageString(dialogsWindowMaximizeButton, messages[dialogsWindowMaximizeButton])
    });
    var maximizeButtonRender = getTemplate.call(this, {
      h: h,
      template: maximizeButton,
      defaultRendering: maximizeButtonDefault
    });
    var restoreButtonDefault =
    // @ts-ignore
    h(Button, {
      type: "button",
      attrs: this.v3 ? undefined : {
        type: "button",
        icon: 'window-restore',
        svgIcon: windowRestoreIcon,
        fillMode: 'flat',
        "aria-label": ls.toLanguageString(dialogsWindowRestoreButton, messages[dialogsWindowRestoreButton])
      },
      "class": "k-window-titlebar-action",
      icon: 'window-restore',
      svgIcon: windowRestoreIcon,
      fillMode: 'flat',
      onClick: this.onRestoreClick,
      on: this.v3 ? undefined : {
        "click": this.onRestoreClick
      },
      "aria-label": ls.toLanguageString(dialogsWindowRestoreButton, messages[dialogsWindowRestoreButton])
    });
    var restoreButtonRender = getTemplate.call(this, {
      h: h,
      template: restoreButton,
      defaultRendering: restoreButtonDefault
    });
    var closeButtonDefault =
    // @ts-ignore
    h(Button, {
      type: "button",
      attrs: this.v3 ? undefined : {
        type: "button",
        icon: 'x',
        svgIcon: xIcon,
        fillMode: 'flat',
        "aria-label": ls.toLanguageString(dialogsWindowCloseButton, messages[dialogsWindowCloseButton])
      },
      "class": "k-window-titlebar-action",
      icon: 'x',
      svgIcon: xIcon,
      fillMode: 'flat',
      onClick: this.onCloseClick,
      on: this.v3 ? undefined : {
        "click": this.onCloseClick
      },
      "aria-label": ls.toLanguageString(dialogsWindowCloseButton, messages[dialogsWindowCloseButton])
    });
    var closeButtonRender = getTemplate.call(this, {
      h: h,
      template: closeButton,
      defaultRendering: closeButtonDefault
    });
    return h("div", {
      "class": "k-window-titlebar",
      style: {
        touchAction: 'none'
      },
      onDblclick: this.onDoubleClick,
      on: this.v3 ? undefined : {
        "dblclick": this.onDoubleClick
      }
    }, [h("span", {
      "class": "k-window-title",
      id: this.$props.titleId,
      attrs: this.v3 ? undefined : {
        id: this.$props.titleId
      }
    }, [titleElement]), h("div", {
      "class": "k-window-titlebar-actions"
    }, [stage === windowStage.DEFAULT && minimizeButtonRender, stage === windowStage.DEFAULT && maximizeButtonRender, stage !== windowStage.DEFAULT && restoreButtonRender, closeButtonRender])]);
  }
};
/**
 * @hidden
 */
var WindowTitleBar = WindowTitleBarVue2;
export { WindowTitleBar, WindowTitleBarVue2 };
