// @progress/kendo-vue-inputs/dist/npm/colors/utils/color-cache.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCachedRgba = exports.getCachedHsva = exports.getCachedHex = exports.removeCachedColor = exports.cacheHex = exports.cacheRgba = exports.cacheHsva = exports.DEFAULT_PALETTE_SETTINGS = exports.DEFAULT_GRADIENT_SETTINGS = void 0;
var ColorPalette_1 = require("../ColorPalette");
var cache = {};
/**
 * @hidden
 */
exports.DEFAULT_GRADIENT_SETTINGS = {
    opacity: true
};
/**
 * @hidden
 */
exports.DEFAULT_PALETTE_SETTINGS = {
    palette: ColorPalette_1.DEFAULT_PRESET,
    tileSize: ColorPalette_1.DEFAULT_TILE_SIZE
};
/**
 * @hidden
 */
function cacheHsva(componentGuid, value, hsva) {
    cache[componentGuid] = { value: value, hsva: hsva };
}
exports.cacheHsva = cacheHsva;
/**
 * @hidden
 */
function cacheRgba(componentGuid, value, rgba) {
    cache[componentGuid] = { value: value, rgba: rgba };
}
exports.cacheRgba = cacheRgba;
/**
 * @hidden
 */
function cacheHex(componentGuid, value, hex) {
    cache[componentGuid] = { value: value, hex: hex };
}
exports.cacheHex = cacheHex;
/**
 * @hidden
 */
function removeCachedColor(componentGuid) {
    delete cache[componentGuid];
}
exports.removeCachedColor = removeCachedColor;
/**
 * @hidden
 */
function getCachedHex(componentGuid, value) {
    if (componentGuid) {
        var cachedColor = cache[componentGuid];
        return cachedColor && cachedColor.value === value ? cachedColor.hex : undefined;
    }
    else {
        return undefined;
    }
}
exports.getCachedHex = getCachedHex;
/**
 * @hidden
 */
function getCachedHsva(componentGuid, value) {
    if (componentGuid) {
        var cachedColor = cache[componentGuid];
        return cachedColor && cachedColor.value === value ? cachedColor.hsva : undefined;
    }
    else {
        return undefined;
    }
}
exports.getCachedHsva = getCachedHsva;
/**
 * @hidden
 */
function getCachedRgba(componentGuid, value) {
    if (componentGuid) {
        var cachedColor = cache[componentGuid];
        return cachedColor && cachedColor.value === value ? cachedColor.rgba : undefined;
    }
    else {
        return undefined;
    }
}
exports.getCachedRgba = getCachedRgba;
