// @progress/kendo-vue-data-tools/dist/es/pager/PagerInfo.js
// @ts-ignore
import * as Vue from 'vue';
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var inject = allVue.inject;
import { provideIntlService, provideLocalizationService } from '@progress/kendo-vue-intl';
import { messages, pagerInfo } from '../messages/main';
/**
 * @hidden
 */
var PagerInfoVue2 = {
  name: 'KendoPagerInfo',
  props: {
    totalPages: Number,
    currentPage: Number,
    skip: Number,
    messagesMap: Function
  },
  inject: {
    kendoLocalizationService: {
      default: null
    },
    kendoIntlService: {
      default: null
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    var kendoIntlService = inject('kendoIntlService', {});
    var kendoLocalizationService = inject('kendoLocalizationService', {});
    return {
      v3: v3,
      kendoIntlService: kendoIntlService,
      kendoLocalizationService: kendoLocalizationService
    };
  },
  // @ts-ignore
  render: function render(createElement) {
    var h = gh || createElement;
    var intlService = provideIntlService(this);
    var localizationService = provideLocalizationService(this);
    var infoMessage = this.$props.messagesMap ? this.$props.messagesMap(pagerInfo) : {
      messageKey: pagerInfo,
      defaultMessage: messages[pagerInfo]
    };
    return h("span", {
      "class": "k-pager-info"
    }, [intlService.format(localizationService.toLanguageString(infoMessage.messageKey, infoMessage.defaultMessage), [Math.min(this.$props.skip + 1, this.$props.totalPages), Math.min(this.$props.skip + this.$props.currentPage, this.$props.totalPages), this.$props.totalPages])]);
  }
};
/**
 * @hidden
 */
var PagerInfo = PagerInfoVue2;
export { PagerInfo, PagerInfoVue2 };
