// @progress/kendo-vue-grid/dist/es/cells/GridGroupCell.js
// @ts-ignore
import * as Vue from 'vue';
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
import { getTemplate, Icon, Keys, noop } from '@progress/kendo-vue-common';
import { KEYBOARD_NAV_DATA_ID, KEYBOARD_NAV_DATA_LEVEL } from '@progress/kendo-vue-data-tools';
import { messages, groupCaretAriaLabelExpand, groupCaretAriaLabelCollapse } from '../messages/main';
import { provideLocalizationService } from '@progress/kendo-vue-intl';
import { caretAltDownIcon, caretAltRightIcon, caretAltLeftIcon } from '@progress/kendo-svg-icons';
/**
 * @hidden
 */
var GridGroupCellVue2 = {
  name: 'KendoGridGroupCell',
  inheritAttrs: false,
  props: {
    id: String,
    field: String,
    dataItem: Object,
    format: String,
    type: String,
    colSpan: Number,
    className: String,
    columnIndex: Number,
    columnsCount: Number,
    rowType: String,
    level: Number,
    expanded: Boolean,
    editor: String,
    dataIndex: Number,
    isSelected: Boolean,
    isRtl: Boolean,
    ariaColumnIndex: Number,
    render: [String, Function, Object]
  },
  // @ts-ignore
  emits: {
    change: null,
    cellkeydown: null
  },
  inject: {
    kendoIntlService: {
      default: null
    },
    kendoLocalizationService: {
      default: null
    },
    getKeyboardNavigationAttributes: {
      default: noop
    }
  },
  computed: {
    tdClass: function tdClass() {
      var _a;
      var className = this.$props.className;
      return _a = {
        'k-table-td': true
      }, _a[className] = className, _a;
    }
  },
  methods: {
    triggerKeydown: function triggerKeydown(event, expanded) {
      this.$emit('cellkeydown', {
        event: event,
        dataItem: this.$props.dataItem,
        dataIndex: this.$props.dataIndex,
        field: this.$props.field,
        expanded: this.$props.expanded
      });
      if (event.defaultPrevented) {
        return;
      }
      if (event.keyCode === Keys.enter) {
        event.preventDefault();
        this.$emit('change', {
          dataItem: this.$props.dataItem,
          dataIndex: this.$props.dataIndex,
          event: event,
          level: this.$props.level,
          field: undefined,
          value: !expanded
        });
      }
    },
    clickHandler: function clickHandler(e, dataItem, expanded) {
      e.preventDefault();
      this.$emit('change', {
        dataItem: dataItem,
        dataIndex: this.$props.dataIndex,
        event: e,
        level: this.$props.level,
        field: undefined,
        value: !expanded
      });
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  // @ts-ignore
  render: function render(createElement) {
    var _this = this;
    var h = gh || createElement;
    var defaultRendering = null;
    var _a = this.$props,
      columnIndex = _a.columnIndex,
      level = _a.level,
      columnsCount = _a.columnsCount,
      rowType = _a.rowType,
      dataItem = _a.dataItem,
      field = _a.field,
      expanded = _a.expanded,
      render = _a.render;
    var navAttrs = this.getKeyboardNavigationAttributes(this.$props.id);
    var groupCaretAriaLabelCollapseValue = provideLocalizationService(this).toLanguageString(groupCaretAriaLabelCollapse, messages[groupCaretAriaLabelCollapse]);
    var groupCaretAriaLabelExpandValue = provideLocalizationService(this).toLanguageString(groupCaretAriaLabelExpand, messages[groupCaretAriaLabelExpand]);
    if (columnIndex === undefined || level === undefined || columnIndex < level || columnsCount === undefined || rowType !== 'groupHeader' || dataItem[field] === undefined) {
      defaultRendering = h("td", {
        style: this.$attrs.style,
        key: 'g' + columnIndex,
        "class": "k-table-td k-group-cell"
      });
    } else if (columnIndex <= level) {
      defaultRendering = h("td", {
        style: this.$attrs.style,
        onKeydown: function onKeydown(ev) {
          _this.triggerKeydown(ev, expanded);
        },
        on: this.v3 ? undefined : {
          "keydown": function onKeydown(ev) {
            _this.triggerKeydown(ev, expanded);
          }
        },
        key: 'g-colspan',
        "class": this.tdClass,
        colspan: columnsCount - columnIndex,
        attrs: this.v3 ? undefined : {
          colspan: columnsCount - columnIndex,
          role: 'gridcell',
          "aria-colindex": this.$props.ariaColumnIndex,
          "aria-selected": this.$props.isSelected,
          "aria-expanded": expanded,
          "data-grid-col-index": this.$props.columnIndex,
          tabindex: navAttrs.tabIndex,
          "data-keyboardnavlevel": navAttrs[KEYBOARD_NAV_DATA_LEVEL],
          "data-keyboardnavid": navAttrs[KEYBOARD_NAV_DATA_ID]
        },
        role: 'gridcell',
        "aria-colindex": this.$props.ariaColumnIndex,
        "aria-selected": this.$props.isSelected,
        "aria-expanded": expanded,
        "data-grid-col-index": this.$props.columnIndex,
        tabindex: navAttrs.tabIndex,
        "data-keyboardnavlevel": navAttrs[KEYBOARD_NAV_DATA_LEVEL],
        "data-keyboardnavid": navAttrs[KEYBOARD_NAV_DATA_ID]
      }, [h("p", {
        "class": "k-reset"
      }, [h("a", {
        onClick: function onClick(e) {
          _this.clickHandler(e, dataItem, expanded);
        },
        on: this.v3 ? undefined : {
          "click": function onClick(e) {
            _this.clickHandler(e, dataItem, expanded);
          }
        },
        href: "#",
        attrs: this.v3 ? undefined : {
          href: "#",
          tabindex: -1,
          title: expanded ? groupCaretAriaLabelCollapseValue : groupCaretAriaLabelExpandValue,
          ariaLable: expanded ? groupCaretAriaLabelCollapseValue : groupCaretAriaLabelExpandValue
        },
        tabindex: -1,
        title: expanded ? groupCaretAriaLabelCollapseValue : groupCaretAriaLabelExpandValue,
        ariaLable: expanded ? groupCaretAriaLabelCollapseValue : groupCaretAriaLabelExpandValue
      }, [h(Icon, {
        name: expanded ? 'caret-alt-down' : !this.isRtl ? 'caret-alt-right' : 'caret-alt-left',
        attrs: this.v3 ? undefined : {
          name: expanded ? 'caret-alt-down' : !this.isRtl ? 'caret-alt-right' : 'caret-alt-left',
          icon: expanded ? caretAltDownIcon : !this.isRtl ? caretAltRightIcon : caretAltLeftIcon
        },
        icon: expanded ? caretAltDownIcon : !this.isRtl ? caretAltRightIcon : caretAltLeftIcon
      })]), dataItem[field].toString()])]);
    }
    return getTemplate.call(this, {
      h: h,
      template: render,
      defaultRendering: defaultRendering,
      additionalProps: this.$props,
      additionalListeners: {
        keydown: this.triggerKeydown,
        click: this.clickHandler
      }
    });
  }
};
/**
 * @hidden
 */
var GridGroupCell = GridGroupCellVue2;
export { GridGroupCell, GridGroupCellVue2 };
