// @progress/kendo-vue-data-tools/dist/npm/cells/FilterCell.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FilterCellVue2 = exports.FilterCell = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var inject = allVue.inject;
var kendo_vue_buttons_1 = require("@progress/kendo-vue-buttons");
var main_1 = require("../messages/main");
var kendo_vue_dropdowns_1 = require("@progress/kendo-vue-dropdowns");
var kendo_vue_inputs_1 = require("@progress/kendo-vue-inputs");
var kendo_vue_dateinputs_1 = require("@progress/kendo-vue-dateinputs");
var kendo_vue_intl_1 = require("@progress/kendo-vue-intl");
var filterCommon_1 = require("../filterCommon");
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
var kendo_svg_icons_1 = require("@progress/kendo-svg-icons");
/**
 * @hidden
 */
var FilterCellVue2 = {
  name: 'KendoFilterCell',
  inheritAttrs: false,
  // @ts-ignore
  emits: {
    'change': null
  },
  props: {
    id: String,
    grid: Object,
    field: String,
    filterType: String,
    colSpan: Number,
    title: String,
    value: [String, Number, Boolean, Date],
    operator: [String, Function],
    operators: Array,
    booleanValues: Array,
    onChange: Function,
    render: [String, Function, Object],
    ariaLabel: String,
    size: String
  },
  inject: {
    kendoLocalizationService: {
      default: null
    },
    kendoIntlService: {
      default: null
    }
  },
  methods: {
    inputChange: function inputChange(value, e) {
      var filter = (0, filterCommon_1.cellInputChange)(value, e, this.$props);
      this.triggerChange(filter);
    },
    operatorChange: function operatorChange(operatorValue, e) {
      var filter = (0, filterCommon_1.cellOperatorChange)(operatorValue.value.operator, e, this.$props.value);
      this.triggerChange(filter);
    },
    boolDropdownChange: function boolDropdownChange(value, e) {
      var filter = (0, filterCommon_1.cellBoolDropdownChange)(value.value.operator, e);
      this.triggerChange(filter);
    },
    clear: function clear(e) {
      e.preventDefault();
      this.triggerChange({
        value: '',
        operator: '',
        event: e
      });
    },
    triggerChange: function triggerChange(filter) {
      filter.field = this.$props.field;
      this.$emit('change', filter);
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    var kendoIntlService = inject('kendoIntlService', {});
    var kendoLocalizationService = inject('kendoLocalizationService', {});
    return {
      v3: v3,
      kendoIntlService: kendoIntlService,
      kendoLocalizationService: kendoLocalizationService
    };
  },
  // @ts-ignore
  render: function render(createElement) {
    var _this = this;
    var h = gh || createElement;
    var size = this.$props.size;
    var ls = (0, kendo_vue_intl_1.provideLocalizationService)(this);
    var selectedOperator = this.$props.operators.find(function (item) {
      return item.operator === _this.$props.operator;
    }) || null;
    var renderOperatorEditor = function renderOperatorEditor() {
      if (this.$props.filterType === 'boolean') {
        return;
      }
      return (
        // @ts-ignore
        h(kendo_vue_dropdowns_1.DropDownList, {
          onChange: this.operatorChange,
          on: this.v3 ? undefined : {
            "change": this.operatorChange
          },
          value: selectedOperator,
          attrs: this.v3 ? undefined : {
            value: selectedOperator,
            size: size,
            icon: 'filter',
            svgIcon: kendo_svg_icons_1.filterIcon,
            iconClassName: "k-button-icon",
            "data-items": this.$props.operators,
            textField: "text",
            title: ls.toLanguageString(main_1.columnMenuFilterChooseOperator, main_1.messages[main_1.columnMenuFilterChooseOperator]),
            popupSettings: {
              width: ''
            },
            "aria-label": this.ariaLabel || ls.toLanguageString(main_1.filterDropDownOperatorAriaLabel, main_1.messages[main_1.filterDropDownOperatorAriaLabel])
          },
          size: size,
          "class": "k-dropdown-operator",
          icon: 'filter',
          svgIcon: kendo_svg_icons_1.filterIcon,
          iconClassName: "k-button-icon",
          "data-items": this.$props.operators,
          textField: "text",
          title: ls.toLanguageString(main_1.columnMenuFilterChooseOperator, main_1.messages[main_1.columnMenuFilterChooseOperator]),
          popupSettings: {
            width: ''
          },
          "aria-label": this.ariaLabel || ls.toLanguageString(main_1.filterDropDownOperatorAriaLabel, main_1.messages[main_1.filterDropDownOperatorAriaLabel])
        })
      );
    };
    var filterComponent = function filterComponent(filterType, value) {
      var _this = this;
      switch (filterType) {
        case 'numeric':
          return (
            // @ts-ignore
            h(kendo_vue_inputs_1.NumericTextBox, {
              size: size,
              attrs: this.v3 ? undefined : {
                size: size,
                value: value,
                title: this.$props.title,
                "aria-label": this.ariaLabel || ls.toLanguageString(main_1.filterNumericFilterAriaLabel, main_1.messages[main_1.filterNumericFilterAriaLabel])
              },
              value: value,
              onChange: function onChange(e) {
                _this.inputChange(e.value, e.event);
              },
              on: this.v3 ? undefined : {
                "change": function onChange(e) {
                  _this.inputChange(e.value, e.event);
                }
              },
              title: this.$props.title,
              "aria-label": this.ariaLabel || ls.toLanguageString(main_1.filterNumericFilterAriaLabel, main_1.messages[main_1.filterNumericFilterAriaLabel])
            })
          );
        case 'date':
          return (
            // @ts-ignore
            h(kendo_vue_dateinputs_1.DatePicker, {
              size: size,
              attrs: this.v3 ? undefined : {
                size: size,
                value: value,
                title: this.$props.title,
                "aria-label": this.ariaLabel || ls.toLanguageString(main_1.filterDateFilterAriaLabel, main_1.messages[main_1.filterDateFilterAriaLabel])
              },
              value: value,
              onChange: function onChange(e) {
                _this.inputChange(e.value, e);
              },
              on: this.v3 ? undefined : {
                "change": function onChange(e) {
                  _this.inputChange(e.value, e);
                }
              },
              title: this.$props.title,
              "aria-label": this.ariaLabel || ls.toLanguageString(main_1.filterDateFilterAriaLabel, main_1.messages[main_1.filterDateFilterAriaLabel])
            })
          );
        case 'boolean':
          var booleanValues = this.$props.booleanValues;
          return (
            // @ts-ignore
            h(kendo_vue_dropdowns_1.DropDownList, {
              onChange: this.boolDropdownChange,
              on: this.v3 ? undefined : {
                "change": this.boolDropdownChange
              },
              size: size,
              attrs: this.v3 ? undefined : {
                size: size,
                value: booleanValues.find(function (item) {
                  return item.operator === (value !== null ? value : '');
                }),
                "data-items": booleanValues,
                textField: "text",
                title: this.$props.title,
                "aria-label": this.ariaLabel || ls.toLanguageString(main_1.filterEnumFilterDropdownAriaLabel, main_1.messages[main_1.filterEnumFilterDropdownAriaLabel])
              },
              value: booleanValues.find(function (item) {
                return item.operator === (value !== null ? value : '');
              }),
              "data-items": booleanValues,
              textField: "text",
              title: this.$props.title,
              "aria-label": this.ariaLabel || ls.toLanguageString(main_1.filterEnumFilterDropdownAriaLabel, main_1.messages[main_1.filterEnumFilterDropdownAriaLabel])
            })
          );
        default:
          return h(kendo_vue_inputs_1.TextBox, {
            size: size,
            attrs: this.v3 ? undefined : {
              size: size,
              value: value || '',
              title: this.$props.title,
              "aria-label": this.ariaLabel || ls.toLanguageString(main_1.filterTextFilterAriaLabel, main_1.messages[main_1.filterTextFilterAriaLabel])
            },
            value: value || '',
            onInput: function onInput(e) {
              _this.inputChange(e.target.value, e);
            },
            on: this.v3 ? undefined : {
              "input": function onInput(e) {
                _this.inputChange(e.target.value, e);
              }
            },
            title: this.$props.title,
            "aria-label": this.ariaLabel || ls.toLanguageString(main_1.filterTextFilterAriaLabel, main_1.messages[main_1.filterTextFilterAriaLabel])
          });
      }
    };
    var defaultRendering = h("div", {
      "class": "k-filtercell",
      style: this.$attrs.style
    }, [h("div", {
      "class": "k-filtercell-wrapper"
    }, [filterComponent.call(this, this.$props.filterType, this.$props.value), h("div", {
      "class": "k-filtercell-operator"
    }, [renderOperatorEditor.call(this),
    // @ts-ignore function children 
    h(kendo_vue_buttons_1.Button, {
      size: size,
      attrs: this.v3 ? undefined : {
        size: size,
        type: "button",
        icon: 'filter-clear',
        svgIcon: kendo_svg_icons_1.filterClearIcon,
        title: ls.toLanguageString(main_1.columnMenuFilterClearButton, main_1.messages[main_1.columnMenuFilterClearButton])
      },
      type: "button",
      icon: 'filter-clear',
      svgIcon: kendo_svg_icons_1.filterClearIcon,
      "class": {
        /* button is always visible if there is either value or operator */
        'k-disabled': !(!(this.$props.value === null || this.$props.value === '') || this.$props.operator)
      },
      title: ls.toLanguageString(main_1.columnMenuFilterClearButton, main_1.messages[main_1.columnMenuFilterClearButton]),
      onClick: this.clear,
      on: this.v3 ? undefined : {
        "click": this.clear
      }
    })])])]);
    var gridListeners = this.$props.grid ? kendo_vue_common_1.getListeners.call(this.$props.grid) : null;
    var cellRenderFunction = kendo_vue_common_1.templateRendering.call(this.$props.grid, this.$props.render, gridListeners);
    return kendo_vue_common_1.getTemplate.call(this, {
      h: h,
      template: cellRenderFunction,
      defaultRendering: defaultRendering,
      additionalProps: this.$props,
      additionalListeners: {
        change: this.triggerChange
      }
    });
  }
};
exports.FilterCellVue2 = FilterCellVue2;
/**
 * @hidden
 */
var FilterCell = FilterCellVue2;
exports.FilterCell = FilterCell;
