// @progress/kendo-vue-data-tools/dist/es/header/HeaderTdElement.js
// @ts-ignore
import * as Vue from 'vue';
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var inject = allVue.inject;
import { getDefaultSlots, noop } from '@progress/kendo-vue-common';
import { KEYBOARD_NAV_DATA_ID, KEYBOARD_NAV_DATA_LEVEL } from '../navigation/constants';
/**
 * @hidden
 */
var HeaderTdElementVue2 = {
  name: 'KendoHeaderTdElement',
  // @ts-ignore
  emits: {
    keydown: null
  },
  props: {
    ariaColumnIndex: Number,
    role: String,
    columnId: String,
    navigatable: Boolean
  },
  computed: {
    tdClass: function tdClass() {
      var _a;
      var className = this.$props.className;
      return _a = {
        'k-table-td': true
      }, _a[className] = className, _a;
    }
  },
  inject: {
    getKeyboardNavigationAttributes: {
      default: noop
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    var kendoIntlService = inject('kendoIntlService', {});
    var kendoLocalizationService = inject('kendoLocalizationService', {});
    return {
      v3: v3,
      kendoIntlService: kendoIntlService,
      kendoLocalizationService: kendoLocalizationService
    };
  },
  // @ts-ignore
  render: function render(createElement) {
    var h = gh || createElement;
    var _a = this.$props,
      columnId = _a.columnId,
      navigatable = _a.navigatable,
      ariaColumnIndex = _a.ariaColumnIndex,
      role = _a.role;
    var navAttrs = this.getKeyboardNavigationAttributes(columnId, navigatable);
    var defaultSlot = getDefaultSlots(this);
    return h("td", {
      style: {
        top: '0px'
      },
      "aria-colindex": ariaColumnIndex,
      attrs: this.v3 ? undefined : {
        "aria-colindex": ariaColumnIndex,
        role: role,
        tabindex: navAttrs.tabIndex,
        "data-keyboardnavlevel": navAttrs[KEYBOARD_NAV_DATA_LEVEL],
        "data-keyboardnavid": navAttrs[KEYBOARD_NAV_DATA_ID]
      },
      "class": this.tdClass,
      onKeydown: this.onKeyDown,
      on: this.v3 ? undefined : {
        "keydown": this.onKeyDown
      },
      role: role,
      tabindex: navAttrs.tabIndex,
      "data-keyboardnavlevel": navAttrs[KEYBOARD_NAV_DATA_LEVEL],
      "data-keyboardnavid": navAttrs[KEYBOARD_NAV_DATA_ID]
    }, [defaultSlot]);
  },
  methods: {
    onKeyDown: function onKeyDown(e) {
      this.$emit('keydown', e);
    }
  }
};
/**
 * @hidden
 */
var HeaderTdElement = HeaderTdElementVue2;
export { HeaderTdElement, HeaderTdElementVue2 };
