// @progress/kendo-vue-dropdowns/dist/npm/messages/main.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.messages = exports.adaptiveModeFooterCancel = exports.adaptiveModeFooterApply = exports.selectButton = exports.clear = exports.expandButton = exports.nodata = void 0;
/**
 * @hidden
 */
exports.nodata = 'dropdowns.nodata';
/**
 * @hidden
 */
exports.expandButton = 'dropdowns.expandbutton';
/**
 * @hidden
 */
exports.clear = 'dropdowns.clear';
/**
 * @hidden
 */
exports.selectButton = 'dropdowns.select';
/**
 * @hidden
 */
exports.adaptiveModeFooterApply = 'dropdowns.adaptiveModeFooterApply';
/**
 * @hidden
 */
exports.adaptiveModeFooterCancel = 'dropdowns.adaptiveModeFooterCancel';
/**
 * @hidden
 */
exports.messages = (_a = {},
    _a[exports.clear] = 'clear',
    _a[exports.expandButton] = 'expand button',
    _a[exports.nodata] = 'NO DATA FOUND.',
    _a[exports.selectButton] = 'Select',
    _a[exports.adaptiveModeFooterApply] = 'Apply',
    _a[exports.adaptiveModeFooterCancel] = 'Cancel',
    _a);
