// @progress/kendo-vue-grid/dist/npm/GridNoRecords.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GridNoRecordsVue2 = exports.GridNoRecords = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var inject = allVue.inject;
var kendo_vue_intl_1 = require("@progress/kendo-vue-intl");
var main_1 = require("./messages/main");
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
/**
 * @hidden
 */
var GridNoRecordsVue2 = {
  name: 'GridNoRecords',
  inject: {
    kendoLocalizationService: {
      default: null
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    var kendoLocalizationService = inject('kendoLocalizationService', {});
    return {
      v3: v3,
      kendoLocalizationService: kendoLocalizationService
    };
  },
  // @ts-ignore
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlot = (0, kendo_vue_common_1.getDefaultSlots)(this);
    var localizationService = (0, kendo_vue_intl_1.provideLocalizationService)(this);
    var noRecordsMessage = localizationService.toLanguageString(main_1.noRecords, main_1.messages[main_1.noRecords]);
    return defaultSlot ? h("div", [defaultSlot])
    // @ts-ignore
    : this.v3 ? noRecordsMessage : this._v(noRecordsMessage);
  }
};
exports.GridNoRecordsVue2 = GridNoRecordsVue2;
/**
 * Represents the GridNoRecords component which is rendered
 * when the `data` property of the Grid is empty or `null`.
 *
 * @example
 * ```tsx-no-run
 *   <div id="app">
 *       <Grid
 *           :data-items="products"
 *           :columns="columns">
 *       <grid-norecords>
 *        There is no data available custom
 *        </grid-norecords>
 *       </Grid>
 *   </div>
 * <script>
 *       Vue.component('Grid', Grid);
 *       Vue.component('grid-norecords', GridNoRecords);
 *       new Vue({
 *           el: '#app',
 *           data: function () {
 *               return {
 *                  products: [
 *                    { 'foo': 'A1', 'bar': 'B1' },
 *                       { 'foo': 'A2', 'bar': 'B2' },
 *                       { 'foo': 'A3', 'bar': 'B2' }
 *                   ],
 *                   columns: [
 *                       { field: 'foo'},
 *                       { field: 'bar'}
 *                   ]
 *               };
 *           }
 *       });
 *   </script>
 * ```
 */
/**
 * Represents the default `GridNoRecords` component.
 */
var GridNoRecords = GridNoRecordsVue2;
exports.GridNoRecords = GridNoRecords;
