// @progress/kendo-vue-grid/dist/npm/GridCellValue.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GridCellValueVue2 = exports.GridCellValue = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
/**
 * @hidden
 */
var GridCellValueVue2 = {
  name: 'GridCellValue',
  props: {
    dataItem: Object,
    column: Object,
    editField: String,
    editCell: Object,
    inbuildCell: Object
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  // @ts-ignore
  render: function render(createElement) {
    var h = gh || createElement;
    var inEdit = this.editField ? (0, kendo_vue_common_1.getNestedValue)(this.editField, this.dataItem) : false;
    return this.column.editable && this.editField && (inEdit === true || inEdit === this.column.field) ? this.editCell : this.inbuildCell;
  }
};
exports.GridCellValueVue2 = GridCellValueVue2;
/**
 * @hidden
 */
var GridCellValue = GridCellValueVue2;
exports.GridCellValue = GridCellValue;
