// @progress/kendo-vue-buttons/dist/npm/ListButton/SplitButton.js
"use strict";

var __assign = undefined && undefined.__assign || function () {
  __assign = Object.assign || function (t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
      s = arguments[i];
      for (var p in s) {
        if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
      }
    }
    return t;
  };
  return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SplitButtonVue2 = exports.SplitButton = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var ref = allVue.ref;
var Button_1 = require("../Button");
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
var ButtonItem_1 = require("./ButtonItem");
var navigation_1 = require("./utils/navigation");
var kendo_vue_popup_1 = require("@progress/kendo-vue-popup");
var popup_1 = require("./utils/popup");
var kendo_vue_common_2 = require("@progress/kendo-vue-common");
var package_metadata_1 = require("../package-metadata");
/**
 * @hidden
 */
var SplitButtonVue2 = {
  name: 'KendoSplitButton',
  // @ts-ignore
  emits: {
    focus: null,
    blur: null,
    buttonclick: null,
    itemclick: null,
    open: null,
    close: null
  },
  props: {
    accessKey: String,
    ariaLabel: String,
    text: String,
    items: {
      type: Array,
      default: function _default() {
        return [];
      }
    },
    textField: String,
    tabIndex: Number,
    disabled: Boolean,
    icon: String,
    svgIcon: Object,
    size: {
      type: String,
      default: 'medium'
    },
    rounded: {
      type: String,
      default: 'medium'
    },
    fillMode: {
      type: String,
      default: 'solid',
      validator: function validator(value) {
        return [null, 'flat', 'link', 'outline', 'solid'].includes(value);
      }
    },
    // eslint-disable-next-line max-len
    themeColor: {
      type: String,
      default: 'base',
      validator: function validator(value) {
        return [null, 'base', 'dark', 'error', 'info', 'inverse', 'inverse', 'light', 'primary', 'secondary', 'success', 'tertiary', 'warning'].includes(value);
      }
    },
    opened: {
      type: Boolean,
      default: undefined
    },
    iconClass: String,
    imageUrl: String,
    popupSettings: Object,
    itemRender: [String, Function, Object],
    item: [String, Function, Object],
    look: String,
    className: String,
    buttonClass: String,
    dir: String
  },
  data: function data() {
    return {
      focused: false,
      focusedIndex: -1,
      currentOpened: false
    };
  },
  created: function created() {
    this._blurTimeout = null;
    this._anchor = (0, kendo_vue_common_1.guid)();
    this.mainButton = null;
    this.guid = (0, kendo_vue_common_1.guid)();
    this.buttonsData = [];
    (0, kendo_vue_common_2.validatePackage)(package_metadata_1.packageMetadata);
  },
  mounted: function mounted() {
    this.mainButton = this.$refs[this._anchor];
    // If this.$props.opened is true during the initial render, the popup is not aligned.
    if (this.$props.dir === undefined && this.isRtl() || this.computedOpened) {
      this.$forceUpdate();
    }
  },
  updated: function updated() {
    if (this.focused && this.element()) {
      this.mainButton = this.$refs[this._anchor];
      this.mainButton.focus();
    }
  },
  computed: {
    computedOpened: function computedOpened() {
      return this.$props.opened === undefined ? this.currentOpened : this.$props.opened;
    },
    wrapperClass: function wrapperClass() {
      return {
        'k-split-button': true,
        'k-button-group': true,
        'k-focus': this.focused
      };
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    var kendoAnchorRef = ref(null);
    return {
      v3: v3,
      kendoAnchorRef: kendoAnchorRef
    };
  },
  render: function render(createElement) {
    var _this3 = this;
    var _this = this;
    var h = gh || createElement;
    this.buttonsData = this.$props.items;
    var rtl = this.isRtl();
    var dir = rtl ? 'rtl' : undefined;
    var _a = this.$props,
      tabIndex = _a.tabIndex,
      disabled = _a.disabled;
    var defaultSlot = (0, kendo_vue_common_1.getDefaultSlots)(this);
    var renderChildItems = function renderChildItems() {
      var _a = this.$props,
        item = _a.item,
        itemRender = _a.itemRender,
        textField = _a.textField;
      return this.buttonsData.length > 0 ? this.buttonsData.map(function (dataItem, index) {
        var currentDataItem = typeof dataItem !== 'string' ? __assign(__assign({}, dataItem), {
          render: kendo_vue_common_2.templateRendering.call(this, dataItem.render, kendo_vue_common_2.getListeners.call(this))
        }) : dataItem;
        return (
          // @ts-ignore
          h(ButtonItem_1.ButtonItem, {
            "class": "k-menu-item",
            role: "menuitem",
            attrs: this.v3 ? undefined : {
              role: "menuitem",
              dataItem: currentDataItem,
              textField: textField,
              focused: this.focusedIndex === index,
              render: kendo_vue_common_2.templateRendering.call(this, itemRender, kendo_vue_common_2.getListeners.call(this)),
              item: item,
              index: index,
              id: "".concat(this.guid, "-").concat(index)
            },
            dataItem: currentDataItem,
            textField: textField,
            focused: this.focusedIndex === index,
            onClick: this.onItemClick,
            on: this.v3 ? undefined : {
              "click": this.onItemClick,
              "down": this.onItemDown
            },
            onDown: this.onItemDown,
            render: kendo_vue_common_2.templateRendering.call(this, itemRender, kendo_vue_common_2.getListeners.call(this)),
            item: item,
            key: index,
            index: index,
            id: "".concat(this.guid, "-").concat(index)
          })
        );
      }, this) : null;
    };
    var renderPopup = function renderPopup() {
      var _this2 = this;
      var _a = this.$props,
        _b = _a.popupSettings,
        popupSettings = _b === void 0 ? {} : _b,
        size = _a.size;
      return (
        // @ts-ignore function children
        h(kendo_vue_popup_1.Popup, {
          anchor: this._anchor,
          attrs: this.v3 ? undefined : {
            anchor: this._anchor,
            show: this.computedOpened,
            animate: popupSettings.animate,
            popupClass: (0, kendo_vue_common_1.classNames)('k-menu-popup', popupSettings.popupClass),
            anchorAlign: popupSettings.anchorAlign || (0, popup_1.getAnchorAlign)(rtl),
            popupAlign: popupSettings.popupAlign || (0, popup_1.getPopupAlign)(rtl)
          },
          show: this.computedOpened,
          animate: popupSettings.animate,
          popupClass: (0, kendo_vue_common_1.classNames)('k-menu-popup', popupSettings.popupClass),
          anchorAlign: popupSettings.anchorAlign || (0, popup_1.getAnchorAlign)(rtl),
          popupAlign: popupSettings.popupAlign || (0, popup_1.getPopupAlign)(rtl),
          style: rtl ? {
            direction: 'rtl'
          } : undefined
        }, this.v3 ? function () {
          return [h("ul", {
            "class": "k-group k-menu-group k-reset k-menu-group-".concat(kendo_vue_common_1.kendoThemeMaps.sizeMap[size] || size),
            role: "menu",
            attrs: _this2.v3 ? undefined : {
              role: "menu",
              id: _this2.guid,
              "aria-labelledby": _this2._anchor
            },
            id: _this2.guid,
            "aria-labelledby": _this2._anchor
          }, [renderChildItems.call(_this2)])];
        } : [h("ul", {
          "class": "k-group k-menu-group k-reset k-menu-group-".concat(kendo_vue_common_1.kendoThemeMaps.sizeMap[size] || size),
          role: "menu",
          attrs: _this2.v3 ? undefined : {
            role: "menu",
            id: _this2.guid,
            "aria-labelledby": _this2._anchor
          },
          id: _this2.guid,
          "aria-labelledby": _this2._anchor
        }, [renderChildItems.call(_this2)])])
      );
    };
    return h("div", {
      "class": this.wrapperClass,
      onKeydown: this.onKeyDown,
      on: this.v3 ? undefined : {
        "keydown": this.onKeyDown,
        "focusin": this.onFocus,
        "focusout": this.onBlur
      },
      onFocusin: this.onFocus,
      onFocusout: this.onBlur,
      dir: dir,
      attrs: this.v3 ? undefined : {
        dir: dir
      }
    }, [
    // @ts-ignore function children
    h(Button_1.Button, {
      size: this.$props.size,
      attrs: this.v3 ? undefined : {
        size: this.$props.size,
        rounded: this.$props.rounded,
        fillMode: this.$props.fillMode,
        themeColor: this.$props.themeColor,
        disabled: disabled || undefined,
        tabIndex: tabIndex,
        accessKey: this.$props.accessKey,
        icon: this.$props.icon,
        svgIcon: this.$props.svgIcon,
        iconClass: this.$props.iconClass,
        imageUrl: this.$props.imageUrl,
        look: this.$props.look,
        dir: dir,
        id: this._anchor,
        type: "button",
        "aria-disabled": disabled,
        "aria-haspopup": 'menu',
        "aria-expanded": this.computedOpened,
        "aria-label": this.$props.ariaLabel || "".concat(this.$props.text || '', " splitbutton"),
        "aria-controls": this.guid,
        "aria-activedescendant": this.focusedIndex !== undefined && this.focusedIndex >= 0 ? "".concat(this.guid, "-").concat(this.focusedIndex) : undefined
      },
      rounded: this.$props.rounded,
      fillMode: this.$props.fillMode,
      themeColor: this.$props.themeColor,
      onClick: function onClick(event) {
        return _this.onItemClick(event, -1);
      },
      on: this.v3 ? undefined : {
        "click": function onClick(event) {
          return _this.onItemClick(event, -1);
        }
      },
      disabled: disabled || undefined,
      tabIndex: tabIndex,
      accessKey: this.$props.accessKey,
      "class": this.$props.buttonClass,
      icon: this.$props.icon,
      svgIcon: this.$props.svgIcon,
      iconClass: this.$props.iconClass,
      imageUrl: this.$props.imageUrl,
      look: this.$props.look,
      dir: dir,
      id: this._anchor,
      ref: this._anchor,
      type: "button",
      "aria-disabled": disabled,
      "aria-haspopup": 'menu',
      "aria-expanded": this.computedOpened,
      "aria-label": this.$props.ariaLabel || "".concat(this.$props.text || '', " splitbutton"),
      "aria-controls": this.guid,
      "aria-activedescendant": this.focusedIndex !== undefined && this.focusedIndex >= 0 ? "".concat(this.guid, "-").concat(this.focusedIndex) : undefined
    }, this.v3 ? function () {
      return [_this3.$props.text, defaultSlot];
    } : [_this3.$props.text, defaultSlot]),
    // @ts-ignore
    h(Button_1.Button, {
      "class": 'k-split-button-arrow',
      size: this.$props.size,
      attrs: this.v3 ? undefined : {
        size: this.$props.size,
        rounded: this.$props.rounded,
        fillMode: this.$props.fillMode,
        themeColor: this.$props.themeColor,
        icon: "caret-alt-down",
        disabled: disabled || undefined,
        tabIndex: -1,
        look: this.$props.look,
        dir: dir,
        "aria-label": "menu toggling button"
      },
      rounded: this.$props.rounded,
      fillMode: this.$props.fillMode,
      themeColor: this.$props.themeColor,
      icon: "caret-alt-down",
      disabled: disabled || undefined,
      tabIndex: -1,
      look: this.$props.look,
      onClick: this.onSplitPartClick,
      on: this.v3 ? undefined : {
        "click": this.onSplitPartClick,
        "mousedown": this.onDownSplitPart,
        "pointerdown": this.onDownSplitPart
      },
      onMousedown: this.onDownSplitPart,
      onPointerdown: this.onDownSplitPart,
      dir: dir,
      "aria-label": "menu toggling button"
    }), renderPopup.call(this)]);
  },
  methods: {
    element: function element() {
      return this.mainButton;
    },
    onKeyDown: function onKeyDown(event) {
      if (event.altKey) {
        if (!this.computedOpened && event.keyCode === kendo_vue_common_1.Keys.down) {
          this.dispatchPopupEvent(event, true);
          this.focusedIndex = 0;
          this.currentOpened = true;
        } else if (this.computedOpened && event.keyCode === kendo_vue_common_1.Keys.up) {
          this.dispatchPopupEvent(event, false);
          this.focusedIndex = -1;
          this.currentOpened = false;
        }
        return;
      }
      var newState = undefined;
      if (event.keyCode === kendo_vue_common_1.Keys.enter || event.keyCode === kendo_vue_common_1.Keys.space) {
        // Prevent default because otherwise when an item is selected
        // click on the default button gets emitted which opens the popup again.
        event.preventDefault();
        this.dispatchClickEvent(event, this.focusedIndex);
        if (this.focusedIndex !== undefined && this.focusedIndex >= 0) {
          newState = {
            focusedIndex: this.computedOpened ? -1 : 0,
            currentOpened: !this.computedOpened
          };
          this.dispatchPopupEvent(event, newState.currentOpened);
        }
      } else if (this.computedOpened && event.keyCode === kendo_vue_common_1.Keys.esc) {
        newState = {
          focusedIndex: -1,
          currentOpened: false
        };
        this.dispatchPopupEvent(event, newState.currentOpened);
      }
      if (this.computedOpened) {
        var newFocused = (0, navigation_1.default)(this.focusedIndex, event.keyCode, event.altKey, this.buttonsData.length);
        if (newFocused !== this.focusedIndex) {
          newState = newState || {};
          // @ts-ignore
          newState.focusedIndex = newFocused;
        }
        var arrowKey = event.keyCode === kendo_vue_common_1.Keys.up || event.keyCode === kendo_vue_common_1.Keys.down || event.keyCode === kendo_vue_common_1.Keys.left || event.keyCode === kendo_vue_common_1.Keys.right;
        if (!event.altKey && arrowKey) {
          // Needed to notify the parent listeners that event is handled.
          event.preventDefault();
        }
      }
      if (newState) {
        this.focusedIndex = newState.focusedIndex;
        this.focused = newState.focused;
        if (newState.currentOpened !== undefined) {
          this.currentOpened = newState.currentOpened;
        }
      }
    },
    onFocus: function onFocus(event) {
      if (!this.focused) {
        this.$emit('focus', event, this, undefined);
        this.focused = true;
      }
      this.focusedIndex = -1;
      clearTimeout(this._blurTimeout);
    },
    onItemClick: function onItemClick(event, clickedItemIndex) {
      var opened = this.computedOpened;
      if (opened) {
        this.focusedIndex = 0;
        this.currentOpened = false;
      }
      this.dispatchClickEvent(event, clickedItemIndex);
      if (opened) {
        this.dispatchPopupEvent(event, false);
      }
    },
    onBlur: function onBlur(event) {
      clearTimeout(this._blurTimeout);
      this.createBlurTimeout(event);
    },
    createBlurTimeout: function createBlurTimeout(event) {
      var _this = this;
      var that = this;
      this._blurTimeout = setTimeout(function () {
        if (kendo_vue_common_2.canUseDOM && document.activeElement !== that.mainButton) {
          that.focused = false;
          that.focusedIndex = -1;
          that.$emit('blur', event, _this, undefined);
          var fireCloseEvent = that.computedOpened;
          if (fireCloseEvent) {
            that.currentOpened = false;
            that.dispatchPopupEvent(event, false);
          }
        }
      }, 200);
    },
    dispatchClickEvent: function dispatchClickEvent(dispatchedEvent, clickedItemIndex) {
      if (!this.isItemDisabled(clickedItemIndex)) {
        if (clickedItemIndex === -1) {
          this.$emit('buttonclick', dispatchedEvent, this, undefined);
        } else {
          this.$emit('itemclick', {
            event: dispatchedEvent,
            component: this,
            item: this.buttonsData[clickedItemIndex],
            itemIndex: clickedItemIndex
          });
        }
      }
    },
    onSplitPartClick: function onSplitPartClick(event) {
      if (this.buttonsData.length) {
        var toOpen = !this.computedOpened;
        this.dispatchPopupEvent(event, toOpen);
        this.focusedIndex = toOpen ? 0 : -1;
        this.currentOpened = toOpen;
        this.focused = true;
      }
    },
    onDownSplitPart: function onDownSplitPart(event) {
      event.preventDefault();
      if (this.element() && document.activeElement !== this.element()) {
        // @ts-ignore
        this.element().focus();
      }
    },
    onItemDown: function onItemDown(event) {
      if (document.activeElement === this.element()) {
        event.preventDefault();
      }
    },
    dispatchPopupEvent: function dispatchPopupEvent(dispatchedEvent, open) {
      this.$emit(open ? 'open' : 'close', dispatchedEvent, this, undefined);
    },
    isItemDisabled: function isItemDisabled(index) {
      return this.buttonsData[index] ? this.buttonsData[index].disabled : this.$props.disabled;
    },
    isRtl: function isRtl() {
      return this.$props.dir !== undefined ? this.$props.dir === 'rtl' : !!this.$el && getComputedStyle(this.$el).direction === 'rtl';
    }
  }
};
exports.SplitButtonVue2 = SplitButtonVue2;
/**
 * @hidden
 */
var SplitButton = SplitButtonVue2;
exports.SplitButton = SplitButton;
