// @progress/kendo-vue-inputs/dist/npm/colors/ColorContrastLabels.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ColorContrastLabelsVue2 = exports.ColorContrastLabels = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var kendo_vue_intl_1 = require("@progress/kendo-vue-intl");
var kendo_svg_icons_1 = require("@progress/kendo-svg-icons");
var main_1 = require("../messages/main");
var color_parser_1 = require("./utils/color-parser");
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
/**
 * @hidden
 */
var ColorContrastLabelsVue2 = {
  name: 'KendoColorContrastLabels',
  props: {
    bgColor: Object,
    rgba: Object
  },
  inject: {
    kendoLocalizationService: {
      default: null
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  // @ts-ignore
  render: function render(createElement) {
    var h = gh || createElement;
    var ls = (0, kendo_vue_intl_1.provideLocalizationService)(this);
    var contrastRatioMessage = ls.toLanguageString(main_1.colorGradientContrastRatio, main_1.messages[main_1.colorGradientContrastRatio]);
    var AALevelMessage = ls.toLanguageString(main_1.colorGradientAALevel, main_1.messages[main_1.colorGradientAALevel]);
    var AAALevelMessage = ls.toLanguageString(main_1.colorGradientAAALevel, main_1.messages[main_1.colorGradientAAALevel]);
    var passMessage = ls.toLanguageString(main_1.colorGradientPass, main_1.messages[main_1.colorGradientPass]);
    var failMessage = ls.toLanguageString(main_1.colorGradientFail, main_1.messages[main_1.colorGradientFail]);
    var contrast = (0, color_parser_1.getContrastFromTwoRGBAs)(this.$props.rgba, this.$props.bgColor);
    var requiredAAContrast = 4.5.toFixed(1);
    var requiredAAAContrast = 7.0.toFixed(1);
    var contrastText = "".concat(contrastRatioMessage, ": ").concat(contrast.toFixed(2));
    var aaText = "".concat(AALevelMessage, ": ").concat(requiredAAContrast);
    var aaaText = "".concat(AAALevelMessage, ": ").concat(requiredAAAContrast);
    var aaPass = contrast >= 4.5;
    var aaaPass = contrast >= 7;
    var success = h("span", {
      key: 1,
      "class": "k-contrast-validation k-text-success"
    }, [passMessage, " ", h(kendo_vue_common_1.Icon, {
      name: "check",
      attrs: this.v3 ? undefined : {
        name: "check",
        icon: kendo_svg_icons_1.checkIcon
      },
      icon: kendo_svg_icons_1.checkIcon
    })]);
    var error = h("span", {
      key: 2,
      "class": "k-contrast-validation k-text-error"
    }, [failMessage, " ", h(kendo_vue_common_1.Icon, {
      name: "x",
      attrs: this.v3 ? undefined : {
        name: "x",
        icon: kendo_svg_icons_1.xIcon
      },
      icon: kendo_svg_icons_1.xIcon
    })]);
    return h("div", {
      "class": "k-vbox k-colorgradient-color-contrast"
    }, [h("div", {
      "class": "k-contrast-ratio"
    }, [h("span", {
      "class": "k-contrast-ratio-text"
    }, [contrastText]), aaPass ? h("span", {
      "class": "k-contrast-validation k-text-success"
    }, [h(kendo_vue_common_1.Icon, {
      name: "check",
      attrs: this.v3 ? undefined : {
        name: "check",
        icon: kendo_svg_icons_1.checkIcon
      },
      icon: kendo_svg_icons_1.checkIcon
    }), aaaPass && h(kendo_vue_common_1.Icon, {
      name: "check",
      attrs: this.v3 ? undefined : {
        name: "check",
        icon: kendo_svg_icons_1.checkIcon
      },
      icon: kendo_svg_icons_1.checkIcon
    })]) : h("span", {
      "class": "k-contrast-validation k-text-error"
    }, [h(kendo_vue_common_1.Icon, {
      name: "x",
      attrs: this.v3 ? undefined : {
        name: "x",
        icon: kendo_svg_icons_1.xIcon
      },
      icon: kendo_svg_icons_1.xIcon
    })])]), h("div", [h("span", [aaText]), aaPass ? success : error]), h("div", [h("span", [aaaText]), aaaPass ? success : error])]);
  }
};
exports.ColorContrastLabelsVue2 = ColorContrastLabelsVue2;
/**
 * @hidden
 */
var ColorContrastLabels = ColorContrastLabelsVue2;
exports.ColorContrastLabels = ColorContrastLabels;
