// @progress/kendo-vue-layout/dist/npm/actionsheet/ActionSheetContent.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ActionSheetContentVue2 = exports.ActionSheetContent = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
/**
 * @hidden
 */
var ActionSheetContentVue2 = {
  name: 'KendoActionSheetContent',
  props: {
    contentClassName: {
      type: String,
      default: undefined
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlot = (0, kendo_vue_common_1.getDefaultSlots)(this);
    var contentClassName = this.$props.contentClassName;
    return h("div", {
      "class": (0, kendo_vue_common_1.classNames)('k-actionsheet-content', contentClassName)
    }, [defaultSlot]);
  }
};
exports.ActionSheetContentVue2 = ActionSheetContentVue2;
/**
 * @hidden
 */
var ActionSheetContent = ActionSheetContentVue2;
exports.ActionSheetContent = ActionSheetContent;
