// @progress/kendo-vue-layout/dist/npm/actionsheet/ActionSheetHeader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ActionSheetHeaderVue2 = exports.ActionSheetHeader = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
/**
 * @hidden
 */
var ActionSheetHeaderVue2 = {
  name: 'KendoActionSheetHeader',
  props: {
    headerClassName: {
      type: String,
      default: undefined
    },
    headerStyle: {
      type: String,
      default: undefined
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlot = (0, kendo_vue_common_1.getDefaultSlots)(this);
    var _a = this.$props,
      headerStyle = _a.headerStyle,
      headerClassName = _a.headerClassName;
    return h("div", {
      style: headerStyle,
      "class": (0, kendo_vue_common_1.classNames)('k-actionsheet-titlebar', headerClassName)
    }, [defaultSlot]);
  }
};
exports.ActionSheetHeaderVue2 = ActionSheetHeaderVue2;
/**
 * @hidden
 */
var ActionSheetHeader = ActionSheetHeaderVue2;
exports.ActionSheetHeader = ActionSheetHeader;
