// @progress/kendo-vue-layout/dist/npm/actionsheet/ActionSheetFooter.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ActionSheetFooterVue2 = exports.ActionSheetFooter = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
/**
 * @hidden
 */
var ActionSheetFooterVue2 = {
  name: 'KendoActionSheetFooter',
  props: {
    footerClassName: {
      type: String,
      default: undefined
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlot = (0, kendo_vue_common_1.getDefaultSlots)(this);
    var footerClassName = this.$props.footerClassName;
    return h("div", {
      "class": (0, kendo_vue_common_1.classNames)('k-actionsheet-footer', footerClassName)
    }, [defaultSlot]);
  }
};
exports.ActionSheetFooterVue2 = ActionSheetFooterVue2;
/**
 * @hidden
 */
var ActionSheetFooter = ActionSheetFooterVue2;
exports.ActionSheetFooter = ActionSheetFooter;
