// @progress/kendo-vue-dateinputs/dist/npm/calendar/utils/main.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractActiveRange = exports.extractFocusedDate = exports.extractRangeFromValue = exports.extractMultipleFromValue = exports.extractDateFromValue = exports.calculateValue = void 0;
var utils_1 = require("../../utils");
var main_1 = require("../models/main");
var kendo_date_math_1 = require("@progress/kendo-date-math");
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
/**
 * @hidden
 */
var calculateValue = function (min, max, stateValue, propValue) {
    return propValue !== undefined
        ? propValue !== null && (0, utils_1.isInRange)(propValue, min, max)
            ? propValue
            : null
        : stateValue !== null && (0, utils_1.isInRange)(stateValue, min, max)
            ? stateValue
            : null;
};
exports.calculateValue = calculateValue;
/**
 * @hidden
 */
var extractDateFromValue = function (min, max, value) {
    return value instanceof Date && !Array.isArray(value) && (0, utils_1.isInRange)((0, kendo_date_math_1.getDate)(value), min, max)
        ? (0, kendo_date_math_1.getDate)(value)
        : null;
};
exports.extractDateFromValue = extractDateFromValue;
/**
 * @hidden
 */
var extractMultipleFromValue = function (min, max, value) {
    return Array.isArray(value)
        ? value.filter(function (date) { return (0, utils_1.isInRange)(date, min, max); }).map(function (date) { return (0, kendo_date_math_1.getDate)(date); })
        : null;
};
exports.extractMultipleFromValue = extractMultipleFromValue;
/**
 * @hidden
 */
var extractRangeFromValue = function (value) {
    return (0, kendo_vue_common_1.isObject)(value) && !(value instanceof Date) && value !== null && !Array.isArray(value)
        ? value
        : main_1.EMPTY_SELECTIONRANGE;
};
exports.extractRangeFromValue = extractRangeFromValue;
/**
 * @hidden
 */
var extractFocusedDate = function (single, multiple, range) {
    return single || (multiple && multiple[0]) || (range && range.start);
};
exports.extractFocusedDate = extractFocusedDate;
/**
 * @hidden
 */
var extractActiveRange = function (range, single) {
    return range.start === null && single === null
        ? 'start'
        : range.end === null
            ? 'end'
            : 'start';
};
exports.extractActiveRange = extractActiveRange;
