// @progress/kendo-vue-data-tools/dist/npm/columnmenu/ColumnMenu.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ColumnMenuVue2 = exports.ColumnMenu = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var inject = allVue.inject;
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
var kendo_vue_popup_1 = require("@progress/kendo-vue-popup");
var kendo_svg_icons_1 = require("@progress/kendo-svg-icons");
var ColumnMenuContent_1 = require("./ColumnMenuContent");
var utils_1 = require("../navigation/utils");
var kendo_vue_intl_1 = require("@progress/kendo-vue-intl");
var main_1 = require("../messages/main");
/**
 * @hidden
 */
var ColumnMenuVue2 = {
  name: 'KendoColumnMenu',
  props: {
    animate: {
      type: [Boolean, Object],
      default: function _default() {
        return true;
      }
    },
    column: Object,
    sortable: [Boolean, Object],
    sort: {
      type: Array
    },
    opened: Boolean,
    filter: Object,
    filterable: Boolean,
    filterOperators: Object,
    render: [Boolean, String, Function, Object]
  },
  inject: {
    kendoIntlService: {
      default: null
    },
    kendoLocalizationService: {
      default: null
    }
  },
  data: function data() {
    return {
      show: false,
      blurTimeout: undefined
    };
  },
  created: function created() {
    this._anchor = (0, kendo_vue_common_1.guid)();
  },
  computed: {
    currentOpen: function currentOpen() {
      return this.show || this.$props.opened;
    }
  },
  watch: {
    currentOpen: function currentOpen(newOpened) {
      this.$nextTick(function () {
        this._content = this.$refs.content;
        if (newOpened && this._content) {
          var focusableElements = utils_1.tableKeyboardNavigationTools.getFocusableElements(this._content, {
            focusable: true
          });
          if (focusableElements.length) {
            focusableElements[0].focus();
          } else {
            this._content.focus();
          }
        }
      });
    }
  },
  methods: {
    blur: function blur() {
      var _this = this;
      clearTimeout(this.blurTimeout);
      this.blurTimeout = setTimeout(function () {
        _this.closeMenu();
      }, 200);
    },
    focus: function focus() {
      clearTimeout(this.blurTimeout);
    },
    handleFocus: function handleFocus(_) {
      clearTimeout(this.blurTimeout);
    },
    anchorClick: function anchorClick(e) {
      e.preventDefault();
      this.show = !this.show;
    },
    closeMenu: function closeMenu() {
      this.$emit('close');
      this.show = false;
    },
    sortChange: function sortChange(newDescriptor, e) {
      this.$emit('sortchange', newDescriptor, e);
    },
    filterChange: function filterChange(newDescriptor, e) {
      this.$emit('filterchange', newDescriptor, e);
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    var kendoLocalizationService = inject('kendoLocalizationService', {});
    return {
      v3: v3,
      kendoLocalizationService: kendoLocalizationService
    };
  },
  // @ts-ignore
  render: function render(createElement) {
    var _this2 = this;
    var h = gh || createElement;
    var renderTemplate = this.$props.render;
    var ls = (0, kendo_vue_intl_1.provideLocalizationService)(this);
    return h("div", {
      style: {
        display: 'inline'
      }
    }, [h("div", {
      "class": 'k-grid-column-menu k-grid-filter',
      tabindex: 0,
      attrs: this.v3 ? undefined : {
        tabindex: 0,
        title: "".concat(this.$props.column.field, " ").concat(ls.toLanguageString(main_1.columnMenu, main_1.messages[main_1.columnMenu]))
      },
      title: "".concat(this.$props.column.field, " ").concat(ls.toLanguageString(main_1.columnMenu, main_1.messages[main_1.columnMenu])),
      ref: (0, kendo_vue_common_1.setRef)(this, 'kendoAnchor', this._anchor),
      onClick: this.anchorClick,
      on: this.v3 ? undefined : {
        "click": this.anchorClick
      }
    }, [h(kendo_vue_common_1.Icon, {
      name: 'more-vertical',
      attrs: this.v3 ? undefined : {
        name: 'more-vertical',
        icon: kendo_svg_icons_1.moreVerticalIcon
      },
      icon: kendo_svg_icons_1.moreVerticalIcon
    })]),
    // @ts-ignore function children
    h(kendo_vue_popup_1.Popup, {
      animate: this.$props.animate,
      attrs: this.v3 ? undefined : {
        animate: this.$props.animate,
        anchor: this._anchor,
        show: this.currentOpen
      },
      anchor: this._anchor,
      show: this.currentOpen
    }, this.v3 ? function () {
      return [h("div", {
        ref: 'content',
        "class": 'k-grid-columnmenu-popup',
        tabindex: 0,
        attrs: _this2.v3 ? undefined : {
          tabindex: 0
        },
        onFocusout: _this2.blur,
        on: _this2.v3 ? undefined : {
          "focusout": _this2.blur,
          "focusin": _this2.focus
        },
        onFocusin: _this2.focus,
        style: {
          outline: 'none',
          width: '230px'
        }
      }, [
      // @ts-ignore
      h(ColumnMenuContent_1.ColumnMenuContent, {
        column: _this2.$props.column,
        attrs: _this2.v3 ? undefined : {
          column: _this2.$props.column,
          sortable: _this2.$props.sortable,
          sort: _this2.$props.sort,
          filter: _this2.$props.filter,
          filterable: _this2.$props.filterable,
          filterOperators: _this2.$props.filterOperators,
          render: renderTemplate
        },
        sortable: _this2.$props.sortable,
        sort: _this2.$props.sort,
        filter: _this2.$props.filter,
        filterable: _this2.$props.filterable,
        filterOperators: _this2.$props.filterOperators,
        render: renderTemplate,
        onContentfocus: _this2.handleFocus,
        on: _this2.v3 ? undefined : {
          "contentfocus": _this2.handleFocus,
          "closemenu": _this2.closeMenu,
          "sortchange": _this2.sortChange,
          "filterchange": _this2.filterChange
        },
        onClosemenu: _this2.closeMenu,
        onSortchange: _this2.sortChange,
        onFilterchange: _this2.filterChange
      })])];
    } : [h("div", {
      ref: 'content',
      "class": 'k-grid-columnmenu-popup',
      tabindex: 0,
      attrs: _this2.v3 ? undefined : {
        tabindex: 0
      },
      onFocusout: _this2.blur,
      on: _this2.v3 ? undefined : {
        "focusout": _this2.blur,
        "focusin": _this2.focus
      },
      onFocusin: _this2.focus,
      style: {
        outline: 'none',
        width: '230px'
      }
    }, [h(ColumnMenuContent_1.ColumnMenuContent, {
      column: _this2.$props.column,
      attrs: _this2.v3 ? undefined : {
        column: _this2.$props.column,
        sortable: _this2.$props.sortable,
        sort: _this2.$props.sort,
        filter: _this2.$props.filter,
        filterable: _this2.$props.filterable,
        filterOperators: _this2.$props.filterOperators,
        render: renderTemplate
      },
      sortable: _this2.$props.sortable,
      sort: _this2.$props.sort,
      filter: _this2.$props.filter,
      filterable: _this2.$props.filterable,
      filterOperators: _this2.$props.filterOperators,
      render: renderTemplate,
      onContentfocus: _this2.handleFocus,
      on: _this2.v3 ? undefined : {
        "contentfocus": _this2.handleFocus,
        "closemenu": _this2.closeMenu,
        "sortchange": _this2.sortChange,
        "filterchange": _this2.filterChange
      },
      onClosemenu: _this2.closeMenu,
      onSortchange: _this2.sortChange,
      onFilterchange: _this2.filterChange
    })])])]);
  }
};
exports.ColumnMenuVue2 = ColumnMenuVue2;
/**
 * @hidden
 */
var ColumnMenu = ColumnMenuVue2;
exports.ColumnMenu = ColumnMenu;
