// @progress/kendo-vue-data-tools/dist/npm/columnmenu/ColumnMenuItemContent.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ColumnMenuItemContentVue2 = exports.ColumnMenuItemContent = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
var kendo_vue_animation_1 = require("@progress/kendo-vue-animation");
/**
 * @hidden
 */
var ColumnMenuItemContentVue2 = {
  props: {
    show: Boolean
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  // @ts-ignore
  render: function render(createElement) {
    var _this = this;
    var h = gh || createElement;
    var defaultSlot = (0, kendo_vue_common_1.getDefaultSlots)(this);
    return h("div", {
      "class": 'k-columnmenu-item-content'
    }, [
    // @ts-ignore function children
    h(kendo_vue_animation_1.Reveal, {
      appear: this.$props.show,
      attrs: this.v3 ? undefined : {
        appear: this.$props.show
      },
      style: {
        position: 'relative',
        display: 'block'
      }
    }, this.v3 ? function () {
      return [_this.$props.show ? defaultSlot : null];
    } : [_this.$props.show ? defaultSlot : null])]);
  }
};
exports.ColumnMenuItemContentVue2 = ColumnMenuItemContentVue2;
/**
 * @hidden
 */
var ColumnMenuItemContent = ColumnMenuItemContentVue2;
exports.ColumnMenuItemContent = ColumnMenuItemContent;
