// @progress/kendo-vue-grid/dist/npm/drag/DragClue.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DragClueVue2 = exports.DragClue = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
var kendo_svg_icons_1 = require("@progress/kendo-svg-icons");
/**
 * @hidden
 */
var DragClueVue2 = {
  data: function data() {
    return {
      visible: false,
      top: 0,
      left: 0,
      innerText: '',
      status: 'cancel'
    };
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  // @ts-ignore
  render: function render(createElement) {
    var h = gh || createElement;
    return this.visible && h("div", {
      "class": "k-header k-drag-clue",
      style: {
        display: 'block',
        position: 'absolute',
        zIndex: 20000,
        padding: '8px 12px',
        top: this.top + 'px',
        left: this.left + 'px'
      }
    }, [h(kendo_vue_common_1.Icon, {
      name: this.status,
      attrs: this.v3 ? undefined : {
        name: this.status,
        icon: this.status === 'cancel' ? kendo_svg_icons_1.cancelIcon : kendo_svg_icons_1.plusIcon
      },
      "class": 'k-drag-status',
      icon: this.status === 'cancel' ? kendo_svg_icons_1.cancelIcon : kendo_svg_icons_1.plusIcon
    }), this.innerText]);
  }
};
exports.DragClueVue2 = DragClueVue2;
/**
 * @hidden
 */
var DragClue = DragClueVue2;
exports.DragClue = DragClue;
