// @progress/kendo-vue-grid/dist/npm/drag/GroupingIndicator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GroupingIndicatorVue2 = exports.GroupingIndicator = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var ref = allVue.ref;
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
var kendo_svg_icons_1 = require("@progress/kendo-svg-icons");
/**
 * @hidden
 */
var GroupingIndicatorVue2 = {
  props: {
    title: String,
    dir: String // 'asc' | 'desc';
  },

  mounted: function mounted() {
    this.element = (0, kendo_vue_common_1.getRef)(this, 'indicatorContainer');
    if (this.element) {
      this.draggable = this.$refs.draggable;
    }
  },
  methods: {
    press: function press(event) {
      if (this.element) {
        this.$emit('press', event, this.element);
      }
    },
    drag: function drag(event) {
      if (this.element) {
        this.$emit('drag', event, this.element);
      }
    },
    release: function release(event) {
      if (this.element) {
        this.$emit('release', event, this.element);
      }
    },
    sortChange: function sortChange(event) {
      event.preventDefault();
      var sort = this.$props.dir === 'asc' ? 'desc' : 'asc';
      this.$emit('sortChange', event, sort);
    },
    groupRemove: function groupRemove(event) {
      event.preventDefault();
      this.$emit('remove', event);
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    var indicatorContainerRef = ref(null);
    return {
      v3: v3,
      indicatorContainerRef: indicatorContainerRef
    };
  },
  // @ts-ignore
  render: function render(createElement) {
    var _this = this;
    var h = gh || createElement;
    return (
      // @ts-ignore  function children
      h(kendo_vue_common_1.Draggable, {
        onPress: this.press,
        on: this.v3 ? undefined : {
          "press": this.press,
          "drag": this.drag,
          "release": this.release
        },
        onDrag: this.drag,
        onRelease: this.release,
        ref: 'draggable'
      }, this.v3 ? function () {
        return [h("div", {
          "class": "k-chip k-chip-md k-chip-solid k-chip-solid-base k-rounded-md",
          ref: (0, kendo_vue_common_1.setRef)(_this, 'indicatorContainer')
        }, [h(kendo_vue_common_1.Icon, {
          name: 'sort-' + _this.$props.dir + '-small',
          attrs: _this.v3 ? undefined : {
            name: 'sort-' + _this.$props.dir + '-small',
            icon: _this.$props.dir === 'asc' ? kendo_svg_icons_1.sortAscSmallIcon : kendo_svg_icons_1.sortDescSmallIcon
          },
          "class": 'k-chip-icon',
          icon: _this.$props.dir === 'asc' ? kendo_svg_icons_1.sortAscSmallIcon : kendo_svg_icons_1.sortDescSmallIcon
        }), h("span", {
          "class": "k-chip-content",
          tabindex: -1,
          attrs: _this.v3 ? undefined : {
            tabindex: -1
          },
          onClick: _this.sortChange,
          on: _this.v3 ? undefined : {
            "click": _this.sortChange
          }
        }, [_this.$props.title]), h("span", {
          "class": "k-chip-actions"
        }, [h("span", {
          "class": "k-chip-action k-chip-remove-action",
          tabindex: -1,
          attrs: _this.v3 ? undefined : {
            tabindex: -1
          },
          onClick: _this.groupRemove,
          on: _this.v3 ? undefined : {
            "click": _this.groupRemove
          }
        }, [h(kendo_vue_common_1.Icon, {
          name: "x-circle",
          attrs: _this.v3 ? undefined : {
            name: "x-circle",
            icon: kendo_svg_icons_1.xCircleIcon
          },
          icon: kendo_svg_icons_1.xCircleIcon
        })])])])];
      } : [h("div", {
        "class": "k-chip k-chip-md k-chip-solid k-chip-solid-base k-rounded-md",
        ref: (0, kendo_vue_common_1.setRef)(_this, 'indicatorContainer')
      }, [h(kendo_vue_common_1.Icon, {
        name: 'sort-' + _this.$props.dir + '-small',
        attrs: _this.v3 ? undefined : {
          name: 'sort-' + _this.$props.dir + '-small',
          icon: _this.$props.dir === 'asc' ? kendo_svg_icons_1.sortAscSmallIcon : kendo_svg_icons_1.sortDescSmallIcon
        },
        "class": 'k-chip-icon',
        icon: _this.$props.dir === 'asc' ? kendo_svg_icons_1.sortAscSmallIcon : kendo_svg_icons_1.sortDescSmallIcon
      }), h("span", {
        "class": "k-chip-content",
        tabindex: -1,
        attrs: _this.v3 ? undefined : {
          tabindex: -1
        },
        onClick: _this.sortChange,
        on: _this.v3 ? undefined : {
          "click": _this.sortChange
        }
      }, [_this.$props.title]), h("span", {
        "class": "k-chip-actions"
      }, [h("span", {
        "class": "k-chip-action k-chip-remove-action",
        tabindex: -1,
        attrs: _this.v3 ? undefined : {
          tabindex: -1
        },
        onClick: _this.groupRemove,
        on: _this.v3 ? undefined : {
          "click": _this.groupRemove
        }
      }, [h(kendo_vue_common_1.Icon, {
        name: "x-circle",
        attrs: _this.v3 ? undefined : {
          name: "x-circle",
          icon: kendo_svg_icons_1.xCircleIcon
        },
        icon: kendo_svg_icons_1.xCircleIcon
      })])])])])
    );
  }
};
exports.GroupingIndicatorVue2 = GroupingIndicatorVue2;
/**
 * @hidden
 */
var GroupingIndicator = GroupingIndicatorVue2;
exports.GroupingIndicator = GroupingIndicator;
