// @progress/kendo-vue-grid/dist/npm/footer/Footer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FooterVue2 = exports.Footer = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var ref = allVue.ref;
var browser_support_service_1 = require("../utils/browser-support.service");
var main_1 = require("../utils/main");
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
/**
 * @hidden
 */
var FooterVue2 = {
  name: 'Footer',
  props: {
    staticHeaders: Boolean,
    row: Object,
    columnResize: Object,
    cols: Array,
    size: String
  },
  data: function data() {
    return {
      scrollbarWidth: 0,
      rtl: false,
      tableWidth: null
    };
  },
  computed: {
    tableClass: function tableClass() {
      var _a;
      var size = this.$props.size;
      return _a = {
        'k-table': true,
        'k-grid-footer-table': true
      }, _a["k-table-".concat(kendo_vue_common_1.kendoThemeMaps.sizeMap[size] || size)] = size, _a;
    },
    divStyle: {
      get: function get() {
        var padding = this.$data.scrollbarWidth + 'px';
        var right = this.rtl ? 0 : padding;
        var left = this.rtl ? padding : 0;
        return {
          padding: "0 ".concat(right, " 0 ").concat(left)
        };
      }
    },
    tableStyle: {
      get: function get() {
        return this.tableWidth ? {
          width: this.tableWidth
        } : null;
      }
    }
  },
  mounted: function mounted() {
    this.$data.scrollbarWidth = new browser_support_service_1.BrowserSupportService().scrollbarWidth;
    this.rtl = (0, main_1.isRtl)(this.$el);
    this.$props.columnResize.colGroupFooter = (0, kendo_vue_common_1.getRef)(this, 'colGroupHeader');
    this._footerWrap = (0, kendo_vue_common_1.getRef)(this, 'footerWrap');
  },
  methods: {
    setScrollLeft: function setScrollLeft(scrollLeft) {
      if (this._footerWrap) {
        this._footerWrap.scrollLeft = scrollLeft;
      }
    },
    setWidth: function setWidth(width) {
      this.$data.tableWidth = width + 'px';
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    var footerWrapRef = ref(null);
    var colGroupHeaderRef = ref(null);
    return {
      v3: v3,
      colGroupHeaderRef: colGroupHeaderRef,
      footerWrapRef: footerWrapRef
    };
  },
  // @ts-ignore
  render: function render(createElement) {
    var h = gh || createElement;
    if (!this.$props.staticHeaders) {
      return h("tfoot", {
        "class": "k-table-tfoot k-grid-footer",
        role: "rowgroup",
        attrs: this.v3 ? undefined : {
          role: "rowgroup"
        }
      }, [this.$props.row]);
    }
    return h("div", {
      "class": "k-grid-footer",
      style: this.divStyle,
      role: "presentation",
      attrs: this.v3 ? undefined : {
        role: "presentation"
      }
    }, [h("div", {
      "class": "k-grid-footer-wrap",
      ref: (0, kendo_vue_common_1.setRef)(this, 'footerWrap'),
      role: "presentation",
      attrs: this.v3 ? undefined : {
        role: "presentation"
      }
    }, [h("table", {
      "class": this.tableClass,
      style: this.tableStyle,
      role: "presentation",
      attrs: this.v3 ? undefined : {
        role: "presentation"
      }
    }, [h("colgroup", {
      ref: (0, kendo_vue_common_1.setRef)(this, 'colGroupHeader')
    }, [this.$props.cols]), h("tfoot", {
      "class": 'k-table-tfoot',
      role: "rowgroup",
      attrs: this.v3 ? undefined : {
        role: "rowgroup"
      }
    }, [this.$props.row])])])]);
  }
};
exports.FooterVue2 = FooterVue2;
/**
 * @hidden
 */
var Footer = FooterVue2;
exports.Footer = Footer;
