// @progress/kendo-vue-layout/dist/es/actionsheet/ActionSheetHeader.js
// @ts-ignore
import * as Vue from 'vue';
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
import { classNames, getDefaultSlots } from '@progress/kendo-vue-common';
/**
 * @hidden
 */
var ActionSheetHeaderVue2 = {
  name: 'KendoActionSheetHeader',
  props: {
    headerClassName: {
      type: String,
      default: undefined
    },
    headerStyle: {
      type: String,
      default: undefined
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlot = getDefaultSlots(this);
    var _a = this.$props,
      headerStyle = _a.headerStyle,
      headerClassName = _a.headerClassName;
    return h("div", {
      style: headerStyle,
      "class": classNames('k-actionsheet-titlebar', headerClassName)
    }, [defaultSlot]);
  }
};
/**
 * @hidden
 */
var ActionSheetHeader = ActionSheetHeaderVue2;
export { ActionSheetHeader, ActionSheetHeaderVue2 };
