// @progress/kendo-vue-dialogs/dist/npm/DialogActionsBar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DialogActionsBarVue2 = exports.DialogActionsBar = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
/**
 * @hidden
 */
var DialogActionsBarVue2 = {
  name: 'DialogActionsBar',
  props: {
    layout: {
      type: String,
      default: 'stretched',
      validator: function validator(value) {
        return ['stretched', 'start', 'center', 'end'].includes(value);
      }
    },
    orientation: {
      type: String,
      default: 'horizontal',
      validator: function validator(value) {
        return ['horizontal', 'vertical'].includes(value);
      }
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  computed: {
    wrapperClasses: function wrapperClasses() {
      var _a;
      var _b = this.$props,
        layout = _b.layout,
        orientation = _b.orientation;
      return _a = {
        'k-actions': true,
        'k-dialog-actions': true,
        'k-window-actions': true
      }, _a["k-actions-".concat(orientation)] = orientation, _a["k-actions-".concat(layout)] = layout, _a;
    }
  },
  // @ts-ignore
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlot = (0, kendo_vue_common_1.getDefaultSlots)(this);
    return h("div", {
      "class": this.wrapperClasses
    }, [defaultSlot]);
  }
};
exports.DialogActionsBarVue2 = DialogActionsBarVue2;
/**
 * @hidden
 */
var DialogActionsBar = DialogActionsBarVue2;
exports.DialogActionsBar = DialogActionsBar;
