// @progress/kendo-vue-dialogs/dist/npm/DialogTitleBar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DialogTitleBarVue2 = exports.DialogTitleBar = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
var kendo_vue_buttons_1 = require("@progress/kendo-vue-buttons");
var kendo_svg_icons_1 = require("@progress/kendo-svg-icons");
/**
 * @hidden
 */
var DialogTitleBarVue2 = {
  props: {
    id: String,
    closeIcon: {
      type: Boolean,
      default: true
    },
    title: String,
    titleRender: [String, Function, Object]
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  methods: {
    onCloseButtonClick: function onCloseButtonClick(e) {
      this.$emit('closebuttonclick', e);
    }
  },
  // @ts-ignore
  render: function render(createElement) {
    var h = gh || createElement;
    var _a = this.$props,
      id = _a.id,
      closeIcon = _a.closeIcon,
      titleRender = _a.titleRender,
      title = _a.title;
    var titleElement;
    titleElement = kendo_vue_common_1.getTemplate.call(this, {
      h: h,
      template: titleRender,
      defaultRendering: title
    });
    return h("div", {
      "class": "k-window-titlebar k-dialog-titlebar",
      id: id,
      attrs: this.v3 ? undefined : {
        id: id
      }
    }, [h("span", {
      "class": "k-window-title k-dialog-title"
    }, [titleElement]), h("div", {
      "class": "k-window-titlebar-actions k-dialog-titlebar-actions"
    }, [closeIcon &&
    // @ts-ignore
    h(kendo_vue_buttons_1.Button, {
      type: "button",
      attrs: this.v3 ? undefined : {
        type: "button",
        fillMode: 'flat',
        "aria-label": "Close",
        icon: 'x',
        svgIcon: kendo_svg_icons_1.xIcon
      },
      fillMode: 'flat',
      "aria-label": "Close",
      icon: 'x',
      svgIcon: kendo_svg_icons_1.xIcon,
      onClick: this.onCloseButtonClick,
      on: this.v3 ? undefined : {
        "click": this.onCloseButtonClick
      },
      "class": "k-window-titlebar-action k-dialog-titlebar-action"
    })])]);
  }
};
exports.DialogTitleBarVue2 = DialogTitleBarVue2;
/**
 * @hidden
 */
var DialogTitleBar = DialogTitleBarVue2;
exports.DialogTitleBar = DialogTitleBar;
