// @progress/kendo-vue-dialogs/dist/npm/main.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WindowVue2 = exports.Window = exports.DialogActionsBarVue2 = exports.DialogActionsBar = exports.DialogVue2 = exports.Dialog = void 0;
var Dialog_1 = require("./Dialog");
Object.defineProperty(exports, "Dialog", { enumerable: true, get: function () { return Dialog_1.Dialog; } });
Object.defineProperty(exports, "DialogVue2", { enumerable: true, get: function () { return Dialog_1.DialogVue2; } });
var DialogActionsBar_1 = require("./DialogActionsBar");
Object.defineProperty(exports, "DialogActionsBar", { enumerable: true, get: function () { return DialogActionsBar_1.DialogActionsBar; } });
Object.defineProperty(exports, "DialogActionsBarVue2", { enumerable: true, get: function () { return DialogActionsBar_1.DialogActionsBarVue2; } });
var Window_1 = require("./Window");
Object.defineProperty(exports, "Window", { enumerable: true, get: function () { return Window_1.Window; } });
Object.defineProperty(exports, "WindowVue2", { enumerable: true, get: function () { return Window_1.WindowVue2; } });
