// @progress/kendo-vue-layout/dist/npm/card/Avatar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AvatarVue2 = exports.Avatar = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var Enums_1 = require("./interfaces/Enums");
var package_metadata_1 = require("../package-metadata");
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
/**
 * @hidden
 */
var AvatarVue2 = {
  name: 'KendoAvatar',
  props: {
    type: {
      type: String,
      default: Enums_1.avatarType.TEXT,
      validator: function validator(value) {
        return ['text', 'image', 'icon'].includes(value);
      }
    },
    border: Boolean,
    rounded: {
      type: String,
      default: 'medium',
      validator: function validator(value) {
        return ['small', 'medium', 'large', 'full'].includes(value);
      }
    },
    fillMode: {
      type: String,
      default: 'solid',
      validator: function validator(value) {
        return ['solid', 'outline'].includes(value);
      }
    },
    size: {
      type: String,
      default: 'medium',
      validator: function validator(value) {
        return ['small', 'medium', 'large'].includes(value);
      }
    },
    themeColor: {
      type: String,
      default: 'primary',
      validator: function validator(value) {
        return ['primary', 'secondary', 'tertiary', 'info', 'success', 'warning', 'error', 'dark', 'light', 'inverse', 'inherit'].includes(value);
      }
    }
  },
  created: function created() {
    (0, kendo_vue_common_1.validatePackage)(package_metadata_1.packageMetadata);
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  computed: {
    wrapperClass: function wrapperClass() {
      var _a;
      var _b = this.$props,
        border = _b.border,
        type = _b.type,
        size = _b.size,
        rounded = _b.rounded,
        fillMode = _b.fillMode,
        themeColor = _b.themeColor;
      return _a = {
        'k-avatar-bordered': border,
        'k-avatar': true
      }, _a["k-avatar-".concat(kendo_vue_common_1.kendoThemeMaps.sizeMap[size] || size)] = size, _a["k-rounded-".concat(kendo_vue_common_1.kendoThemeMaps.roundedMap[rounded] || rounded)] = rounded, _a["k-avatar-".concat(fillMode)] = fillMode, _a["k-avatar-".concat(fillMode, "-").concat(themeColor)] = Boolean(fillMode && themeColor), _a;
    },
    innerClass: function innerClass() {
      var _a;
      return _a = {}, _a["k-avatar-".concat(this.$props.type)] = this.$props.type, _a;
    }
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = (0, kendo_vue_common_1.getDefaultSlots)(this);
    return h("div", {
      "class": this.wrapperClass
    }, [h("span", {
      "class": this.innerClass
    }, [defaultSlots])]);
  }
};
exports.AvatarVue2 = AvatarVue2;
/**
 * @hidden
 */
var Avatar = AvatarVue2;
exports.Avatar = Avatar;
