// @progress/kendo-vue-layout/dist/npm/card/CardActions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CardActionsVue2 = exports.CardActions = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
var Enums_1 = require("./interfaces/Enums");
/**
 * @hidden
 */
var CardActionsVue2 = {
  name: 'KendoCardActions',
  props: {
    layout: {
      type: String,
      default: Enums_1.cardActionsLayout.START,
      validator: function validator(value) {
        return ['stretched', 'start', 'center', 'end'].includes(value);
      }
    },
    orientation: {
      type: String,
      default: Enums_1.cardOrientation.HORIZONTAL,
      validator: function validator(value) {
        return ['horizontal', 'vertical'].includes(value);
      }
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  computed: {
    wrapperClass: function wrapperClass() {
      var _a;
      return _a = {
        'k-card-actions': true
      }, _a["k-card-actions-".concat(this.$props.layout)] = true, _a["k-card-actions-".concat(this.$props.orientation !== Enums_1.cardOrientation.HORIZONTAL ? 'vertical' : 'horizontal')] = true, _a;
    }
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = (0, kendo_vue_common_1.getDefaultSlots)(this);
    return h("div", {
      "class": this.wrapperClass
    }, [defaultSlots]);
  }
};
exports.CardActionsVue2 = CardActionsVue2;
/**
 * @hidden
 */
var CardActions = CardActionsVue2;
exports.CardActions = CardActions;
