// @progress/kendo-vue-layout/dist/npm/card/Card.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CardVue2 = exports.Card = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var Enums_1 = require("./interfaces/Enums");
var Enums_2 = require("./interfaces/Enums");
var package_metadata_1 = require("../package-metadata");
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
/**
 * @hidden
 */
var CardVue2 = {
  name: 'KendoCard',
  props: {
    dir: String,
    type: {
      type: String,
      default: Enums_1.cardType.DEFAULT,
      validator: function validator(value) {
        return ['default', 'primary', 'info', 'success', 'warning', 'error'].includes(value);
      }
    },
    orientation: {
      type: String,
      default: Enums_2.cardOrientation.VERTICAL,
      validator: function validator(value) {
        return ['horizontal', 'vertical'].includes(value);
      }
    }
  },
  created: function created() {
    (0, kendo_vue_common_1.validatePackage)(package_metadata_1.packageMetadata);
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  computed: {
    wrapperClass: function wrapperClass() {
      var _a;
      var orientationClass = this.$props.orientation !== Enums_2.cardOrientation.HORIZONTAL ? 'vertical' : 'horizontal';
      return _a = {
        'k-card': true
      }, _a["k-card-".concat(this.$props.type)] = this.$props.type !== Enums_1.cardType.DEFAULT, _a["k-card-".concat(orientationClass)] = true, _a;
    }
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = (0, kendo_vue_common_1.getDefaultSlots)(this);
    return h("div", {
      dir: this.$props.dir,
      attrs: this.v3 ? undefined : {
        dir: this.$props.dir
      },
      "class": this.wrapperClass
    }, [defaultSlots]);
  }
};
exports.CardVue2 = CardVue2;
/**
 * @hidden
 */
var Card = CardVue2;
exports.Card = Card;
