// @progress/kendo-vue-grid/dist/npm/header/GridHeaderSelectionCell.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GridHeaderSelectionCellVue2 = exports.GridHeaderSelectionCell = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
var kendo_vue_intl_1 = require("@progress/kendo-vue-intl");
var main_1 = require("../messages/main");
/**
 * @hidden
 */
var GridHeaderSelectionCellVue2 = {
  props: {
    field: String,
    title: String,
    selectionValue: Boolean,
    render: [Object, Function, String]
  },
  created: function created() {
    this.inputId = (0, kendo_vue_common_1.guid)();
  },
  inject: {
    kendoLocalizationService: {
      default: null
    }
  },
  methods: {
    changeHandle: function changeHandle(e) {
      this.$emit('selectionchange', {
        field: this.$props.field,
        event: e
      });
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  // @ts-ignore
  render: function render(createElement) {
    var h = gh || createElement;
    var renderTemplate = this.$props.render;
    var ls = (0, kendo_vue_intl_1.provideLocalizationService)(this);
    var selectAllRowsMessage = ls.toLanguageString(main_1.selectAllRows, main_1.messages[main_1.selectAllRows]);
    var input = h("input", {
      key: 0,
      checked: this.v3 ? this.$props.selectionValue : null,
      domProps: this.v3 ? undefined : {
        "checked": this.$props.selectionValue
      },
      id: this.inputId,
      attrs: this.v3 ? undefined : {
        id: this.inputId,
        type: "checkbox",
        "aria-label": selectAllRowsMessage
      },
      type: "checkbox",
      "class": "k-checkbox k-checkbox-md k-rounded-md",
      "aria-label": selectAllRowsMessage,
      onChange: this.changeHandle,
      on: this.v3 ? undefined : {
        "change": this.changeHandle
      }
    });
    var defaultRendering = kendo_vue_common_1.getTemplate.call(this, {
      h: h,
      template: renderTemplate,
      defaultRendering: input,
      additionalProps: this.$props,
      additionalListeners: {
        selectionchange: this.changeHandle
      }
    });
    return h("span", {
      "class": "k-link"
    }, [h("span", {
      "class": "k-column-title"
    }, [defaultRendering])]);
  }
};
exports.GridHeaderSelectionCellVue2 = GridHeaderSelectionCellVue2;
/**
 * @hidden
 */
var GridHeaderSelectionCell = GridHeaderSelectionCellVue2;
exports.GridHeaderSelectionCell = GridHeaderSelectionCell;
