// @progress/kendo-vue-inputs/dist/npm/messages/main.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.messages = exports.signatureMinimize = exports.signatureMaximize = exports.signatureClear = exports.ratingAriaLabel = exports.colorGradientToggleInputsButton = exports.colorPickerDropdownButtonAriaLabel = exports.switchValidation = exports.radioButtonValidation = exports.checkboxOptionalText = exports.checkboxValidation = exports.flatColorPickerClearBtn = exports.flatColorPickerPaletteBtn = exports.flatColorPickerGradientBtn = exports.flatColorPickerApplyBtn = exports.flatColorPickerCancelBtn = exports.colorGradientAlphaSliderLabel = exports.colorGradientHueSliderLabel = exports.colorGradientFail = exports.colorGradientPass = exports.colorGradientAAALevel = exports.colorGradientAALevel = exports.colorGradientContrastRatio = exports.colorGradientALabel = exports.colorGradientBLabel = exports.colorGradientGLabel = exports.colorGradientRLabel = exports.colorGradientInputColorButton = exports.colorGradientHexLabel = exports.colorGradientHex = exports.colorGradientA = exports.colorGradientB = exports.colorGradientG = exports.colorGradientR = exports.colorGradientDragHandle = exports.sliderDragTitle = exports.sliderDecreaseValue = exports.sliderIncreaseValue = exports.numericDecreaseValue = exports.numericIncreaseValue = void 0;
/**
 * @hidden
 */
exports.numericIncreaseValue = 'numerictextbox.increment';
/**
 * @hidden
 */
exports.numericDecreaseValue = 'numerictextbox.decrement';
/**
 * @hidden
 */
exports.sliderIncreaseValue = 'slider.increment';
/**
 * @hidden
 */
exports.sliderDecreaseValue = 'slider.decrement';
/**
 * @hidden
 */
exports.sliderDragTitle = 'slider.dragTitle';
/**
 * @hidden
 */
exports.colorGradientDragHandle = 'colorGradient.dragHandle';
/**
 * @hidden
 */
exports.colorGradientR = 'colorGradient.r';
/**
 * @hidden
 */
exports.colorGradientG = 'colorGradient.g';
/**
 * @hidden
 */
exports.colorGradientB = 'colorGradient.b';
/**
 * @hidden
 */
exports.colorGradientA = 'colorGradient.a';
/**
 * @hidden
 */
exports.colorGradientHex = 'colorGradient.hex';
/**
 * @hidden
 */
exports.colorGradientHexLabel = 'colorGradient.hexLabel';
/**
 * @hidden
 */
exports.colorGradientInputColorButton = 'colorGradient.inputColorButton';
/**
 * @hidden
 */
exports.colorGradientRLabel = 'colorGradient.rLabel';
/**
 * @hidden
 */
exports.colorGradientGLabel = 'colorGradient.gLabel';
/**
 * @hidden
 */
exports.colorGradientBLabel = 'colorGradient.bLabel';
/**
 * @hidden
 */
exports.colorGradientALabel = 'colorGradient.aLabel';
/**
 * @hidden
 */
exports.colorGradientContrastRatio = 'colorGradient.contrastRatio';
/**
 * @hidden
 */
exports.colorGradientAALevel = 'colorGradient.colorGradientAALevel';
/**
 * @hidden
 */
exports.colorGradientAAALevel = 'colorGradient.colorGradientAAALevel';
/**
 * @hidden
 */
exports.colorGradientPass = 'colorGradient.colorGradientPass';
/**
 * @hidden
 */
exports.colorGradientFail = 'colorGradient.colorGradientFail';
/**
 * @hidden
 */
exports.colorGradientHueSliderLabel = 'colorGradient.hueSliderLabel';
/**
 * @hidden
 */
exports.colorGradientAlphaSliderLabel = 'colorGradient.alphaSliderLabel';
/**
 * @hidden
 */
exports.flatColorPickerCancelBtn = 'flatColorPicker.cancelBtn';
/**
 * @hidden
 */
exports.flatColorPickerApplyBtn = 'flatColorPicker.applyBtn';
/**
 * @hidden
 */
exports.flatColorPickerGradientBtn = 'flatColorPicker.gradientBtn';
/**
 * @hidden
 */
exports.flatColorPickerPaletteBtn = 'flatColorPicker.paletterBtn';
/**
 * @hidden
 */
exports.flatColorPickerClearBtn = 'flatColorPicker.clearBtn';
/**
 * @hidden
 */
exports.checkboxValidation = 'checkbox.validation';
/**
 * @hidden
 */
exports.checkboxOptionalText = 'checkbox.optionalText';
/**
 * @hidden
 */
exports.radioButtonValidation = 'radioButton.validation';
/**
 * @hidden
 */
exports.switchValidation = 'switch.validation';
/**
 * @hidden
 */
exports.colorPickerDropdownButtonAriaLabel = 'colorPicker.dropdownButtonAriaLabel';
/**
 * @hidden
 */
exports.colorGradientToggleInputsButton = 'colorGradient.toggleInputsButton';
/**
 * @hidden
 */
exports.ratingAriaLabel = 'rating.ariaLabel';
/**
 * @hidden
 */
exports.signatureClear = 'signature.clear';
/**
 * @hidden
 */
exports.signatureMaximize = 'signature.maximize';
/**
 * @hidden
 */
exports.signatureMinimize = 'signature.minimize';
/**
 * @hidden
 */
exports.messages = (_a = {},
    _a[exports.numericIncreaseValue] = 'Increase value',
    _a[exports.numericDecreaseValue] = 'Decrease value',
    _a[exports.sliderIncreaseValue] = 'Increase',
    _a[exports.sliderDecreaseValue] = 'Decrease',
    _a[exports.sliderDragTitle] = 'Drag',
    _a[exports.colorGradientDragHandle] = "Color well with two-dimensional slider for selecting\n     saturation and lightness. Selected color is",
    _a[exports.colorGradientInputColorButton] = 'Change the color input',
    _a[exports.colorGradientR] = 'r',
    _a[exports.colorGradientG] = 'g',
    _a[exports.colorGradientB] = 'b',
    _a[exports.colorGradientA] = 'a',
    _a[exports.colorGradientRLabel] = 'red chanel',
    _a[exports.colorGradientGLabel] = 'green chanel',
    _a[exports.colorGradientBLabel] = 'blue chanel',
    _a[exports.colorGradientALabel] = 'alpha chanel',
    _a[exports.colorGradientHex] = 'hex',
    _a[exports.colorGradientHexLabel] = 'HEX',
    _a[exports.colorGradientContrastRatio] = 'Contrast ratio',
    _a[exports.colorGradientAALevel] = 'AA',
    _a[exports.colorGradientAAALevel] = 'AAA',
    _a[exports.colorGradientPass] = 'Pass',
    _a[exports.colorGradientFail] = 'Fail',
    _a[exports.colorGradientHueSliderLabel] = 'Hue slider',
    _a[exports.colorGradientAlphaSliderLabel] = 'Alpha slider',
    _a[exports.colorGradientToggleInputsButton] = 'Toggle colorgradient inputs',
    _a[exports.flatColorPickerCancelBtn] = 'Cancel',
    _a[exports.flatColorPickerApplyBtn] = 'Apply',
    _a[exports.flatColorPickerGradientBtn] = 'Gradient',
    _a[exports.flatColorPickerPaletteBtn] = 'Palette',
    _a[exports.flatColorPickerClearBtn] = 'Clear',
    _a[exports.checkboxValidation] = 'Please check this box if you want to proceed!',
    _a[exports.checkboxOptionalText] = '(Optional)',
    _a[exports.radioButtonValidation] = 'Please select option if you want to proceed!',
    _a[exports.switchValidation] = 'Please turn on if you want to proceed!',
    _a[exports.colorPickerDropdownButtonAriaLabel] = 'Select',
    _a[exports.ratingAriaLabel] = 'Rating',
    _a[exports.signatureClear] = 'Clear',
    _a[exports.signatureMaximize] = 'Maximize',
    _a[exports.signatureMinimize] = 'Minimize',
    _a);
