// @progress/kendo-vue-layout/dist/npm/appbar/AppBar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AppBarVue2 = exports.AppBar = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
var package_metadata_1 = require("../package-metadata");
/**
 * @hidden
 */
var AppBarVue2 = {
  name: 'KendoAppBar',
  props: {
    id: String,
    themeColor: {
      type: String,
      default: 'light'
    },
    position: {
      type: String,
      default: 'top',
      validator: function validator(value) {
        return ['top', 'bottom'].includes(value);
      }
    },
    positionMode: {
      type: String,
      default: 'static',
      validator: function validator(value) {
        return ['static', 'sticky', 'fixed'].includes(value);
      }
    }
  },
  created: function created() {
    (0, kendo_vue_common_1.validatePackage)(package_metadata_1.packageMetadata);
    this.calculatedId = (0, kendo_vue_common_1.guid)();
  },
  computed: {
    wrapperClass: function wrapperClass() {
      var _a = this.$props,
        position = _a.position,
        positionMode = _a.positionMode,
        themeColor = _a.themeColor;
      return {
        'k-appbar': true,
        'k-appbar-top': position === 'top',
        'k-appbar-bottom': position === 'bottom',
        'k-appbar-static': positionMode === 'static',
        'k-appbar-sticky': positionMode === 'sticky',
        'k-appbar-fixed': positionMode === 'fixed',
        'k-appbar-light': themeColor === 'light',
        'k-appbar-dark': themeColor === 'dark',
        'k-appbar-inherit': themeColor === 'inherit'
      };
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlot = (0, kendo_vue_common_1.getDefaultSlots)(this);
    var id = this.$props.id;
    return h("div", {
      "class": this.wrapperClass,
      id: id || this.calculatedId,
      attrs: this.v3 ? undefined : {
        id: id || this.calculatedId
      }
    }, [defaultSlot]);
  },
  methods: {
    focus: function focus() {
      if (this.$el) {
        (0, kendo_vue_common_1.focusFirstFocusableChild)(this.$el);
      }
    }
  }
};
exports.AppBarVue2 = AppBarVue2;
/**
 * @hidden
 */
var AppBar = AppBarVue2;
exports.AppBar = AppBar;
