// @progress/kendo-vue-layout/dist/npm/appbar/AppBarSpacer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AppBarSpacerVue2 = exports.AppBarSpacer = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
/**
 * @hidden
 */
var AppBarSpacerVue2 = {
  name: 'KendoAppBarSpacer',
  props: {
    width: [String, Number]
  },
  computed: {
    spacerClasses: function spacerClasses() {
      return {
        'k-appbar-spacer': true,
        'k-appbar-spacer-sized': this.width !== undefined
      };
    },
    spacerStyles: function spacerStyles() {
      return {
        flexBasis: this.width !== undefined ? this.width : undefined
      };
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlot = (0, kendo_vue_common_1.getDefaultSlots)(this);
    return h("span", {
      "class": this.spacerClasses,
      style: this.spacerStyles
    }, [defaultSlot]);
  },
  methods: {
    focus: function focus() {
      if (this.$el) {
        (0, kendo_vue_common_1.focusFirstFocusableChild)(this.$el);
      }
    }
  }
};
exports.AppBarSpacerVue2 = AppBarSpacerVue2;
/**
 * @hidden
 */
var AppBarSpacer = AppBarSpacerVue2;
exports.AppBarSpacer = AppBarSpacer;
