// @progress/kendo-vue-layout/dist/npm/appbar/AppBarSection.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AppBarSectionVue2 = exports.AppBarSection = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
/**
 * @hidden
 */
var AppBarSectionVue2 = {
  name: 'KendoAppBarSection',
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlot = (0, kendo_vue_common_1.getDefaultSlots)(this);
    return h("div", {
      "class": 'k-appbar-section'
    }, [defaultSlot]);
  },
  methods: {
    focus: function focus() {
      if (this.$el) {
        (0, kendo_vue_common_1.focusFirstFocusableChild)(this.$el);
      }
    }
  }
};
exports.AppBarSectionVue2 = AppBarSectionVue2;
/**
 * @hidden
 */
var AppBarSection = AppBarSectionVue2;
exports.AppBarSection = AppBarSection;
