// @progress/kendo-vue-upload/dist/npm/UploadInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UploadInputVue2 = exports.UploadInput = void 0;
// @ts-ignore
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
var kendo_vue_intl_1 = require("@progress/kendo-vue-intl");
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var ref = allVue.ref;
var main_1 = require("./messages/main");
var CHROME_REGEX = /(chrome)[ \/]([\w.]+)/i;
var SAFARI_REGEX = /(webkit)[ \/]([\w.]+)/i;
/**
 * @hidden
 */
var UploadInputVue2 = {
  name: 'KendoVueUploadInput',
  props: {
    async: Object,
    id: String,
    multiple: {
      type: Boolean,
      default: undefined
    },
    disabled: {
      type: Boolean,
      default: undefined
    },
    accept: {
      type: String,
      default: undefined
    },
    hasFiles: {
      type: Boolean,
      default: false
    },
    ariaLabelledBy: {
      type: String,
      default: undefined
    },
    ariaDescribedBy: {
      type: String,
      default: undefined
    }
  },
  // @ts-ignore
  emits: {
    'mousedown': null,
    'add': null
  },
  inject: {
    kendoLocalizationService: {
      default: null
    }
  },
  mounted: function mounted() {
    this._input = this.v3 ? this.inputRef : this.$refs.input;
  },
  methods: {
    onMouseDown: function onMouseDown(e) {
      this.$emit('mousedown', e);
    },
    onAdd: function onAdd() {
      var ua = navigator.userAgent;
      var input = this._input;
      if (input) {
        if (input.files) {
          this.$emit('add', input.files);
        }
        /*
            Chrome and Internet Explorer do not trigger a `change` event
            when a file with the same name is selected a number of consecutive times.
            As a workaround, clear the input value after handling the file.
        */
        if (!(!ua.match(CHROME_REGEX) && ua.match(SAFARI_REGEX))) {
          input.type = '';
          input.type = 'file';
        }
      }
    },
    actionElement: function actionElement() {
      return this._input;
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    var inputRef = ref(null);
    return {
      v3: v3,
      inputRef: inputRef
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var _a = this.$props,
      multiple = _a.multiple,
      async = _a.async,
      disabled = _a.disabled,
      accept = _a.accept,
      hasFiles = _a.hasFiles,
      ariaLabelledBy = _a.ariaLabelledBy,
      ariaDescribedBy = _a.ariaDescribedBy;
    var localizationService = (0, kendo_vue_intl_1.provideLocalizationService)(this);
    var message = hasFiles ? main_1.selectTitle : main_1.selectNoFilesTitle;
    var selectMessage = localizationService.toLanguageString(message, main_1.messages[message]);
    var selectLabel = localizationService.toLanguageString(main_1.select, main_1.messages[main_1.select]);
    return h("input", {
      ref: (0, kendo_vue_common_1.setRef)(this, 'input'),
      id: this.id,
      attrs: this.v3 ? undefined : {
        id: this.id,
        autocomplete: 'off',
        name: async.saveField,
        accept: accept,
        type: 'file',
        tabindex: -1,
        multiple: multiple,
        disabled: disabled,
        title: selectMessage,
        "aria-label": ariaLabelledBy !== undefined ? undefined : selectLabel,
        "aria-labelledby": ariaLabelledBy,
        "aria-describedby": ariaDescribedBy
      },
      "class": 'k-hidden',
      autocomplete: 'off',
      name: async.saveField,
      accept: accept,
      type: 'file',
      tabindex: -1,
      multiple: multiple,
      disabled: disabled,
      onChange: this.onAdd,
      on: this.v3 ? undefined : {
        "change": this.onAdd,
        "mousedown": this.onMouseDown
      },
      onMousedown: this.onMouseDown,
      title: selectMessage,
      "aria-label": ariaLabelledBy !== undefined ? undefined : selectLabel,
      "aria-labelledby": ariaLabelledBy,
      "aria-describedby": ariaDescribedBy
    });
  }
};
exports.UploadInputVue2 = UploadInputVue2;
/**
 * @hidden
 */
var UploadInput = UploadInputVue2;
exports.UploadInput = UploadInput;
