// @progress/kendo-vue-data-tools/dist/npm/messages/main.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.columnMenuFilterIsEmptyOperator = exports.columnMenuFilterIsNotNullOperator = exports.columnMenuFilterIsNullOperator = exports.columnMenuFilterNotEqOperator = exports.columnMenuFilterEqOperator = exports.columnMenuSortDescending = exports.columnMenuSortAscending = exports.columnMenuFilterTitle = exports.columnMenuFilterSubmitButton = exports.columnMenuFilterClearButton = exports.columnMenuFilterChooseOperator = exports.filterClose = exports.filterAddGroup = exports.filterAddExpression = exports.filterOrLogic = exports.filterAndLogic = exports.filterBeforeOrEqualOperator = exports.filterBeforeOperator = exports.filterAfterOperator = exports.filterAfterOrEqualOperator = exports.filterIsFalse = exports.filterIsTrue = exports.filterLtOperator = exports.filterLteOperator = exports.filterGtOperator = exports.filterGteOperator = exports.filterEndsWithOperator = exports.filterNotContainsOperator = exports.filterContainsOperator = exports.filterStartsWithOperator = exports.filterIsNotEmptyOperator = exports.filterIsEmptyOperator = exports.filterIsNotNullOperator = exports.filterIsNullOperator = exports.filterNotEqOperator = exports.filterGroupFilterAriaLabel = exports.filterEqOperator = exports.pagerTotalPages = exports.pagerOf = exports.pagerPage = exports.pagerPageSelection = exports.pagerSmallPageSelectorAriaLabel = exports.pagerPageInputAriaLabel = exports.pagerItemPerPage = exports.pagerLastPage = exports.pagerNextPage = exports.pagerPreviousPage = exports.pagerFirstPage = exports.pagerInfo = exports.columnMenu = void 0;
exports.messages = exports.filterDropDownOperatorAriaLabel = exports.filterDateFilterAriaLabel = exports.filterTextFilterAriaLabel = exports.filterNumericFilterAriaLabel = exports.filterEnumFilterDropdownAriaLabel = exports.filterExpressionOperatorDropdownAriaLabel = exports.filterExpressionDropdownAriaLabel = exports.filterExpressionAriaLabel = exports.filterGroupAriaLabel = exports.filterAriaLabel = exports.sortAriaLabel = exports.filterCheckAll = exports.selectAllRows = exports.searchPlaceholder = exports.columnMenuFilterOrLogic = exports.columnMenuFilterAndLogic = exports.columnMenuFilterBeforeOrEqualOperator = exports.columnMenuFilterBeforeOperator = exports.columnMenuFilterAfterOperator = exports.columnMenuFilterAfterOrEqualOperator = exports.columnMenuFilterIsTrue = exports.columnMenuFilterLtOperator = exports.columnMenuFilterLteOperator = exports.columnMenuFilterGtOperator = exports.columnMenuFilterGteOperator = exports.columnMenuFilterEndsWithOperator = exports.columnMenuFilterNotContainsOperator = exports.columnMenuFilterContainsOperator = exports.columnMenuFilterStartsWithOperator = exports.columnMenuFilterIsNotEmptyOperator = void 0;
/**
 * @hidden
 */
exports.columnMenu = 'grid.columnMenu';
/**
 * @hidden
 */
exports.pagerInfo = 'pager.info';
/**
 * @hidden
 */
exports.pagerFirstPage = 'pager.firstPage';
/**
 * @hidden
 */
exports.pagerPreviousPage = 'pager.previousPage';
/**
 * @hidden
 */
exports.pagerNextPage = 'pager.nextPage';
/**
 * @hidden
 */
exports.pagerLastPage = 'pager.lastPage';
/**
 * @hidden
 */
exports.pagerItemPerPage = 'pager.itemsPerPage';
/**
 * @hidden
 */
exports.pagerPageInputAriaLabel = 'pager.pageInputAriaLabel';
/**
 * @hidden
 */
exports.pagerSmallPageSelectorAriaLabel = 'pager.smallPageSelectorAriaLabel';
/**
 * @hidden
 */
exports.pagerPageSelection = 'pager.pageSelection';
/**
 * @hidden
 */
exports.pagerPage = 'pager.page';
/**
 * @hidden
 */
exports.pagerOf = 'pager.of';
/**
 * @hidden
 */
exports.pagerTotalPages = 'pager.totalPages';
/**
 * @hidden
 */
exports.filterEqOperator = 'filter.eqOperator';
/**
 * @hidden
 */
exports.filterGroupFilterAriaLabel = 'filter.groupFilterAriaLabel';
/**
 * @hidden
 */
exports.filterNotEqOperator = 'filter.notEqOperator';
/**
 * @hidden
 */
exports.filterIsNullOperator = 'filter.isNullOperator';
/**
 * @hidden
 */
exports.filterIsNotNullOperator = 'filter.isNotNullOperator';
/**
 * @hidden
 */
exports.filterIsEmptyOperator = 'filter.isEmptyOperator';
/**
 * @hidden
 */
exports.filterIsNotEmptyOperator = 'filter.isNotEmptyOperator';
/**
 * @hidden
 */
exports.filterStartsWithOperator = 'filter.startsWithOperator';
/**
 * @hidden
 */
exports.filterContainsOperator = 'filter.containsOperator';
/**
 * @hidden
 */
exports.filterNotContainsOperator = 'filter.notContainsOperator';
/**
 * @hidden
 */
exports.filterEndsWithOperator = 'filter.endsWithOperator';
/**
 * @hidden
 */
exports.filterGteOperator = 'filter.gteOperator';
/**
 * @hidden
 */
exports.filterGtOperator = 'filter.gtOperator';
/**
 * @hidden
 */
exports.filterLteOperator = 'filter.lteOperator';
/**
 * @hidden
 */
exports.filterLtOperator = 'filter.ltOperator';
/**
 * @hidden
 */
exports.filterIsTrue = 'filter.isTrue';
/**
 * @hidden
 */
exports.filterIsFalse = 'filter.isFalse';
/**
 * @hidden
 */
exports.filterAfterOrEqualOperator = 'filter.afterOrEqualOperator';
/**
 * @hidden
 */
exports.filterAfterOperator = 'filter.afterOperator';
/**
 * @hidden
 */
exports.filterBeforeOperator = 'filter.beforeOperator';
/**
 * @hidden
 */
exports.filterBeforeOrEqualOperator = 'filter.beforeOrEqualOperator';
/**
 * @hidden
 */
exports.filterAndLogic = 'filter.andLogic';
/**
 * @hidden
 */
exports.filterOrLogic = 'filter.orLogic';
/**
 * @hidden
 */
exports.filterAddExpression = 'filter.addExpression';
/**
 * @hidden
 */
exports.filterAddGroup = 'filter.addGroup';
/**
 * @hidden
 */
exports.filterClose = 'filter.close';
/**
 * @hidden
 */
exports.columnMenuFilterChooseOperator = 'columnMenu.filterChooseOperator';
/**
 * @hidden
 */
exports.columnMenuFilterClearButton = 'columnMenu.filterClearButton';
/**
 * @hidden
 */
exports.columnMenuFilterSubmitButton = 'columnMenu.filterSubmitButton';
/**
 * @hidden
 */
exports.columnMenuFilterTitle = 'columnMenu.filterTitle';
/**
 * @hidden
 */
exports.columnMenuSortAscending = 'columnMenu.sortAscending';
/**
 * @hidden
 */
exports.columnMenuSortDescending = 'columnMenu.sortDescending';
/**
 * @hidden
 */
exports.columnMenuFilterEqOperator = 'columnMenu.filterEqOperator';
/**
 * @hidden
 */
exports.columnMenuFilterNotEqOperator = 'columnMenu.filterNotEqOperator';
/**
 * @hidden
 */
exports.columnMenuFilterIsNullOperator = 'columnMenu.filterIsNullOperator';
/**
 * @hidden
 */
exports.columnMenuFilterIsNotNullOperator = 'columnMenu.filterIsNotNullOperator';
/**
 * @hidden
 */
exports.columnMenuFilterIsEmptyOperator = 'columnMenu.filterIsEmptyOperator';
/**
 * @hidden
 */
exports.columnMenuFilterIsNotEmptyOperator = 'columnMenu.filterIsNotEmptyOperator';
/**
 * @hidden
 */
exports.columnMenuFilterStartsWithOperator = 'columnMenu.filterStartsWithOperator';
/**
 * @hidden
 */
exports.columnMenuFilterContainsOperator = 'columnMenu.filterContainsOperator';
/**
 * @hidden
 */
exports.columnMenuFilterNotContainsOperator = 'columnMenu.filterNotContainsOperator';
/**
 * @hidden
 */
exports.columnMenuFilterEndsWithOperator = 'columnMenu.filterEndsWithOperator';
/**
 * @hidden
 */
exports.columnMenuFilterGteOperator = 'columnMenu.filterGteOperator';
/**
 * @hidden
 */
exports.columnMenuFilterGtOperator = 'columnMenu.filterGtOperator';
/**
 * @hidden
 */
exports.columnMenuFilterLteOperator = 'columnMenu.filterLteOperator';
/**
 * @hidden
 */
exports.columnMenuFilterLtOperator = 'columnMenu.filterLtOperator';
/**
 * @hidden
 */
exports.columnMenuFilterIsTrue = 'columnMenu.filterIsTrue';
/**
 * @hidden
 */
exports.columnMenuFilterAfterOrEqualOperator = 'columnMenu.filterAfterOrEqualOperator';
/**
 * @hidden
 */
exports.columnMenuFilterAfterOperator = 'columnMenu.filterAfterOperator';
/**
 * @hidden
 */
exports.columnMenuFilterBeforeOperator = 'columnMenu.filterBeforeOperator';
/**
 * @hidden
 */
exports.columnMenuFilterBeforeOrEqualOperator = 'columnMenu.filterBeforeOrEqualOperator';
/**
 * @hidden
 */
exports.columnMenuFilterAndLogic = 'columnMenu.filterAndLogic';
/**
 * @hidden
 */
exports.columnMenuFilterOrLogic = 'columnMenu.filterOrLogic';
/**
 * @hidden
 */
exports.searchPlaceholder = 'grid.searchPlaceholder';
/**
 * @hidden
 */
exports.selectAllRows = 'grid.selectAllRows';
/**
 * @hidden
 */
exports.filterCheckAll = 'grid.filterCheckAll';
/**
 * @hidden
 */
exports.sortAriaLabel = 'sort.ariaLabel';
/**
 * @hidden
 */
exports.filterAriaLabel = 'filter.ariaLabel';
/**
 * @hidden
 */
exports.filterGroupAriaLabel = 'filter.groupAriaLabel';
/**
 * @hidden
 */
exports.filterExpressionAriaLabel = 'filter.expressionAriaLabel';
/**
 * @hidden
 */
exports.filterExpressionDropdownAriaLabel = 'filter.expressionDropdownAriaLabel';
/**
 * @hidden
 */
exports.filterExpressionOperatorDropdownAriaLabel = 'filter.expressionOperatorDropdownAriaLabel';
/**
 * @hidden
 */
exports.filterEnumFilterDropdownAriaLabel = 'filter.enumFilterDropdownAriaLabel';
/**
 * @hidden
 */
exports.filterNumericFilterAriaLabel = 'filter.numericFilterAriaLabel';
/**
 * @hidden
 */
exports.filterTextFilterAriaLabel = 'filter.textFilterAriaLabel';
/**
 * @hidden
 */
exports.filterDateFilterAriaLabel = 'filter.dateFilterAriaLabel';
/**
 * @hidden
 */
exports.filterDropDownOperatorAriaLabel = 'filter.dropdownOperatorAriaLabel';
/**
 * @hidden
 */
exports.messages = (_a = {},
    _a[exports.filterGroupFilterAriaLabel] = 'Group Filter',
    _a[exports.pagerSmallPageSelectorAriaLabel] = 'Select page',
    _a[exports.pagerPageInputAriaLabel] = 'Pager - Page number input',
    _a[exports.pagerItemPerPage] = 'items per page',
    _a[exports.pagerPageSelection] = 'Select page',
    _a[exports.pagerInfo] = '{0} - {1} of {2} items',
    _a[exports.pagerFirstPage] = 'Go to the first page',
    _a[exports.pagerPreviousPage] = 'Go to the previous page',
    _a[exports.pagerNextPage] = 'Go to the next page',
    _a[exports.pagerLastPage] = 'Go to the last page',
    _a[exports.pagerPage] = 'Page',
    _a[exports.pagerOf] = 'of',
    _a[exports.pagerTotalPages] = '{0}',
    _a[exports.searchPlaceholder] = 'Search',
    _a[exports.selectAllRows] = 'Select All Rows',
    _a[exports.filterCheckAll] = 'Check All',
    _a[exports.sortAriaLabel] = 'Sortable',
    _a[exports.filterAriaLabel] = 'Filter',
    _a[exports.filterEqOperator] = 'Is equal to',
    _a[exports.filterNotEqOperator] = 'Is not equal to',
    _a[exports.filterIsNullOperator] = 'Is null',
    _a[exports.filterIsNotNullOperator] = 'Is not null',
    _a[exports.filterIsEmptyOperator] = 'Is empty',
    _a[exports.filterIsNotEmptyOperator] = 'Is not empty',
    _a[exports.filterStartsWithOperator] = 'Starts with',
    _a[exports.filterContainsOperator] = 'Contains',
    _a[exports.filterNotContainsOperator] = 'Does not contain',
    _a[exports.filterEndsWithOperator] = 'Ends with',
    _a[exports.filterGteOperator] = 'Is greater than or equal to',
    _a[exports.filterGtOperator] = 'Is greater than',
    _a[exports.filterLteOperator] = 'Is less than or equal to',
    _a[exports.filterLtOperator] = 'Is less than',
    _a[exports.filterIsTrue] = 'Is true',
    _a[exports.filterIsFalse] = 'Is false',
    _a[exports.filterAfterOrEqualOperator] = 'Is after or equal to',
    _a[exports.filterAfterOperator] = 'Is after',
    _a[exports.filterBeforeOperator] = 'Is before',
    _a[exports.filterBeforeOrEqualOperator] = 'Is before or equal to',
    _a[exports.filterAndLogic] = 'And',
    _a[exports.filterOrLogic] = 'Or',
    _a[exports.filterAddExpression] = 'Add Expression',
    _a[exports.filterAddGroup] = 'Add Group',
    _a[exports.filterClose] = 'Close',
    _a[exports.filterGroupAriaLabel] = 'Filter toolbar',
    _a[exports.filterExpressionAriaLabel] = 'Filter expression row',
    _a[exports.filterExpressionDropdownAriaLabel] = 'Filter dropdown',
    _a[exports.filterExpressionOperatorDropdownAriaLabel] = 'Filter operator dropdown',
    _a[exports.filterEnumFilterDropdownAriaLabel] = 'Select True/False',
    _a[exports.filterNumericFilterAriaLabel] = 'Enter number',
    _a[exports.filterTextFilterAriaLabel] = 'Enter text',
    _a[exports.filterDateFilterAriaLabel] = 'Enter date',
    _a[exports.filterDropDownOperatorAriaLabel] = 'Choose Operator',
    _a[exports.columnMenuFilterChooseOperator] = 'Choose Operator',
    _a[exports.columnMenuFilterClearButton] = 'Clear',
    _a[exports.columnMenuFilterSubmitButton] = 'Filter',
    _a[exports.columnMenuFilterTitle] = 'Filter',
    _a[exports.columnMenuSortAscending] = 'Sort Ascending',
    _a[exports.columnMenuSortDescending] = 'Sort Descending',
    _a[exports.columnMenuFilterEqOperator] = 'Is equal to',
    _a[exports.columnMenuFilterNotEqOperator] = 'Is not equal to',
    _a[exports.columnMenuFilterIsNullOperator] = 'Is null',
    _a[exports.columnMenuFilterIsNotNullOperator] = 'Is not null',
    _a[exports.columnMenuFilterIsEmptyOperator] = 'Is empty',
    _a[exports.columnMenuFilterIsNotEmptyOperator] = 'Is not empty',
    _a[exports.columnMenuFilterStartsWithOperator] = 'Starts with',
    _a[exports.columnMenuFilterContainsOperator] = 'Contains',
    _a[exports.columnMenuFilterNotContainsOperator] = 'Does not contain',
    _a[exports.columnMenuFilterEndsWithOperator] = 'Ends with',
    _a[exports.columnMenuFilterGteOperator] = 'Is greater than or equal to',
    _a[exports.columnMenuFilterGtOperator] = 'Is greater than',
    _a[exports.columnMenuFilterLteOperator] = 'Is less than or equal to',
    _a[exports.columnMenuFilterLtOperator] = 'Is less than',
    _a[exports.columnMenuFilterIsTrue] = 'Is true',
    _a[exports.columnMenuFilterAfterOrEqualOperator] = 'Is after or equal to',
    _a[exports.columnMenuFilterAfterOperator] = 'Is after',
    _a[exports.columnMenuFilterBeforeOperator] = 'Is before',
    _a[exports.columnMenuFilterBeforeOrEqualOperator] = 'Is before or equal to',
    _a[exports.columnMenuFilterAndLogic] = 'And',
    _a[exports.columnMenuFilterOrLogic] = 'Or',
    _a['grid.filterEqOperator'] = 'Is equal to',
    _a['grid.filterNotEqOperator'] = 'Is not equal to',
    _a['grid.filterIsNullOperator'] = 'Is null',
    _a['grid.filterIsNotNullOperator'] = 'Is not null',
    _a['grid.filterIsEmptyOperator'] = 'Is empty',
    _a['grid.filterIsNotEmptyOperator'] = 'Is not empty',
    _a['grid.filterStartsWithOperator'] = 'Starts with',
    _a['grid.filterContainsOperator'] = 'Contains',
    _a['grid.filterNotContainsOperator'] = 'Does not contain',
    _a['grid.filterEndsWithOperator'] = 'Ends with',
    _a['grid.filterGteOperator'] = 'Is greater than or equal to',
    _a['grid.filterGtOperator'] = 'Is greater than',
    _a['grid.filterLteOperator'] = 'Is less than or equal to',
    _a['grid.filterLtOperator'] = 'Is less than',
    _a['grid.filterIsTrue'] = 'Is true',
    _a['grid.filterIsFalse'] = 'Is false',
    _a['grid.filterBooleanAll'] = '(All)',
    _a['grid.filterAfterOrEqualOperator'] = 'Is after or equal to',
    _a['grid.filterAfterOperator'] = 'Is after',
    _a['grid.filterBeforeOperator'] = 'Is before',
    _a['grid.filterBeforeOrEqualOperator'] = 'Is before or equal to',
    _a['grid.filterAndLogic'] = 'And',
    _a['grid.filterOrLogic'] = 'Or',
    _a);
