// @progress/kendo-vue-data-tools/dist/esm/filter/filters/NumericFilter.js
var __assign = this && this.__assign || function () {
  __assign = Object.assign || function (t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
      s = arguments[i];
      for (var p in s) {
        if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
      }
    }
    return t;
  };
  return __assign.apply(this, arguments);
};
// @ts-ignore
import * as Vue from 'vue';
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
import { NumericTextBox } from '@progress/kendo-vue-inputs';
import { provideLocalizationService } from '@progress/kendo-vue-intl';
import { filterNumericFilterAriaLabel, messages } from '../../messages/main.js';
/**
 * @hidden
 */
var NumericFilterVue2 = {
  name: 'kendoNumericFilter',
  props: {
    filter: {
      type: Object,
      required: true
    }
  },
  inject: {
    kendoLocalizationService: {
      default: null
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var filter = this.$props.filter;
    var ls = provideLocalizationService(this);
    var ariaLabel = ls.toLanguageString(filterNumericFilterAriaLabel, messages[filterNumericFilterAriaLabel]);
    return (
      // @ts-ignore
      h(NumericTextBox, {
        value: typeof filter.value === 'number' ? filter.value : null,
        attrs: this.v3 ? undefined : {
          value: typeof filter.value === 'number' ? filter.value : null,
          ariaLabel: ariaLabel
        },
        onChange: this.onChange,
        on: this.v3 ? undefined : {
          "change": this.onChange
        },
        ariaLabel: ariaLabel
      })
    );
  },
  methods: {
    onChange: function onChange(event) {
      this.$emit('filterchange', {
        nextFilter: __assign(__assign({}, this.$props.filter), {
          value: event.value
        })
      });
    }
  }
};
/**
 * @hidden
 */
var NumericFilter = NumericFilterVue2;
export { NumericFilter, NumericFilterVue2 };
