// @progress/kendo-vue-data-tools/dist/npm/header/HeaderTdElement.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HeaderTdElementVue2 = exports.HeaderTdElement = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var inject = allVue.inject;
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
var constants_1 = require("../navigation/constants");
/**
 * @hidden
 */
var HeaderTdElementVue2 = {
  name: 'KendoHeaderTdElement',
  // @ts-ignore
  emits: {
    keydown: null
  },
  props: {
    ariaColumnIndex: Number,
    role: String,
    columnId: String,
    navigatable: Boolean
  },
  computed: {
    tdClass: function tdClass() {
      var _a;
      var className = this.$props.className;
      return _a = {
        'k-table-td': true
      }, _a[className] = className, _a;
    }
  },
  inject: {
    getKeyboardNavigationAttributes: {
      default: kendo_vue_common_1.noop
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    var kendoIntlService = inject('kendoIntlService', {});
    var kendoLocalizationService = inject('kendoLocalizationService', {});
    return {
      v3: v3,
      kendoIntlService: kendoIntlService,
      kendoLocalizationService: kendoLocalizationService
    };
  },
  // @ts-ignore
  render: function render(createElement) {
    var h = gh || createElement;
    var _a = this.$props,
      columnId = _a.columnId,
      navigatable = _a.navigatable,
      ariaColumnIndex = _a.ariaColumnIndex,
      role = _a.role;
    var navAttrs = this.getKeyboardNavigationAttributes(columnId, navigatable);
    var defaultSlot = (0, kendo_vue_common_1.getDefaultSlots)(this);
    return h("td", {
      style: {
        top: '0px'
      },
      "aria-colindex": ariaColumnIndex,
      attrs: this.v3 ? undefined : {
        "aria-colindex": ariaColumnIndex,
        role: role,
        tabindex: navAttrs.tabIndex,
        "data-keyboardnavlevel": navAttrs[constants_1.KEYBOARD_NAV_DATA_LEVEL],
        "data-keyboardnavid": navAttrs[constants_1.KEYBOARD_NAV_DATA_ID]
      },
      "class": this.tdClass,
      onKeydown: this.onKeyDown,
      on: this.v3 ? undefined : {
        "keydown": this.onKeyDown
      },
      role: role,
      tabindex: navAttrs.tabIndex,
      "data-keyboardnavlevel": navAttrs[constants_1.KEYBOARD_NAV_DATA_LEVEL],
      "data-keyboardnavid": navAttrs[constants_1.KEYBOARD_NAV_DATA_ID]
    }, [defaultSlot]);
  },
  methods: {
    onKeyDown: function onKeyDown(e) {
      this.$emit('keydown', e);
    }
  }
};
exports.HeaderTdElementVue2 = HeaderTdElementVue2;
/**
 * @hidden
 */
var HeaderTdElement = HeaderTdElementVue2;
exports.HeaderTdElement = HeaderTdElement;
