// @progress/kendo-vue-grid/dist/npm/cells/GridDetailCell.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GridDetailCellVue2 = exports.GridDetailCell = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var inject = allVue.inject;
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
var kendo_vue_data_tools_1 = require("@progress/kendo-vue-data-tools");
/**
 * @hidden
 */
var GridDetailCellVue2 = {
  props: {
    colSpan: Number,
    ariaColIndex: Number,
    dataItem: [Object, String, Number],
    dataIndex: Number,
    detail: [String, Function, Object],
    id: String
  },
  inject: {
    getKeyboardNavigationAttributes: {
      default: kendo_vue_common_1.noop
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    var kendoIntlService = inject('kendoIntlService', {});
    return {
      v3: v3,
      kendoIntlService: kendoIntlService
    };
  },
  // @ts-ignore
  render: function render(createElement) {
    var h = gh || createElement;
    var _a = this.$props,
      colSpan = _a.colSpan,
      ariaColIndex = _a.ariaColIndex,
      dataItem = _a.dataItem,
      dataIndex = _a.dataIndex,
      id = _a.id;
    var navAttrs = this.getKeyboardNavigationAttributes(id);
    var detailRender = function detailRender(args) {
      return kendo_vue_common_1.getTemplate.call(this, {
        h: h,
        template: this.$props.detail,
        additionalProps: args
      });
    };
    return h("td", {
      "class": "k-table-td k-detail-cell",
      colspan: colSpan,
      attrs: this.v3 ? undefined : {
        colspan: colSpan,
        "aria-colindex": ariaColIndex,
        role: 'gridcell',
        tabindex: navAttrs.tabIndex,
        "data-keyboardnavlevel": navAttrs[kendo_vue_data_tools_1.KEYBOARD_NAV_DATA_LEVEL],
        "data-keyboardnavid": navAttrs[kendo_vue_data_tools_1.KEYBOARD_NAV_DATA_ID]
      },
      "aria-colindex": ariaColIndex,
      role: 'gridcell',
      tabindex: navAttrs.tabIndex,
      "data-keyboardnavlevel": navAttrs[kendo_vue_data_tools_1.KEYBOARD_NAV_DATA_LEVEL],
      "data-keyboardnavid": navAttrs[kendo_vue_data_tools_1.KEYBOARD_NAV_DATA_ID]
    }, [detailRender.call(this, {
      dataItem: dataItem,
      dataIndex: dataIndex
    })]);
  }
};
exports.GridDetailCellVue2 = GridDetailCellVue2;
/**
 * @hidden
 */
var GridDetailCell = GridDetailCellVue2;
exports.GridDetailCell = GridDetailCell;
