// @progress/kendo-vue-grid/dist/npm/cells/GridSelectionCell.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GridSelectionCellVue2 = exports.GridSelectionCell = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var main_1 = require("../utils/main");
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
var kendo_vue_data_tools_1 = require("@progress/kendo-vue-data-tools");
/**
 * @hidden
 */
var GridSelectionCellVue2 = {
  name: 'KendoGridSelectionCell',
  inheritAttrs: false,
  props: {
    id: String,
    field: String,
    dataItem: Object,
    format: String,
    type: String,
    className: String,
    colSpan: Number,
    columnIndex: Number,
    columnsCount: Number,
    rowType: String,
    level: Number,
    expanded: Boolean,
    render: [String, Function, Object],
    isSelected: Boolean,
    ariaColumnIndex: Number,
    editor: String,
    isRtl: Boolean
  },
  // @ts-ignore
  emits: {
    selectionchange: null,
    cellkeydown: null
  },
  inject: {
    kendoIntlService: {
      default: null
    },
    getKeyboardNavigationAttributes: {
      default: kendo_vue_common_1.noop
    }
  },
  data: function data() {
    return {
      inputId: (0, kendo_vue_common_1.guid)()
    };
  },
  computed: {
    tdClass: function tdClass() {
      var _a;
      var className = this.$props.className;
      return _a = {
        'k-table-td': true
      }, _a[className] = className, _a;
    }
  },
  methods: {
    triggerKeydown: function triggerKeydown(e) {
      this.$emit('cellkeydown', {
        event: e,
        dataItem: this.$props.dataItem,
        field: this.$props.field
      });
    },
    handleOnChange: function handleOnChange(event) {
      this.$emit('selectionchange', {
        event: event,
        dataItem: this.$props.dataItem
      });
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  // @ts-ignore
  render: function render(createElement) {
    var h = gh || createElement;
    var data = (0, main_1.getNestedValue)(this.$props.field, this.$props.dataItem);
    var renderTemplate = this.$props.render;
    var navAttrs = this.getKeyboardNavigationAttributes(this.$props.id);
    var defaultRendering = this.$props.rowType !== 'groupHeader' ? h("td", {
      style: this.$attrs.style,
      onKeydown: this.triggerKeydown,
      on: this.v3 ? undefined : {
        "keydown": this.triggerKeydown
      },
      colspan: this.$props.colSpan,
      attrs: this.v3 ? undefined : {
        colspan: this.$props.colSpan,
        "aria-colindex": this.$props.ariaColumnIndex,
        role: 'gridcell',
        tabindex: navAttrs.tabIndex,
        "data-keyboardnavlevel": navAttrs[kendo_vue_data_tools_1.KEYBOARD_NAV_DATA_LEVEL],
        "data-keyboardnavid": navAttrs[kendo_vue_data_tools_1.KEYBOARD_NAV_DATA_ID]
      },
      "class": this.tdClass,
      "aria-colindex": this.$props.ariaColumnIndex,
      role: 'gridcell',
      tabindex: navAttrs.tabIndex,
      "data-keyboardnavlevel": navAttrs[kendo_vue_data_tools_1.KEYBOARD_NAV_DATA_LEVEL],
      "data-keyboardnavid": navAttrs[kendo_vue_data_tools_1.KEYBOARD_NAV_DATA_ID]
    }, [h("input", {
      checked: this.v3 ? data : null,
      domProps: this.v3 ? undefined : {
        "checked": data
      },
      id: this.inputId,
      attrs: this.v3 ? undefined : {
        id: this.inputId,
        type: "checkbox"
      },
      type: "checkbox",
      "class": "k-checkbox k-checkbox-md k-rounded-md",
      onChange: this.handleOnChange,
      on: this.v3 ? undefined : {
        "change": this.handleOnChange
      }
    }), h("label", {
      "class": "k-checkbox-label",
      "for": this.inputId,
      attrs: this.v3 ? undefined : {
        "for": this.inputId
      }
    })]) : null;
    return kendo_vue_common_1.getTemplate.call(this, {
      h: h,
      template: renderTemplate,
      defaultRendering: defaultRendering,
      additionalProps: this.$props,
      additionalListeners: {
        keydown: this.triggerKeydown,
        change: this.handleOnChange
      }
    });
  }
};
exports.GridSelectionCellVue2 = GridSelectionCellVue2;
/**
 * @hidden
 */
var GridSelectionCell = GridSelectionCellVue2;
exports.GridSelectionCell = GridSelectionCell;
