// @progress/kendo-vue-grid/dist/npm/cells/GridDetailHierarchyCell.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GridDetailHierarchyCellVue2 = exports.GridDetailHierarchyCell = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var inject = allVue.inject;
var kendo_vue_data_tools_1 = require("@progress/kendo-vue-data-tools");
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
/**
 * @hidden
 */
var GridDetailHierarchyCellVue2 = {
  props: {
    id: String
  },
  inject: {
    getKeyboardNavigationAttributes: {
      default: kendo_vue_common_1.noop
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    var kendoIntlService = inject('kendoIntlService', {});
    return {
      v3: v3,
      kendoIntlService: kendoIntlService
    };
  },
  // @ts-ignore
  render: function render(createElement) {
    var h = gh || createElement;
    var navAttrs = this.getKeyboardNavigationAttributes(this.$props.id);
    return h("td", {
      "class": "k-table-td k-hierarchy-cell",
      tabindex: navAttrs.tabIndex,
      attrs: this.v3 ? undefined : {
        tabindex: navAttrs.tabIndex,
        "data-keyboardnavlevel": navAttrs[kendo_vue_data_tools_1.KEYBOARD_NAV_DATA_LEVEL],
        "data-keyboardnavid": navAttrs[kendo_vue_data_tools_1.KEYBOARD_NAV_DATA_ID]
      },
      "data-keyboardnavlevel": navAttrs[kendo_vue_data_tools_1.KEYBOARD_NAV_DATA_LEVEL],
      "data-keyboardnavid": navAttrs[kendo_vue_data_tools_1.KEYBOARD_NAV_DATA_ID]
    });
  }
};
exports.GridDetailHierarchyCellVue2 = GridDetailHierarchyCellVue2;
/**
 * @hidden
 */
var GridDetailHierarchyCell = GridDetailHierarchyCellVue2;
exports.GridDetailHierarchyCell = GridDetailHierarchyCell;
