// @progress/kendo-vue-data-tools/dist/npm/pager/PagerNavigationButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PagerNavigationButtonVue2 = exports.PagerNavigationButton = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var kendo_vue_buttons_1 = require("@progress/kendo-vue-buttons");
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
/**
 * @hidden
 */
var PagerNavigationButtonVue2 = {
  name: 'KendoPagerNavigationButton',
  props: {
    title: String,
    disabled: Boolean,
    icon: String,
    svgIcon: Object,
    page: Number,
    size: String
  },
  inject: {
    kendoLocalizationService: {
      default: null
    },
    kendoIntlService: {
      default: null
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  methods: {
    changePage: function changePage(e) {
      e.preventDefault();
      this.$emit('pagechange', this.$props.page, e);
    }
  },
  // @ts-ignore
  render: function render(createElement) {
    var h = gh || createElement;
    return h(kendo_vue_buttons_1.Button, {
      fillMode: 'flat',
      attrs: this.v3 ? undefined : {
        fillMode: 'flat',
        themeColor: 'base',
        size: this.$props.size,
        rounded: null,
        ariaLabel: this.title,
        icon: (0, kendo_vue_common_1.getIconName)(this.icon),
        svgIcon: this.svgIcon,
        title: this.title,
        disabled: this.disabled,
        role: "button"
      },
      themeColor: 'base',
      size: this.$props.size,
      rounded: null,
      "class": 'k-pager-nav',
      ariaLabel: this.title,
      icon: (0, kendo_vue_common_1.getIconName)(this.icon),
      svgIcon: this.svgIcon,
      title: this.title,
      disabled: this.disabled,
      role: "button",
      onClick: this.changePage,
      on: this.v3 ? undefined : {
        "click": this.changePage
      }
    });
  }
};
exports.PagerNavigationButtonVue2 = PagerNavigationButtonVue2;
/**
 * @hidden
 */
var PagerNavigationButton = PagerNavigationButtonVue2;
exports.PagerNavigationButton = PagerNavigationButton;
