// @progress/kendo-vue-data-tools/dist/npm/pager/PagerPageSizes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PagerPageSizesVue2 = exports.PagerPageSizes = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var inject = allVue.inject;
var main_1 = require("../messages/main");
var kendo_vue_intl_1 = require("@progress/kendo-vue-intl");
var kendo_vue_dropdowns_1 = require("@progress/kendo-vue-dropdowns");
/**
 * @hidden
 */
var PagerPageSizesVue2 = {
  name: 'KendoPagerPageSizes',
  props: {
    value: [String, Number],
    pageSize: Number,
    pageSizes: Array,
    size: String,
    messagesMap: Function
  },
  inject: {
    kendoLocalizationService: {
      default: null
    }
  },
  methods: {
    pageSizeChange: function pageSizeChange(e) {
      this.$emit('pagechange', {
        skip: 0,
        take: parseInt(e.target.value, 10)
      }, e);
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    var kendoLocalizationService = inject('kendoLocalizationService', {});
    return {
      v3: v3,
      kendoLocalizationService: kendoLocalizationService
    };
  },
  // @ts-ignore
  render: function render(createElement) {
    var h = gh || createElement;
    var _a = this.$props,
      value = _a.value,
      pageSizes = _a.pageSizes,
      pageSize = _a.pageSize;
    var ls = (0, kendo_vue_intl_1.provideLocalizationService)(this);
    var sizes = pageSizes.slice();
    if (value === undefined && sizes.filter(function (s) {
      return s === pageSize;
    }).length === 0) {
      sizes.unshift(pageSize);
    }
    var itemPerPageMessage = this.$props.messagesMap ? this.$props.messagesMap(main_1.pagerItemPerPage) : {
      messageKey: main_1.pagerItemPerPage,
      defaultMessage: main_1.messages[main_1.pagerItemPerPage]
    };
    var pageSelectionMessage = this.$props.messagesMap ? this.$props.messagesMap(main_1.pagerPageSelection) : {
      messageKey: main_1.pagerPageSelection,
      defaultMessage: main_1.messages[main_1.pagerPageSelection]
    };
    return h("span", {
      "class": "k-pager-sizes k-label"
    }, [
    // @ts-ignore function children
    h(kendo_vue_dropdowns_1.DropDownList, {
      value: value !== undefined ? value : pageSize,
      attrs: this.v3 ? undefined : {
        value: value !== undefined ? value : pageSize,
        dataItems: sizes,
        size: this.$props.size,
        ariaLabel: ls.toLanguageString(pageSelectionMessage.messageKey, pageSelectionMessage.defaultMessage)
      },
      dataItems: sizes,
      size: this.$props.size,
      ariaLabel: ls.toLanguageString(pageSelectionMessage.messageKey, pageSelectionMessage.defaultMessage),
      onChange: this.pageSizeChange,
      on: this.v3 ? undefined : {
        "change": this.pageSizeChange
      }
    }), ls.toLanguageString(itemPerPageMessage.messageKey, itemPerPageMessage.defaultMessage)]);
  }
};
exports.PagerPageSizesVue2 = PagerPageSizesVue2;
/**
 * @hidden
 */
var PagerPageSizes = PagerPageSizesVue2;
exports.PagerPageSizes = PagerPageSizes;
