// @progress/kendo-vue-dropdowns/dist/npm/common/ClearButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ClearButtonVue2 = exports.ClearButton = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var inject = allVue.inject;
var kendo_svg_icons_1 = require("@progress/kendo-svg-icons");
var kendo_vue_intl_1 = require("@progress/kendo-vue-intl");
var main_1 = require("../messages/main");
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
/**
 * @hidden
 */
var ClearButtonVue2 = {
  name: 'clear-button',
  inject: {
    kendoLocalizationService: {
      default: null
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    var kendoLocalizationService = inject('kendoLocalizationService', {});
    return {
      v3: v3,
      kendoLocalizationService: kendoLocalizationService
    };
  },
  methods: {
    onMouseDown: function onMouseDown(e) {
      e.preventDefault();
    },
    onClickHandler: function onClickHandler(e) {
      this.$emit('clearclick', e);
    }
  },
  // @ts-ignore
  render: function render(createElement) {
    var h = gh || createElement;
    var title = (0, kendo_vue_intl_1.provideLocalizationService)(this).toLanguageString(main_1.clear, main_1.messages[main_1.clear]);
    return h("span", {
      "class": "k-clear-value",
      role: "button",
      attrs: this.v3 ? undefined : {
        role: "button",
        tabindex: -1,
        title: title
      },
      onClick: this.onClickHandler,
      on: this.v3 ? undefined : {
        "click": this.onClickHandler,
        "mousedown": this.onMouseDown
      },
      onMousedown: this.onMouseDown,
      tabindex: -1,
      title: title,
      key: "clearbutton"
    }, [h(kendo_vue_common_1.Icon, {
      name: "x",
      attrs: this.v3 ? undefined : {
        name: "x",
        icon: kendo_svg_icons_1.xIcon
      },
      icon: kendo_svg_icons_1.xIcon
    })]);
  }
};
exports.ClearButtonVue2 = ClearButtonVue2;
/**
 * @hidden
 */
var ClearButton = ClearButtonVue2;
exports.ClearButton = ClearButton;
