// @progress/kendo-vue-dropdowns/dist/npm/common/ListDefaultItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ListDefaultItemVue2 = exports.ListDefaultItem = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
var utils_1 = require("./utils");
/**
 * @hidden
 */
var ListDefaultItemVue2 = {
  name: 'list-default-item',
  props: {
    defaultItem: [Object, String],
    textField: String,
    selected: Boolean
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  methods: {
    onMouseDown: function onMouseDown(e) {
      e.preventDefault();
    },
    onClick: function onClick(e) {
      this.$emit('defaultitemclick', e);
    }
  },
  // @ts-ignore
  render: function render(createElement) {
    var h = gh || createElement;
    var _a = this.$props,
      selected = _a.selected,
      defaultItem = _a.defaultItem,
      textField = _a.textField;
    return h("div", {
      onClick: this.onClick,
      on: this.v3 ? undefined : {
        "click": this.onClick,
        "mousedown": this.onMouseDown
      },
      onMousedown: this.onMouseDown,
      style: {
        position: 'unset'
      },
      "class": (0, kendo_vue_common_1.classNames)('k-list-optionlabel', {
        'k-selected': selected
      })
    }, [(0, utils_1.getItemValue)(defaultItem, textField) || '']);
  }
};
exports.ListDefaultItemVue2 = ListDefaultItemVue2;
/**
 * @hidden
 */
var ListDefaultItem = ListDefaultItemVue2;
exports.ListDefaultItem = ListDefaultItem;
